---
title: "Sub-challenge 1"
author: "Herdiantri Sufriyana"
date: "2024-06-10"
output: html_document
---

# Programming environment

```{r Set random seed, include=FALSE, paged.print=FALSE}
seed <- 2024-06-10
```

```{r Load R packages, include=FALSE}
library(tidyverse)
library(knitr)
library(kableExtra)
library(ggpubr)
mutate <- dplyr::mutate
library(reticulate)
library(pbapply)
library(ChAMP)
reduce <- purrr::reduce
rename <- dplyr::rename
slice <- dplyr::slice
library(vroom)
library(parallel)
library(doParallel)
library(torch)
library(digest)
library(curl)
library(minfi)
library(GEOquery)
library(broom)
```

```{r Load custom functions, include=FALSE}
lapply(list.files("R/", pattern = "-function.R", full.names = TRUE), source)
```

```{r Import Python libraries, eval=FALSE, include=FALSE}
syn <- import("synapseclient")
```

```{r Set theme, include=FALSE}
dslabs::ds_theme_set()
```

# Download raw data

## This study training data

```{r Login to Synapse, eval=FALSE, include=FALSE}
synapse <- syn$Synapse()
synapse$login(authToken = "")
```

```{r Load a table of dataset list, include=FALSE}
syn59834055 <- read_csv("inst/extdata/syn59834055.csv", show_col_types = FALSE)
```

```{r table-s1, echo=FALSE}
syn59834055 |>
  mutate(
    size=
      case_when(
        size_unit == "KB" ~ 10^3*size
        ,size_unit == "MB" ~ 10^6*size
        ,size_unit == "GB" ~ 10^9*size
      )
  ) |>
  select(-size_unit) |>
  arrange(size) |>
  kable(caption = "Table S1. Raw data list") |>
  kable_classic()
```

```{r Download Preprocess.R, eval=FALSE, include=FALSE}
synapse$get("syn60236613", downloadLocation = "R/")
```

```{r Download Sample_annotation.csv, eval=FALSE, include=FALSE}
synapse$get("syn60157686", downloadLocation = "inst/extdata/")
```

```{r Download Probe_array.csv, eval=FALSE, include=FALSE}
synapse$get("syn60157718", downloadLocation = "inst/extdata/")
```

```{r Download Probe_annotation.csv, eval=FALSE, include=FALSE}
synapse$get("syn60157694", downloadLocation = "inst/extdata/")
```

```{r Download DetectionP_subchallenge1.csv, eval=FALSE, include=FALSE}
synapse$get("syn59870646", downloadLocation = "inst/extdata/")
```

```{r Download Beta_raw_subchallenge1.csv.gz, eval=FALSE, include=FALSE}
synapse$get("syn59868755", downloadLocation = "inst/extdata/")
```

## Download original series for understanding the data generation process

```{r Download GSE128827, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE128nnn/GSE128827/matrix/GSE128827_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE128827_series_matrix.txt.gz"
)
```

```{r Download GSE228149, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE228nnn/GSE228149/matrix/GSE228149_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE228149_series_matrix.txt.gz"
)
```

```{r Download GSE200659, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE200nnn/GSE200659/matrix/GSE200659_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE200659_series_matrix.txt.gz"
)
```

```{r Download E_MTAB_9312, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ebi.ac.uk/biostudies/fire/E-MTAB-/312/E-MTAB-9312/Files/E-MTAB-9312.sdrf.txt"
  ,destfile = "inst/extdata/E-MTAB-9312.sdrf.txt"
)
```

```{r Download GSE204977, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE204nnn/GSE204977/matrix/GSE204977_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE204977_series_matrix.txt.gz"
)
```

```{r Download GSE169598, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE169nnn/GSE169598/matrix/GSE169598_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE169598_series_matrix.txt.gz"
)
```

```{r Download GSE232778, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE232nnn/GSE232778/matrix/GSE232778_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE232778_series_matrix.txt.gz"
)
```

```{r Download GSE144129, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144129/matrix/GSE144129_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE144129_series_matrix.txt.gz"
)
```

```{r Download GSE167885, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE167nnn/GSE167885/matrix/GSE167885_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE167885_series_matrix.txt.gz"
)
```

```{r Download GSE75196, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE75nnn/GSE75196/matrix/GSE75196_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE75196_series_matrix.txt.gz"
)
```

```{r Download GSE108567, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE108nnn/GSE108567/matrix/GSE108567_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE108567_series_matrix.txt.gz"
)
```

```{r Download GSE69502, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE69nnn/GSE69502/matrix/GSE69502_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE69502_series_matrix.txt.gz"
)
```

```{r Download GSE74738, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE74nnn/GSE74738/matrix/GSE74738_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE74738_series_matrix.txt.gz"
)
```

```{r Download GSE100197, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100197/matrix/GSE100197_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE100197_series_matrix.txt.gz"
)
```

```{r Download GSE115508, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE115nnn/GSE115508/matrix/GSE115508_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE115508_series_matrix.txt.gz"
)
```

```{r Download GSE98224, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE98nnn/GSE98224/matrix/GSE98224-GPL13534_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE98224_series_matrix.txt.gz"
)
```

```{r Download GSE75248, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE75nnn/GSE75248/matrix/GSE75248_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE75248_series_matrix.txt.gz"
)
```

```{r Download GSE71678, eval=FALSE, include=FALSE}
resume_download(
  "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE71nnn/GSE71678/matrix/GSE71678_series_matrix.txt.gz"
  ,destfile = "inst/extdata/GSE71678_series_matrix.txt.gz"
)
```

# Load raw data

```{r Load the probe annotation file, eval=FALSE, include=FALSE}
probe_annotation <-
  vroom(
    "inst/extdata/Probe_annotation.csv"
    ,col_select = -1
    ,show_col_types = FALSE
  ) |>
  column_to_rownames(var = "Name")
```

```{r Load the methylation beta values, eval=FALSE, include=FALSE}
df_train0 <-
  vroom(
    "inst/extdata/Beta_raw_subchallenge1.csv.gz"
    ,show_col_types = FALSE
  ) |>
  column_to_rownames(var = "...1")
```

```{r Load the detection p-values, eval=FALSE, include=FALSE}
detp0 <-
  vroom(
    "inst/extdata/DetectionP_subchallenge1.csv"
    ,show_col_types = FALSE
  ) |>
  column_to_rownames(var = "...1")
```

```{r Load the sample annotation file, include=FALSE}
ano <-
  read_csv(
    "inst/extdata/Sample_annotation.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_Name = Sample_ID) |>
  column_to_rownames(var = "Sample_ID")
```

```{r Load the original sample annotation data, eval=FALSE, include=FALSE}
ano_ori <-
  list.files(
    "inst/extdata/"
    ,pattern = "_series_matrix\\.txt\\.gz"
    ,full.names = TRUE
  )

ano_ori <-
  ano_ori |>
  `names<-`(sapply(ano_ori, \(x) str_extract_all(x, "GSE[:digit:]+")[[1]]))

ano_ori <-
  ano_ori[
    ano |>
      filter(Dataset_ID != "E_MTAB_9312") |>
      pull(Dataset_ID) |>
      unique()
  ] |>
  pblapply(
    \(x)
      getGEO(
        filename = x
        ,GSEMatrix = FALSE
        ,parseCharacteristics = FALSE
        ,AnnotGPL = FALSE
        ,getGPL = FALSE
      )
  ) |>
  pblapply(pData)

ano_ori$E_MTAB_9312 <-
  vroom("inst/extdata/E-MTAB-9312.sdrf.txt", show_col_types = FALSE)

ano_ori <-
  ano_ori[
    ano |>
      pull(Dataset_ID) |>
      unique()
  ]


ano_ori <-
  ano_ori |>
  imap(
    ~ .x |>
      rename_at(
        colnames(.x)[
          str_detect(
            colnames(.x)
            ,"geo_accession|Assay Name"
          )
        ]
        ,\(x) "id"
      ) |>
      filter(
        str_remove_all(id, "_Grn$|_Red$")
        %in% filter(ano, Dataset_ID == .y)$Sample_accession
      ) |>
      arrange(
        factor(
          str_remove_all(id, "_Grn$|_Red$")
          ,filter(ano, Dataset_ID == .y)$Sample_accession
        )
      ) |>
      filter(!str_detect(id, "_Red$"))
  )

saveRDS(ano_ori, "data/ano_ori.rds")
```

```{r Load pre-loaded the original sample annotation data, include=FALSE}
ano_ori <- readRDS("data/ano_ori.rds")
```

```{r table-s3, echo=FALSE}
ano |>
  select(`GEO Number` = Dataset_ID) |>
  group_by(`GEO Number`) |>
  summarize(N = n()) |>
  arrange(N) |>
  kable(caption = "Table S3. This study training data list") |>
  kable_classic()
```

# Methylation data preprocessing

```{r Match train data to the sample order, eval=FALSE, include=FALSE}
df_train <-
  df_train0 |>
  select_at(rownames(ano))
```

```{r Match p-vals to the trainprobe & sample order, eval=FALSE, include=FALSE}
detp <-
  detp0[rownames(df_train), ] |>
  select_at(rownames(ano))
```

```{r Save input sampled-hash for validating p_vals, eval=FALSE, include=FALSE}
save_samp_md5(df_train, seed, "data/df_train_samp_md5.rds")
save_samp_md5(detp, seed, "data/detp_samp_md5.rds")
```

## Probe filtering

```{r Filter out probes using multiple criteria, eval=FALSE, include=FALSE}
train_filter <-
  df_train |>
  champ.filter(
    pd = ano
    ,filterDetP = FALSE
    ,detP = NULL
    ,autoimpute = FALSE
    ,filterBeads = FALSE
    ,arraytype = "450K"
    ,fixOutlier = FALSE
    ,filterNoCG = TRUE
    ,filterSNPs = TRUE #  Fall near SNPs
    ,filterMultiHit = TRUE # Align to multiple locations
    ,filterXY = TRUE # Target locations on X and Y chromosomes
  )
```

```{r Obtain the beta values of multiple criteria, eval=FALSE, include=FALSE}
beta_train_filter <-
  train_filter$beta
```

```{r Compute proportion of samples with p < 0.01, eval=FALSE, include=FALSE}
validate_samp_md5(df_train, seed, "data/df_train_samp_md5.rds")
validate_samp_md5(detp, seed, "data/detp_samp_md5.rds")
source("R/tensor_comp_p_val_lt_0_01.R")
```

```{r Load the pre-computed samples with p < 0.01, eval=FALSE, include=FALSE}
p_vals <- readRDS("data/p_vals.rds")
```

```{r Keep reliably-detected probes in all samples, eval=FALSE, include=FALSE}
keep <-
  rownames(detp)[p_vals == 1]
```

```{r Filter probes based on detection p-values, eval=FALSE, include=FALSE}
beta_train_filter_detp <-
  beta_train_filter[rownames(beta_train_filter) %in% keep, ]
```

```{r Save input sampled-hash for validating BMIQ, eval=FALSE, include=FALSE}
save_samp_md5(beta_train_filter_detp, seed, "data/beta_train_filter_detp.rds")
```

```{r Save input probe IDs for deploying BMIQ, eval=FALSE, include=FALSE}
rownames(beta_train_filter_detp) |>
  saveRDS("data/beta_train_filter_detp_rownames.rds")
```

## Probe normalization

```{r Normalize the filtered beta values using BMIQ, eval=FALSE, include=FALSE}
validate_samp_md5(
  beta_train_filter_detp
  ,seed
  ,"data/beta_train_filter_detp.rds"
)
source("R/parallel_bmiq_per_batch-codes.R")
```

```{r Load the pre-computed BMIQ, eval=FALSE, include=FALSE}
beta_norm_bmiq <-
  list.files("data/beta_norm_bmiq/", pattern = "batch", full.names = TRUE) |>
  pblapply(
    vroom_rds
    ,tmp_csv_path = "inst/extdata/tmp.csv"
    ,show_col_types = FALSE
  ) |>
  reduce(cbind) |>
  `rownames<-`(readRDS("data/beta_train_filter_detp_rownames.rds"))

saveRDS(beta_norm_bmiq, "data/beta_norm_bmiq.rds")
```

```{r Load pre-loaded the pre-computed BMIQ, eval=FALSE, include=FALSE}
beta_norm_bmiq <- readRDS("data/beta_norm_bmiq.rds")
```

```{r Save input sample IDs for ordering annotation, eval=FALSE, include=FALSE}
validate_samp_md5(
  beta_train_filter_detp
  ,seed
  ,"data/beta_train_filter_detp.rds"
)
saveRDS(colnames(df_train), "data/df_train_colnames.rds")
```

# Phenotype preprocessing

## Phenotype preprocessing for conditions

```{r Identify phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars0 <-
  ano_ori |>
  pblapply(parse_characteristics)
```

```{r Write old colnames of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars0 |>
  lapply(colnames) |>
  imap(~ data.frame(Dataset_ID = .y, old_colname = .x)) |>
  reduce(rbind) |>
  write_csv("inst/extdata/ano_ori_chars_old_colnames.csv")
```

```{r Read new colnames of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars_colnames <-
  read_csv(
    "inst/extdata/ano_ori_chars_new_colnames.csv"
    ,show_col_types = FALSE
  )
```

```{r Change colnames of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1 <-
  ano_ori_chars0 |>
  imap(
    ~ .x |>
      change_char_names(
        char_names = 
          ano_ori_chars_colnames |>
          filter(Dataset_ID == .y) |>
          select(old_colname, new_colname)
      )
  )
```

```{r Identify prob num cols of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1_num_col_prob <-
  ano_ori_chars1 |>
  bind_rows() |>
  mutate_all(\(x) suppressWarnings(as.numeric(x))) |>
  lapply(\(x) x[!is.na(x)]) |>
  sapply(\(x) length(x)>0) |>
  which() |>
  names()
```

```{r Identify cat cols among prob num cols, eval=FALSE, include=FALSE}
ano_ori_chars1_num_col_cat <-
  c("subject_id"
    ,"sentrix_id"
    ,"slide"
  )
```

```{r Identify num cols of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1_num_col <-
  ano_ori_chars1_num_col_prob |>
  setdiff(ano_ori_chars1_num_col_cat)
```

```{r Identify prob cat cols of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1_cat_col_prob <-
  ano_ori_chars1 |>
  bind_rows() |>
  mutate_all(\(x) suppressWarnings(as.numeric(x))) |>
  lapply(\(x) x[!is.na(x)]) |>
  sapply(\(x) length(x)==0) |>
  which() |>
  names()
```

```{r Identify id cols among prob cat cols, eval=FALSE, include=FALSE}
ano_ori_chars1_id_col <-
  c("subject_id"
    ,"sentrix_id"
    ,"slide"
    ,"plate"
    ,"batch"
    ,"sample_id"
    ,"batch_cohort"
    ,"batch_nih"
  )
```

```{r Identify cat cols of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1_cat_col <-
  ano_ori_chars1_cat_col_prob |>
  setdiff(ano_ori_chars1_id_col)
```

```{r Write old cats of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars1 |>
  lapply(lapply, unique) %>%
  imap(
    ~ .x |>
      imap(
        ~ data.frame(colname = .y, old_cat = .x)
      ) |>
      reduce(rbind) |>
      mutate(Dataset_ID = .y) |>
      select(Dataset_ID, everything())
  ) |>
  reduce(rbind) |>
  filter(colname %in% ano_ori_chars1_cat_col) |>
  write_csv("inst/extdata/ano_ori_chars_old_cats.csv")
```

```{r Read new cats of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars_cats <-
  read_csv(
    "inst/extdata/ano_ori_chars_new_cats.csv"
    ,show_col_types = FALSE
  )
```

```{r Change cats of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars2 <-
  ano_ori_chars1 |>
  imap(
    ~ .x |>
      change_char_cats(
        char_cats = 
          ano_ori_chars_cats |>
          filter(Dataset_ID == .y) |>
          select(colname, old_cat, new_cat)
      )
  )
```

```{r Write old conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars2 |>
  imap(
    ~ .x |>
      select_at(
        colnames(.x)[
          colnames(.x) |>
            str_detect("^condition")
        ]
      ) |>
      rownames_to_column(var = "Sample_accession") |>
      mutate(Dataset_ID = .y) |>
      select(Dataset_ID, Sample_accession, everything())
  ) |>
  bind_rows() |>
  gather(old_colname, old_condition, -Dataset_ID, -Sample_accession) |>
  filter(
    !is.na(old_condition)
    & !old_condition %in% c("Unspecified", "Replicate")
  ) |>
  select(Dataset_ID, old_colname, old_condition) |>
  unique() |>
  mutate_at(c("Dataset_ID", "old_colname"), \(x) factor(x, unique(x))) |>
  arrange(Dataset_ID, old_colname, old_condition) |>
  write_csv("inst/extdata/ano_ori_chars_old_conditions.csv")
```

```{r Read new conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars_conditions <-
  read_csv(
    "inst/extdata/ano_ori_chars_new_conditions.csv"
    ,show_col_types = FALSE
  )
```

```{r Change conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars3 <-
  ano_ori_chars2 |>
  imap(
    ~ .x |>
      change_char_conditions(
        char_conditions = 
          ano_ori_chars_conditions |>
          filter(Dataset_ID == .y) |>
          select(old_colname, old_condition, new_colname, new_condition)
      ) |>
      mutate_all(\(x) ifelse(x == "Unspecified", NA, x))
  ) |>
  imap(
    ~ .x |>
      rownames_to_column(var = "Sample_accession") |>
      right_join(
        ano_ori_chars2[[.y]] |>
          select_at(
            colnames(ano_ori_chars2[[.y]])[
              !str_detect(
                colnames(ano_ori_chars2[[.y]])
                ,"^condition"
              )
            ]
          ) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      ) |>
      column_to_rownames(var = "Sample_accession")
  )
```

```{r Determine final phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars <- ano_ori_chars3
saveRDS(ano_ori_chars, "data/ano_ori_chars.rds")
```

```{r Load pre-determined final phenotype chars, eval=FALSE, include=FALSE}
ano_ori_chars <- readRDS("data/ano_ori_chars.rds")
```

```{r Obtain conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_conds <-
  ano_ori_chars |>
  bind_rows()

ano_ori_conds <-
  ano_ori_conds |>
  select_at(
    ano_ori_conds |>
      colnames() |>
      str_detect("^y_") |>
      which()
  )

ano_ori_conds <-
  ano_ori_conds |>
  `colnames<-`(
    ano_ori_conds |>
      colnames() |>
      str_remove_all("^y_")
  )
```

```{r GA and conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_ga_conds0 <-
  ano[readRDS("data/df_train_colnames.rds"), ] |>
  left_join(
    ano_ori_conds |>
      rownames_to_column(var = "Sample_accession")
    , by = join_by(Sample_accession)
  ) |>
  mutate(
    pe_onset =
      ifelse(
        pe == "No"
        ,"Not applicable"
        ,pe_onset
      )
    ,hellp =
      ifelse(
        pe == "No"
        ,"No"
        ,hellp
      )
    ,anencephaly =
      ifelse(
        is.na(anencephaly)
        ,ifelse(
          Dataset_ID == "GSE69502"
          ,anencephaly
          ,"No"
        )
        ,anencephaly
      )
    ,spina_bifida =
      ifelse(
        is.na(spina_bifida)
        ,ifelse(
          Dataset_ID == "GSE69502"
          ,spina_bifida
          ,"No"
        )
        ,spina_bifida
      )
    ,diandric_triploid =
      ifelse(
        is.na(diandric_triploid)
        ,ifelse(
          Dataset_ID == "GSE74738"
          ,diandric_triploid
          ,"No"
        )
        ,diandric_triploid
      )
    ,miscarriage =
      ifelse(
        is.na(miscarriage)
        ,ifelse(
          GA < 13
          ,miscarriage
          ,"No"
        )
        ,miscarriage
      )
    ,preterm =
      ifelse(
        !is.na(GA)
        ,ifelse(
            GA < 37
            ,ifelse(
                anencephaly == "Yes"
                | spina_bifida == "Yes"
                | miscarriage == "Yes"
                | diandric_triploid == "Yes"
                ,preterm
                ,"Yes"
              )
            ,"No"
          )
        ,preterm
      )
  ) |>
  mutate(ceiled_GA = ceiling(GA)) |>
  column_to_rownames(var = "Sample_accession") |>
  select(-Array, -Sample_Name)
```

## Phenotype imputation for conditions

### FGR imputation

```{r FGR imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
fgr <- list()

fgr$data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(fgr)) |>
  select(Dataset_ID, outcome = fgr)

fgr <- imputation(fgr, "Identify datasets")

fgr <- imputation(fgr, "Identify DMPs", arraytype = "450K")
fgr_dmp <- fgr$dmp
saveRDS(fgr_dmp, "data/fgr_dmp.rds")
fgr$dmp_comparison <- "Yes_to_No"
fgr$dmp_outcome_cat <- "Yes"

fgr <- imputation(fgr, "Imputation train set")

fgr$imp_train_set |>
  write_csv("inst/extdata/fgr_imp_train_set.csv")

fgr$imp_train_set |>
  rownames() |>
  saveRDS("data/fgr_imp_train_set_rownames.rds")

fgr$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(fgr)) |>
  rownames()

fgr <- imputation(fgr, "Identify imputee data")

fgr <- imputation(fgr, "Imputation set")

fgr$imp_set |>
  write_csv("inst/extdata/fgr_imp_set.csv")

fgr$imp_set |>
  rownames() |>
  saveRDS("data/fgr_imp_set_rownames.rds")
```

```{r FGR imputation - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
fgr_dmp <- readRDS("data/fgr_dmp.rds")

data.frame(probe_id = rownames(fgr_dmp[[1]])) |>
  write_csv("inst/extdata/fgr_dmp.csv")
```

```{r FGR imputation - Load pre-written DMPs, include=FALSE}
fgr_dmp <- list()

fgr_dmp$Yes_to_No <-
  read_csv("inst/extdata/fgr_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r FGR imputation - Read imputation results, include=FALSE}
fgr_imp_set_predictions <-
  read_csv("inst/extdata/fgr_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/fgr_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### PE imputation

```{r PE imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
pe <- list()

pe$data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(pe)) |>
  select(Dataset_ID, outcome = pe)

pe <- imputation(pe, "Identify datasets")

pe <- imputation(pe, "Identify DMPs", arraytype = "450K")
pe_dmp <- pe$dmp
saveRDS(pe_dmp, "data/pe_dmp.rds")
pe$dmp_comparison <- "Yes_to_No"
pe$dmp_outcome_cat <- "Yes"

pe <- imputation(pe, "Imputation train set")

pe$imp_train_set |>
  write_csv("inst/extdata/pe_imp_train_set.csv")

pe$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(pe)) |>
  rownames()

pe <- imputation(pe, "Identify imputee data")

pe <- imputation(pe, "Imputation set")

pe$imp_set |>
  write_csv("inst/extdata/pe_imp_set.csv")

pe$imp_set |>
  rownames() |>
  saveRDS("data/pe_imp_set_rownames.rds")
```

```{r PE prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
pe_dmp <- readRDS("data/pe_dmp.rds")

data.frame(probe_id = rownames(pe_dmp[[1]])) |>
  write_csv("inst/extdata/pe_dmp.csv")
```

```{r PE imputation - Load pre-written DMPs, include=FALSE}
pe_dmp <- list()

pe_dmp$Yes_to_No <-
  read_csv("inst/extdata/pe_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r PE imputation - Read imputation results, include=FALSE}
pe_imp_set_predictions <-
  read_csv("inst/extdata/pe_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/pe_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### PE onset imputation

```{r PE onset imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
pe_onset <- list()

pe_onset$data0 <-
  ano_ori_ga_conds0 |>
  filter(!(is.na(pe_onset) | pe_onset == "Not applicable")) |>
  select(Dataset_ID, outcome = pe_onset)

pe_onset <- imputation(pe_onset, "Identify datasets")

pe_onset <- imputation(pe_onset, "Identify DMPs", arraytype = "450K")
pe_onset_dmp <- pe_onset$dmp
saveRDS(pe_onset_dmp, "data/pe_onset_dmp.rds")
pe_onset$dmp_comparison <- "Early_to_Late"
pe_onset$dmp_outcome_cat <- "Early"

pe_onset <- imputation(pe_onset, "Imputation train set")

pe_onset$imp_train_set |>
  write_csv("inst/extdata/pe_onset_imp_train_set.csv")

pe_onset$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(pe_onset)) |>
  rownames()

pe_onset <- imputation(pe_onset, "Identify imputee data")

pe_onset <- imputation(pe_onset, "Imputation set")

pe_onset$imp_set |>
  write_csv("inst/extdata/pe_onset_imp_set.csv")

pe_onset$imp_set |>
  rownames() |>
  saveRDS("data/pe_onset_imp_set_rownames.rds")
```

```{r PE onset pred - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
pe_onset_dmp <- readRDS("data/pe_onset_dmp.rds")

data.frame(probe_id = rownames(pe_onset_dmp[[1]])) |>
  write_csv("inst/extdata/pe_onset_dmp.csv")
```

```{r PE onset imputation - Load pre-written DMPs, include=FALSE}
pe_onset_dmp <- list()

pe_onset_dmp$Early_to_Late <-
  read_csv("inst/extdata/pe_onset_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r PE onset imputation - Read imputation results, include=FALSE}
pe_onset_imp_set_predictions <-
  read_csv(
    "inst/extdata/pe_onset_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/pe_onset_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### HELLP imputation

```{r HELLP imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
hellp <- list()

hellp$data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(hellp)) |>
  select(Dataset_ID, outcome = hellp)

hellp$data0b <-
  ano_ori_ga_conds0 |>
  filter(pe == "Yes") |>
  filter(!is.na(hellp)) |>
  select(Dataset_ID, outcome = hellp)

hellp <- imputation(hellp, "Identify datasets")

hellp <- imputation(hellp, "Identify DMPs", arraytype = "450K")
hellp_dmp <- hellp$dmp
saveRDS(hellp_dmp, "data/hellp_dmp.rds")
hellp$dmp_comparison <- "No_to_Yes"
hellp$dmp_outcome_cat <- "Yes"

hellp <- imputation(hellp, "Imputation train set")

hellp$imp_train_set <-
  hellp$imp_train_set[rownames(hellp$data0b), , drop = FALSE]

hellp$imp_train_set |>
  write_csv("inst/extdata/hellp_imp_train_set.csv")

hellp$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(hellp)) |>
  rownames()

hellp <- imputation(hellp, "Identify imputee data")

hellp <- imputation(hellp, "Imputation set")

hellp$imp_set |>
  write_csv("inst/extdata/hellp_imp_set.csv")

hellp$imp_set |>
  rownames() |>
  saveRDS("data/hellp_imp_set_rownames.rds")
```

```{r HELLP prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
hellp_dmp <- readRDS("data/hellp_dmp.rds")

data.frame(probe_id = rownames(hellp_dmp[[1]])) |>
  write_csv("inst/extdata/hellp_dmp.csv")
```

```{r HELLP imputation - Load pre-written DMPs, include=FALSE}
hellp_dmp <- list()

hellp_dmp$No_to_Yes <-
  read_csv("inst/extdata/hellp_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r HELLP imputation - Read imputation results, include=FALSE}
hellp_imp_set_predictions <-
  read_csv(
    "inst/extdata/hellp_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/hellp_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Diandric triploid imputation

```{r Diatri imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
diandric_triploid <- list()

diandric_triploid$data0 <-
  ano_ori_ga_conds0 |>
  filter(Dataset_ID == "GSE74738") |>
  filter(!is.na(diandric_triploid)) |>
  select(Dataset_ID, outcome = diandric_triploid)

diandric_triploid <- imputation(diandric_triploid, "Identify datasets")

diandric_triploid <-
  imputation(diandric_triploid, "Identify DMPs", arraytype = "450K")

diandric_triploid_dmp <- diandric_triploid$dmp
saveRDS(diandric_triploid_dmp, "data/diandric_triploid_dmp.rds")
diandric_triploid$dmp_comparison <- "No_to_Yes"
diandric_triploid$dmp_outcome_cat <- "Yes"

diandric_triploid <- imputation(diandric_triploid, "Imputation train set")

diandric_triploid$imp_train_set |>
  write_csv("inst/extdata/diandric_triploid_imp_train_set.csv")

diandric_triploid$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(diandric_triploid)) |>
  rownames()

diandric_triploid <- imputation(diandric_triploid, "Identify imputee data")

diandric_triploid <- imputation(diandric_triploid, "Imputation set")

diandric_triploid$imp_set |>
  write_csv("inst/extdata/diandric_triploid_imp_set.csv")

diandric_triploid$imp_set |>
  rownames() |>
  saveRDS("data/diandric_triploid_imp_set_rownames.rds")
```

```{r Diatri prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
diandric_triploid_dmp <- readRDS("data/diandric_triploid_dmp.rds")

data.frame(probe_id = rownames(diandric_triploid_dmp[[1]])) |>
  write_csv("inst/extdata/diandric_triploid_dmp.csv")
```

```{r Diatri imputation - Load pre-written DMPs, include=FALSE}
diandric_triploid_dmp <- list()

diandric_triploid_dmp$No_to_Yes <-
  read_csv("inst/extdata/diandric_triploid_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Diandric triploid imputation - Read imputation results, include=FALSE}
diandric_triploid_imp_set_predictions <-
  read_csv(
    "inst/extdata/diandric_triploid_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/diandric_triploid_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Miscarriage imputation

```{r Miscar imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
miscarriage <- list()

miscarriage$data0 <-
  ano_ori_ga_conds0 |>
  filter(Dataset_ID == "GSE74738") |>
  filter(!is.na(miscarriage)) |>
  select(Dataset_ID, outcome = miscarriage)

miscarriage <- imputation(miscarriage, "Identify datasets")

miscarriage <- imputation(miscarriage, "Identify DMPs", arraytype = "450K")
miscarriage_dmp <- miscarriage$dmp
saveRDS(miscarriage_dmp, "data/miscarriage_dmp.rds")
miscarriage$dmp_comparison <- "No_to_Yes"
miscarriage$dmp_outcome_cat <- "Yes"

miscarriage <- imputation(miscarriage, "Imputation train set")

miscarriage$imp_train_set |>
  write_csv("inst/extdata/miscarriage_imp_train_set.csv")

miscarriage$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(miscarriage)) |>
  rownames()

miscarriage <- imputation(miscarriage, "Identify imputee data")

miscarriage <- imputation(miscarriage, "Imputation set")

miscarriage$imp_set |>
  write_csv("inst/extdata/miscarriage_imp_set.csv")

miscarriage$imp_set |>
  rownames() |>
  saveRDS("data/miscarriage_imp_set_rownames.rds")
```

```{r Miscar prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
miscarriage_dmp <- readRDS("data/miscarriage_dmp.rds")

data.frame(probe_id = rownames(miscarriage_dmp[[1]])) |>
  write_csv("inst/extdata/miscarriage_dmp.csv")
```

```{r Miscar imputation - Load pre-written DMPs, include=FALSE}
miscarriage_dmp <- list()

miscarriage_dmp$No_to_Yes <-
  read_csv("inst/extdata/miscarriage_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Miscarriage imputation - Read imputation results, include=FALSE}
miscarriage_imp_set_predictions <-
  read_csv(
    "inst/extdata/miscarriage_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/miscarriage_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Preterm imputation

```{r Preterm imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
preterm <- list()

preterm$data0 <-
  ano_ori_ga_conds0 |>
  filter(Dataset_ID == "GSE100197") |>
  filter(!is.na(preterm)) |>
  select(Dataset_ID, outcome = preterm)

preterm <- imputation(preterm, "Identify datasets")

preterm <- imputation(preterm, "Identify DMPs", arraytype = "450K")
preterm_dmp <- preterm$dmp
saveRDS(preterm_dmp, "data/preterm_dmp.rds")
preterm$dmp_comparison <- "Yes_to_No"
preterm$dmp_outcome_cat <- "Yes"

preterm <- imputation(preterm, "Imputation train set")

preterm$imp_train_set |>
  write_csv("inst/extdata/preterm_imp_train_set.csv")

preterm$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(preterm)) |>
  rownames()

preterm <- imputation(preterm, "Identify imputee data")

preterm <- imputation(preterm, "Imputation set")

preterm$imp_set |>
  write_csv("inst/extdata/preterm_imp_set.csv")

preterm$imp_set |>
  rownames() |>
  saveRDS("data/preterm_imp_set_rownames.rds")
```

```{r Pret prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
preterm_dmp <- readRDS("data/preterm_dmp.rds")

data.frame(probe_id = rownames(preterm_dmp[[1]])) |>
  write_csv("inst/extdata/preterm_dmp.csv")
```

```{r Pret imputation - Load pre-written DMPs, include=FALSE}
preterm_dmp <- list()

preterm_dmp$Yes_to_No <-
  read_csv("inst/extdata/preterm_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Preterm imputation - Read imputation results, include=FALSE}
preterm_imp_set_predictions <-
  read_csv(
    "inst/extdata/preterm_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/preterm_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GDM imputation

```{r GDM imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
gdm <- list()

gdm$data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(gdm)) |>
  select(Dataset_ID, outcome = gdm)

gdm <- imputation(gdm, "Identify datasets")

gdm <- imputation(gdm, "Identify DMPs", arraytype = "450K")
gdm_dmp <- gdm$dmp
saveRDS(gdm_dmp, "data/gdm_dmp.rds")
gdm$dmp_comparison <- "No_to_Yes"
gdm$dmp_outcome_cat <- "Yes"

gdm <- imputation(gdm, "Imputation train set")

gdm$imp_train_set |>
  write_csv("inst/extdata/gdm_imp_train_set.csv")

gdm$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(gdm)) |>
  rownames()

gdm <- imputation(gdm, "Identify imputee data")

gdm <- imputation(gdm, "Imputation set")

gdm$imp_set |>
  write_csv("inst/extdata/gdm_imp_set.csv")

gdm$imp_set |>
  rownames() |>
  saveRDS("data/gdm_imp_set_rownames.rds")
```

```{r GDM prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
gdm_dmp <- readRDS("data/gdm_dmp.rds")

data.frame(probe_id = rownames(gdm_dmp[[1]])) |>
  write_csv("inst/extdata/gdm_dmp.csv")
```

```{r GDM imputation - Load pre-written DMPs, include=FALSE}
gdm_dmp <- list()

gdm_dmp$No_to_Yes <-
  read_csv("inst/extdata/gdm_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r GDM imputation - Read imputation results, include=FALSE}
gdm_imp_set_predictions <-
  read_csv("inst/extdata/gdm_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/gdm_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### SGA imputation

```{r SGA imputation - Identify ds-DMPs-imp, eval=FALSE, include=FALSE}
sga <- list()

sga$data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(sga)) |>
  select(Dataset_ID, outcome = sga)

sga <- imputation(sga, "Identify datasets")

sga <- imputation(sga, "Identify DMPs", arraytype = "450K")
sga_dmp <- sga$dmp
saveRDS(sga_dmp, "data/sga_dmp.rds")
sga$dmp_comparison <- "No_to_Yes"
sga$dmp_outcome_cat <- "Yes"

sga <- imputation(sga, "Imputation train set")

sga$imp_train_set |>
  write_csv("inst/extdata/sga_imp_train_set.csv")

sga$data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(sga)) |>
  rownames()

sga <- imputation(sga, "Identify imputee data")

sga <- imputation(sga, "Imputation set")

sga$imp_set |>
  write_csv("inst/extdata/sga_imp_set.csv")

sga$imp_set |>
  rownames() |>
  saveRDS("data/sga_imp_set_rownames.rds")
```

```{r SGA prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
sga_dmp <- readRDS("data/sga_dmp.rds")

data.frame(probe_id = rownames(sga_dmp[[1]])) |>
  write_csv("inst/extdata/sga_dmp.csv")
```

```{r SGA imputation - Load pre-written DMPs, include=FALSE}
sga_dmp <- list()

sga_dmp$No_to_Yes <-
  read_csv("inst/extdata/sga_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r SGA imputation - Read imputation results, include=FALSE}
sga_imp_set_predictions <-
  read_csv("inst/extdata/sga_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/sga_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### LGA imputation

```{r LGA imputation - Identify datasets, eval=FALSE, include=FALSE}
lga_data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(lga)) |>
  select(Dataset_ID, outcome = lga)

lga_data <-
  lga_data0 |>
  pull(Dataset_ID) |>
  unique()

lga_data <-
  lga_data |>
  `names<-`(lga_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        lga_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            lga_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r LGA imputation - Identify DMPs, eval=FALSE, include=FALSE}
lga_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[, bind_rows(lga_data)$Sample_Name, drop = FALSE] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      lga_data0[bind_rows(lga_data)$Sample_accession, ,drop = FALSE] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(lga_dmp, "data/lga_dmp.rds")
```

```{r LGA prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
lga_dmp <- readRDS("data/lga_dmp.rds")

data.frame(probe_id = rownames(lga_dmp[[1]])) |>
  write_csv("inst/extdata/lga_dmp.csv")
```

```{r LGA imputation - Load pre-written DMPs, include=FALSE}
lga_dmp <- list()

lga_dmp$No_to_Yes <-
  read_csv("inst/extdata/lga_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r LGA imputation - Imputation train set, eval=FALSE, include=FALSE}
lga_imp_train_set <-
  beta_norm_bmiq[
    rownames(lga_dmp$No_to_Yes)
    ,bind_rows(lga_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(lga_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    lga_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r LGA imputation - Write imputation train set, eval=FALSE, include=FALSE}
lga_imp_train_set |>
  write_csv("inst/extdata/lga_imp_train_set.csv")
```

```{r LGA imputation - Identify imputee data, eval=FALSE, include=FALSE}
lga_data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(lga)) |>
  rownames()

lga_data_imputee <-
  list(
    Sample_accession =
      lga_data_imputee
    ,Sample_Name =
      ano |>
      filter(
        Sample_accession
        %in% c(
          lga_data_imputee
        )
      ) |>
      pull(Sample_Name)
  )
```

```{r LGA imputation - Imputation set, eval=FALSE, include=FALSE}
lga_imp_set <-
  beta_norm_bmiq[
    rownames(lga_dmp$No_to_Yes)
    ,lga_data_imputee$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(lga_data_imputee$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r LGA imputation - Write imputation set, eval=FALSE, include=FALSE}
lga_imp_set |>
  write_csv("inst/extdata/lga_imp_set.csv")

lga_imp_set |>
  rownames() |>
  saveRDS("data/lga_imp_set_rownames.rds")
```

```{r LGA imputation - Read imputation results, include=FALSE}
lga_imp_set_predictions <-
  read_csv("inst/extdata/lga_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/lga_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### SLGA imputation

```{r SLGA imputation - Identify datasets, eval=FALSE, include=FALSE}
slga_data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(sga) & !is.na(lga)) |>
  filter(sga == "Yes" | lga == "Yes") |>
  mutate(slga = ifelse(sga == "Yes", "Yes", "No")) |>
  select(Dataset_ID, outcome = slga)

slga_data <-
  slga_data0 |>
  pull(Dataset_ID) |>
  unique()

slga_data <-
  slga_data |>
  `names<-`(slga_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        slga_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            slga_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r SLGA imputation - Identify DMPs, eval=FALSE, include=FALSE}
slga_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[, bind_rows(slga_data)$Sample_Name, drop = FALSE] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      slga_data0[bind_rows(slga_data)$Sample_accession, ,drop = FALSE] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(slga_dmp, "data/slga_dmp.rds")
```

```{r SLGA prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
slga_dmp <- readRDS("data/slga_dmp.rds")

data.frame(probe_id = rownames(slga_dmp[[1]])) |>
  write_csv("inst/extdata/slga_dmp.csv")
```

```{r SLGA imputation - Load pre-written DMPs, include=FALSE}
slga_dmp <- list()

slga_dmp$Yes_to_No <-
  read_csv("inst/extdata/slga_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r SLGA imputation - Imputation train set, eval=FALSE, include=FALSE}
slga_imp_train_set <-
  beta_norm_bmiq[
    rownames(slga_dmp$Yes_to_No)
    ,bind_rows(slga_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(slga_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    slga_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r SLGA imputation - Write imputation train set, eval=FALSE, include=FALSE}
slga_imp_train_set |>
  write_csv("inst/extdata/slga_imp_train_set.csv")
```

```{r SLGA imputation - Identify imputee data, eval=FALSE, include=FALSE}
slga_data_imputee <-
  ano_ori_ga_conds0 |>
  filter(!(!is.na(sga) & !is.na(lga))) |>
  rownames()

slga_data_imputee <-
  list(
    Sample_accession =
      slga_data_imputee
    ,Sample_Name =
      ano |>
      filter(
        Sample_accession
        %in% c(
          slga_data_imputee
        )
      ) |>
      pull(Sample_Name)
  )
```

```{r SLGA imputation - Imputation set, eval=FALSE, include=FALSE}
slga_imp_set <-
  beta_norm_bmiq[
    rownames(slga_dmp$Yes_to_No)
    ,slga_data_imputee$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(slga_data_imputee$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r SLGA imputation - Write imputation set, eval=FALSE, include=FALSE}
slga_imp_set |>
  write_csv("inst/extdata/slga_imp_set.csv")

slga_imp_set |>
  rownames() |>
  saveRDS("data/slga_imp_set_rownames.rds")
```

```{r SLGA imputation - Read imputation results, include=FALSE}
slga_imp_set_predictions <-
  read_csv("inst/extdata/slga_imp_predictions.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/slga_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### IVF imputation

```{r IVF imputation - Identify datasets, eval=FALSE, include=FALSE}
ivf_data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(ivf)) |>
  select(Dataset_ID, outcome = ivf)

ivf_data <-
  ivf_data0 |>
  pull(Dataset_ID) |>
  unique()

ivf_data <-
  ivf_data |>
  `names<-`(ivf_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ivf_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ivf_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r IVF imputation - Identify DMPs, eval=FALSE, include=FALSE}
ivf_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ivf_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ivf_data0[
        bind_rows(ivf_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ivf_dmp, "data/ivf_dmp.rds")
```

```{r IVF prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
ivf_dmp <- readRDS("data/ivf_dmp.rds")

data.frame(probe_id = rownames(ivf_dmp[[1]])) |>
  write_csv("inst/extdata/ivf_dmp.csv")
```

```{r IVF imputation - Load pre-written DMPs, include=FALSE}
ivf_dmp <- list()

ivf_dmp$No_to_Yes <-
  read_csv("inst/extdata/ivf_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r IVF imputation - Imputation train set, eval=FALSE, include=FALSE}
ivf_imp_train_set <-
  beta_norm_bmiq[
    rownames(ivf_dmp$No_to_Yes)
    ,bind_rows(ivf_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ivf_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ivf_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r IVF imputation - Write imputation train set, eval=FALSE, include=FALSE}
ivf_imp_train_set |>
  write_csv("inst/extdata/ivf_imp_train_set.csv")
```

```{r IVF imputation - Identify imputee data, eval=FALSE, include=FALSE}
ivf_data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(ivf)) |>
  rownames()

ivf_data_imputee <-
  list(
    Sample_accession =
      ivf_data_imputee
    ,Sample_Name =
      ano |>
      filter(
        Sample_accession
        %in% c(
          ivf_data_imputee
        )
      ) |>
      pull(Sample_Name)
  )
```

```{r IVF imputation - Imputation set, eval=FALSE, include=FALSE}
ivf_imp_set <-
  beta_norm_bmiq[
    rownames(ivf_dmp$No_to_Yes)
    ,ivf_data_imputee$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(ivf_data_imputee$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r IVF imputation - Write imputation set, eval=FALSE, include=FALSE}
ivf_imp_set |>
  write_csv("inst/extdata/ivf_imp_set.csv")

ivf_imp_set |>
  rownames() |>
  saveRDS("data/ivf_imp_set_rownames.rds")
```

```{r IVF imputation - Read imputation results, include=FALSE}
ivf_imp_set_predictions <-
  read_csv(
    "inst/extdata/ivf_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ivf_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Subfertility imputation

```{r Subfertility imputation - Identify datasets, eval=FALSE, include=FALSE}
subfertility_data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(subfertility)) |>
  select(Dataset_ID, outcome = subfertility)

subfertility_data <-
  subfertility_data0 |>
  pull(Dataset_ID) |>
  unique()

subfertility_data <-
  subfertility_data |>
  `names<-`(subfertility_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        subfertility_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            subfertility_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Subfertility imputation - Identify DMPs, eval=FALSE, include=FALSE}
subfertility_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(subfertility_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      subfertility_data0[
        bind_rows(subfertility_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
    ,adjPVal = 0.001
    ,adjust.method = "none"
  )

saveRDS(subfertility_dmp, "data/subfertility_dmp.rds")
```

```{r Subfe prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
subfertility_dmp <- readRDS("data/subfertility_dmp.rds")

data.frame(probe_id = rownames(subfertility_dmp[[1]])) |>
  write_csv("inst/extdata/subfertility_dmp.csv")
```

```{r Subfe imputation - Load pre-written DMPs, include=FALSE}
subfertility_dmp <- list()

subfertility_dmp$No_to_Yes <-
  read_csv("inst/extdata/subfertility_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Subfertility imputation - Imputation train set, eval=FALSE, include=FALSE}
subfertility_imp_train_set <-
  beta_norm_bmiq[
    rownames(subfertility_dmp$No_to_Yes)
    ,bind_rows(subfertility_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(subfertility_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    subfertility_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Subfe imputation - Write imputation train set, eval=FALSE, include=FALSE}
subfertility_imp_train_set |>
  write_csv("inst/extdata/subfertility_imp_train_set.csv")
```

```{r Subfe imputation - Identify imputee data, eval=FALSE, include=FALSE}
subfertility_data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(subfertility)) |>
  rownames()

subfertility_data_imputee <-
  list(
    Sample_accession =
      subfertility_data_imputee
    ,Sample_Name =
      ano |>
      filter(
        Sample_accession
        %in% c(
          subfertility_data_imputee
        )
      ) |>
      pull(Sample_Name)
  )
```

```{r Subfertility imputation - Imputation set, eval=FALSE, include=FALSE}
subfertility_imp_set <-
  beta_norm_bmiq[
    rownames(subfertility_dmp$No_to_Yes)
    ,subfertility_data_imputee$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(subfertility_data_imputee$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Subfertility imputation - Write imputation set, eval=FALSE, include=FALSE}
subfertility_imp_set |>
  write_csv("inst/extdata/subfertility_imp_set.csv")

subfertility_imp_set |>
  rownames() |>
  saveRDS("data/subfertility_imp_set_rownames.rds")
```

```{r Subfertility imputation - Read imputation results, include=FALSE}
subfertility_imp_set_predictions <-
  read_csv(
    "inst/extdata/subfertility_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/subfertility_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Chorioamnionitis imputation

```{r Chor imputation - Identify datasets, eval=FALSE, include=FALSE}
chorioamnionitis_data0 <-
  ano_ori_ga_conds0 |>
  filter(!is.na(chorioamnionitis)) |>
  select(Dataset_ID, outcome = chorioamnionitis)

chorioamnionitis_data <-
  chorioamnionitis_data0 |>
  pull(Dataset_ID) |>
  unique()

chorioamnionitis_data <-
  chorioamnionitis_data |>
  `names<-`(chorioamnionitis_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        chorioamnionitis_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            chorioamnionitis_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Chorioamnionitis imputation - Identify DMPs, eval=FALSE, include=FALSE}
chorioamnionitis_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(chorioamnionitis_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      chorioamnionitis_data0[
        bind_rows(chorioamnionitis_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(chorioamnionitis_dmp, "data/chorioamnionitis_dmp.rds")
```

```{r Chori prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
chorioamnionitis_dmp <- readRDS("data/chorioamnionitis_dmp.rds")

data.frame(probe_id = rownames(chorioamnionitis_dmp[[1]])) |>
  write_csv("inst/extdata/chorioamnionitis_dmp.csv")
```

```{r Chori imputation - Load pre-written DMPs, include=FALSE}
chorioamnionitis_dmp <- list()

chorioamnionitis_dmp$No_to_Yes <-
  read_csv("inst/extdata/chorioamnionitis_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Chor imputation - Imputation train set, eval=FALSE, include=FALSE}
chorioamnionitis_imp_train_set <-
  beta_norm_bmiq[
    rownames(chorioamnionitis_dmp$No_to_Yes)
    ,bind_rows(chorioamnionitis_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(chorioamnionitis_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    chorioamnionitis_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Chor imputation - Write imputation train set, eval=FALSE, include=FALSE}
chorioamnionitis_imp_train_set |>
  write_csv("inst/extdata/chorioamnionitis_imp_train_set.csv")
```

```{r Chor imputation - Identify imputee data, eval=FALSE, include=FALSE}
chorioamnionitis_data_imputee <-
  ano_ori_ga_conds0 |>
  filter(is.na(chorioamnionitis)) |>
  rownames()

chorioamnionitis_data_imputee <-
  list(
    Sample_accession =
      chorioamnionitis_data_imputee
    ,Sample_Name =
      ano |>
      filter(
        Sample_accession
        %in% c(
          chorioamnionitis_data_imputee
        )
      ) |>
      pull(Sample_Name)
  )
```

```{r Chorioamnionitis imputation - Imputation set, eval=FALSE, include=FALSE}
chorioamnionitis_imp_set <-
  beta_norm_bmiq[
    rownames(chorioamnionitis_dmp$No_to_Yes)
    ,chorioamnionitis_data_imputee$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(chorioamnionitis_data_imputee$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Chor imputation - Write imputation set, eval=FALSE, include=FALSE}
chorioamnionitis_imp_set |>
  write_csv("inst/extdata/chorioamnionitis_imp_set.csv")

chorioamnionitis_imp_set |>
  rownames() |>
  saveRDS("data/chorioamnionitis_imp_set_rownames.rds")
```

```{r Chorioamnionitis imputation - Read imputation results, include=FALSE}
chorioamnionitis_imp_set_predictions <-
  read_csv(
    "inst/extdata/chorioamnionitis_imp_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/chorioamnionitis_imp_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### Apply imputation

```{r Apply imputation, eval=FALSE, include=FALSE}
ano_ori_ga_conds <-
  ano_ori_ga_conds0 |>
  impute_by_pred("fgr", fgr_imp_set_predictions) |>
  impute_by_pred("pe", pe_imp_set_predictions) |>
  mutate(
    pe_onset =
      ifelse(
        pe == "No"
        ,"Not applicable"
        ,pe_onset
      )
    ,hellp =
      ifelse(
        pe == "No"
        ,"No"
        ,hellp
      )
  ) |>
  impute_by_pred("pe_onset", pe_onset_imp_set_predictions, "Early", "Late") |>
  impute_by_pred("hellp", hellp_imp_set_predictions) |>
  
  impute_by_pred("diandric_triploid", diandric_triploid_imp_set_predictions) |>
  impute_by_pred("miscarriage", miscarriage_imp_set_predictions) |>
  mutate(
    preterm =
      ifelse(
        !is.na(GA)
        ,ifelse(
            GA < 37
            ,ifelse(
                anencephaly == "Yes"
                | spina_bifida == "Yes"
                | miscarriage == "Yes"
                | diandric_triploid == "Yes"
                ,preterm
                ,"Yes"
              )
            ,"No"
          )
        ,preterm
      )
  ) |>
  impute_by_pred("preterm", preterm_imp_set_predictions) |>
  
  impute_by_pred("gdm", gdm_imp_set_predictions) |>
  
  impute_by_pred("sga", sga_imp_set_predictions) |>
  impute_by_pred("lga", lga_imp_set_predictions) |>
  mutate(slga = ifelse(sga == "Yes" | lga == "Yes", NA, "No")) |>
  impute_by_pred("slga", slga_imp_set_predictions) |>
  mutate(
    sga =
      ifelse(
        sga == "Yes" & lga == "Yes"
        ,ifelse(slga == "Yes", "Yes", "No")
        ,sga
      )
    ,lga =
      ifelse(
        sga == "Yes" & lga == "Yes"
        ,ifelse(slga == "Yes", "No", "Yes")
        ,lga
      )
  ) |>
  select(-slga) |>
  
  mutate(ivf0 = ivf) |>
  impute_by_pred("ivf", ivf_imp_set_predictions) |>
  mutate(ivf = ifelse(is.na(ivf0) & ivf == "Yes", NA, ivf)) |>
  select(-ivf0) |>
  
  mutate(subfertility0 = subfertility) |>
  impute_by_pred("subfertility", subfertility_imp_set_predictions) |>
  mutate(
    subfertility =
      ifelse(is.na(subfertility0) & subfertility == "Yes", NA, subfertility)
  ) |>
  select(-subfertility0) |>
  
  impute_by_pred("chorioamnionitis", chorioamnionitis_imp_set_predictions)

saveRDS(ano_ori_ga_conds, "data/ano_ori_ga_conds.rds")
```

```{r Load pre-applied imputation, include=FALSE}
ano_ori_ga_conds <- readRDS("data/ano_ori_ga_conds.rds")
```

## Phenotype preprocessing for non-conditions

```{r GA and non-conditions of phenotype chars, eval=FALSE, include=FALSE}
ano_ori_ga_non_conds0 <-
  ano_ori_chars |>
  bind_rows() |>
  separate(
    fetal_sex_chr
    ,c("fetal_sex2", "chr")
    ,sep = ", "
  ) |>
  select(-fetal_sex2) |>
  separate(
    sentrix_position
    ,c("sentrix_position_R", "sentrix_position_C")
    ,sep = "C"
  ) |>
  mutate_at("sentrix_position_C", \(x) paste0("C", x)) |>
  select(-sentrix_id) |>
  separate(
    sentrix
    ,c("sentrix_id", "sentrix_position")
    ,sep = "_"
  ) |>
  select(-sentrix_position)  |>
  mutate(
    smoker =
      ifelse(
        !is.na(smoker_ever)
        ,smoker_ever
        ,ifelse(
          !is.na(smoker_smoking)
          ,smoker_smoking
          ,ifelse(
            smoke_prepreg %in% c("No", "Yes")
            ,smoke_prepreg
            ,NA
          )
        )
      )
  ) |>
  select(-smoker_ever, -smoke_prepreg, -smoker_smoking) |>
  mutate(
    parity_nulliparous =
      ifelse(
        is.na(parity_nulliparous)
        ,ifelse(
          parity_primiparous == "Yes"
          ,"No"
          ,NA
        )
        ,parity_nulliparous
      )
    ,parity_primiparous =
      ifelse(
        is.na(parity_primiparous)
        ,ifelse(
          parity_nulliparous == "Yes"
          ,"No"
          ,NA
        )
        ,parity_primiparous
      )
  ) |>
  rename(placenta = tissue) |>
  mutate(
    placenta =
      ifelse(str_detect(placenta, "unspecified"), NA, placenta)
  ) |>
  mutate(nightshift = ifelse(nightshift == "Unspecified", NA, nightshift))

ano_ori_ga_non_conds0 <-
  ano_ori_ga_non_conds0 |>
  select_at(
    ano_ori_ga_non_conds0 |>
      colnames() |>
      sapply(\(x) !str_detect(x, "^y_|_id$|GA")) |>
      which()
  ) |>
  select_if(\(x) length(unique(x)[!is.na(unique(x))]) > 1) |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ano[readRDS("data/df_train_colnames.rds"), ] |>
      select(Sample_accession, Dataset_ID, GA)
    , by = join_by(Sample_accession)
  ) |>
  column_to_rownames(var = "Sample_accession") |>
  select(Dataset_ID, GA, everything()) |>
  mutate(ceiled_GA = ceiling(GA))

var_ga_non_conds_abn <-
  c("mod"
    ,"nicu"
    ,"mod_cs"
    ,"attempted_vaginal_delivery"
  )

ano_ori_ga_non_conds0 <-
  ano_ori_ga_non_conds0 |>
  select_at(
    ano_ori_ga_non_conds0 |>
      colnames() |>
      setdiff(var_ga_non_conds_abn)
  )

var_ga_non_conds_multinom <-
  c("sentrix_position_R"
    ,"race_ethnicity"
    ,"education"
    ,"maternal_blood_type"
  )

ano_ori_ga_non_conds0 <-
  ano_ori_ga_non_conds0 |>
  select_if(!colnames(ano_ori_ga_non_conds0) %in% var_ga_non_conds_multinom)
```

```{r Do not need to apply imputation for non-conds, eval=FALSE, include=FALSE}
ano_ori_ga_non_conds <- ano_ori_ga_non_conds0

saveRDS(ano_ori_ga_non_conds, "data/ano_ori_ga_non_conds.rds")
```

```{r Load pre-applied imputation for non-conds, include=FALSE}
ano_ori_ga_non_conds <- readRDS("data/ano_ori_ga_non_conds.rds")
```

# Predictive modeling

## Phenotype correlation matrix among conditions and gestational age

```{r Identify and create missingness variables, include=FALSE}
ano_ori_ga_conds_ms <-
  ano_ori_ga_conds |>
  select(-Dataset_ID, -ceiled_GA) |>
  sapply(\(x) any(is.na(x))) |>
  which() |>
  names() |>
  lapply(
    \(x)
    ano_ori_ga_conds |>
      select_at(x) |>
      `colnames<-`("value") |>
      mutate(value = ifelse(is.na(value), "yes", "no")) |>
      `colnames<-`(paste0("ms_", x))
  ) |>
  reduce(cbind) |>
  cbind(
    ano_ori_ga_conds |>
      select(-Dataset_ID, -ceiled_GA)
  )
```

```{r Check categorical variables with a category of 1 value, include=FALSE}
var_cat_val1=
  ano_ori_ga_conds_ms |>
  select_if(!sapply(ano_ori_ga_conds_ms, is.numeric)) |>
  gather(variable, value) |>
  group_by_all() |>
  summarize(n = n(), .groups = "drop")|>
  filter(n <= 1) |>
  pull(variable) |>
  unique()
```

```{r Pair-wise distribution of categorical variables, include=FALSE}
pairwise_cat_sum <-
  ano_ori_ga_conds_ms |>
  select_if(
    !sapply(ano_ori_ga_conds_ms, is.numeric)
    & !colnames(ano_ori_ga_conds_ms) %in% var_cat_val1
  ) |>
  colnames() |>
  combn(2) |>
  as.data.frame() |>
  lapply(
    \(x)
    ano_ori_ga_conds_ms |>
      select_at(x) |>
      group_by_all() |>
      summarize(n = n(), .groups = "drop") |>
      select_at(c(x, "n")) |>
      `colnames<-`(c("V1_value", "V2_value", "n")) |>
      mutate(V1 = x[1], V2 = x[2])
  ) |>
  lapply(
    \(x)
    expand.grid(
        V1_value = unique(x$V1_value)
        ,V2_value = unique(x$V2_value)
      ) |>
      mutate_all(as.character) |>
      left_join(x, by = join_by(V1_value, V2_value)) |>
      mutate_at("n", \(x) ifelse(is.na(x), 0, x)) |>
      fill(V1, V2)
  ) |>
  reduce(rbind) |>
  select(V1, V1_value, V2, V2_value, everything())
```

```{r Pair-wise perfect separation, include=FALSE}
pairwise_cat_sum_ps <-
  pairwise_cat_sum |>
  group_by(V1, V2) |>
  summarize(ps = any(n == 0), .groups = "drop")
```

```{r table-s7, echo=FALSE}
pairwise_cat_sum_ps |>
  filter(ps) |>
  select(-ps) |>
  kable(
    caption =
      paste0(
        "Table S7. Categorical variables with "
        ,"pair-wise perfect separation."
      )
  ) |>
  kable_classic()
```

```{r Conduct correlation tests for each pair, include=FALSE}
correlation_matrix <-
  ano_ori_ga_conds_ms |>
  colnames() |>
  setdiff(var_cat_val1) |>
  combn(2) |>
  as.data.frame() |>
  t() |>
  as.data.frame() |>
  `rownames<-`(NULL) |>
  filter(
    !paste0(V1, V2)
    %in% c(
      pairwise_cat_sum_ps |>
        filter(ps) |>
        unite(V1V2, V1, V2, sep = "") |>
        pull(V1V2)
    )
  ) |>
  filter(str_remove_all(V1, "^ms_") != str_remove_all(V2, "^ms_")) |>
  pmap(
    \(V1, V2)
    pearson_fisher_aov_t(
        ano_ori_ga_conds_ms[[V1]]
        ,ano_ori_ga_conds_ms[[V2]]
      ) |>
      tidy() |>
      filter(!is.na(p.value)) |>
      select(p.value) |>
      mutate(
        V1 = V1
        ,V2 = V2
      )
  ) |>
  reduce(rbind) |>
  mutate(p.value = p.adjust(p.value, "BH")) |>
  rename(cor.p.value = p.value) |>
  right_join(
    ano_ori_ga_conds_ms |>
      colnames() |>
      combn(2) |>
      as.data.frame() |>
      t() |>
      as.data.frame() |>
      `rownames<-`(NULL)
    ,by = join_by(V1, V2)
  ) |>
  mutate(
    V1 = factor(V1, colnames(ano_ori_ga_conds_ms))
    ,V2 = factor(V2, levels(V1))
  ) |>
  arrange(V1, V2)
```

```{r figure-s1, echo=FALSE, fig.height=5, fig.width=7}
correlation_matrix |>
  mutate(
    V1 = factor(V1, rev(levels(V1)))
    ,sig =
      ifelse(
        is.na(cor.p.value)
        ,ifelse(
          str_remove_all(V1, "^ms_") == str_remove_all(V2, "^ms_")
          ,"Not tested"
          ,ifelse(
            V1 %in% var_cat_val1
            | V2 %in% var_cat_val1
            ,"Not tested"
            ,"Not tested"
          )
        )
        ,ifelse(cor.p.value <= 0.05, "Significant", "Not significant")
      ) |>
      factor()
    ,sig = factor(sig, levels(sig)[c(1, 4, 3, 2)])
    ,cor.p.value = ifelse(cor.p.value<0.001, "<0.001", round(cor.p.value,3))
  ) |>
  ggplot(aes(V1, V2, fill = sig)) +
  geom_tile(color = "white", na.rm = TRUE) +
  # geom_text(aes(label = cor.p.value), size = 3, na.rm = TRUE) +
  coord_flip() +
  xlab("") +
  ylab("") +
  scale_fill_discrete("Significance*") +
  theme(
    panel.grid =
      element_blank()
    ,axis.text.x =
      element_text(angle = 90, vjust = 0.5, hjust=1)
  )
```

```{r Identify conditions related to GA, eval=FALSE, include=FALSE}
var_ga <-
  correlation_matrix |>
  filter(V1 == "GA" | V2 == "GA") |>
  filter(cor.p.value <= 0.05) |>
  gather(variable, value, -cor.p.value) |>
  filter(value != "GA") |>
  filter(!str_detect(value, "^ms_")) |>
  pull(value)

saveRDS(var_ga, "data/var_ga.rds")
```

```{r Load pre-identified conditions related to GA, include=FALSE}
var_ga <- readRDS("data/var_ga.rds")
```

```{r Identify condition missingness related to GA, include=FALSE}
var_ga_ms <-
  correlation_matrix |>
  filter(V1 == "GA" | V2 == "GA") |>
  filter(cor.p.value <= 0.05) |>
  gather(variable, value, -cor.p.value) |>
  filter(value != "GA") |>
  filter(str_detect(value, "^ms_")) |>
  pull(value) |>
  str_remove_all("^ms_")
```

```{r GA distribution of training set per cond, include=FALSE}
ano_ori_ga_conds_plot <-
  ano_ori_ga_conds |>
  colnames() |>
  `names<-`(colnames(ano_ori_ga_conds)) |>
  lapply(
    \(x)
    ano_ori_ga_conds |>
      rename_at(x, \(x) "strata") |>
      ggplot(aes(ceiled_GA, fill = strata)) +
      geom_histogram(binwidth = 1, alpha = 0.5) +
      geom_vline(xintercept = c(12, 26, 37, 42), lty = 2) +
      facet_grid(~ Dataset_ID, scales = "free_x") +
      coord_flip() +
      scale_x_reverse(breaks = seq(0, 42, 2)) +
      scale_fill_discrete(x) +
      theme(
        axis.title.x = element_blank()
        ,axis.text.x = element_blank()
        ,axis.ticks.x = element_blank()
        ,strip.text.x = element_text(angle = 270)
      )
  )
```

```{r figure-s2, echo=FALSE}
ano_ori_ga_conds_plot[var_ga]
```

```{r figure-s3, echo=FALSE}
ano_ori_ga_conds_plot[var_ga_ms]
```

## Model development

### GA prediction

Normal-GA model was trained using samples without 12 of 13 available conditions. They were significantly correlated to GA: (1) fetal growth restriction (FGR); (2) PE; (3) PE onset (early/late/not applicable); (4) hemolysis, elevated liver enzyme, and low platelet (HELLP) syndrome; (5) anencephaly; (6) spina bifida; (7) diandric triploid; (8) miscarriage; (9) preterm delivery; (10) gestational diabetes mellitus (GDM); (11) large-for-gestational-age (LGA) infant; (12) subfertility; and (13) chorioamnionitis. We excluded preterm delivery because it was related to the outcome, i.e., GA, simply by definition.

```{r GA prediction - Identify datasets, eval=FALSE, include=FALSE}
ga_data0 <-
  ano_ori_ga_conds |>
  filter(
    fgr == "No"
    & pe == "No"
    & pe_onset == "Not applicable"
    & anencephaly == "No"
    & spina_bifida == "No"
    & gdm == "No"
    & diandric_triploid == "No"
    & miscarriage == "No"
    & lga == "No"
    & subfertility == "No"
    & hellp == "No"
    & chorioamnionitis == "No"
  ) |>
  select(Dataset_ID, outcome = GA)

ga_data <-
  ga_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_data <-
  ga_data |>
  `names<-`(ga_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r GA prediction - Identify DMPs, eval=FALSE, include=FALSE}
ga_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[, bind_rows(ga_data)$Sample_Name, drop = FALSE] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_data0[bind_rows(ga_data)$Sample_accession, ,drop = FALSE] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_dmp, "data/ga_dmp.rds")
```

```{r GA prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
ga_dmp <- readRDS("data/ga_dmp.rds")

data.frame(probe_id = rownames(ga_dmp[[1]])) |>
  write_csv("inst/extdata/ga_dmp.csv")
```

```{r GA prediction - Load pre-written DMPs, include=FALSE}
ga_dmp <- list()

ga_dmp$NumericVariable <-
  read_csv("inst/extdata/ga_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r GA prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_dmp$NumericVariable)
    ,bind_rows(ga_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA prediction - Prediction set, eval=FALSE, include=FALSE}
ga_est_set <-
  beta_norm_bmiq[
    rownames(ga_dmp$NumericVariable)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r GA prediction - Write prediction train set, eval=FALSE, include=FALSE}
ga_est_train_set |>
  write_csv("inst/extdata/ga_est_train_set.csv")
```

```{r GA prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_est_set |>
  write_csv("inst/extdata/ga_est_set.csv")

ga_est_set |>
  rownames() |>
  saveRDS("data/ga_est_set_rownames.rds")
```

```{r GA prediction - Read prediction results, include=FALSE}
ga_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA prediction - Obtain coefficients, include=FALSE}
ga_est_coefficients <-
  read_csv(
    "inst/extdata/ga_est_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(ga_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### FGR prediction

```{r FGR prediction - Identify datasets, eval=FALSE, include=FALSE}
fgr_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(fgr)) |>
  select(Dataset_ID, outcome = fgr)

fgr_pred_data <-
  fgr_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

fgr_pred_data <-
  fgr_pred_data |>
  `names<-`(fgr_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        fgr_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            fgr_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r FGR prediction - Prediction train set, eval=FALSE, include=FALSE}
fgr_pred_train_set <-
  beta_norm_bmiq[
    rownames(fgr_dmp$Yes_to_No)
    ,bind_rows(fgr_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(fgr_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    fgr_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r FGR prediction - Write prediction train set, eval=FALSE, include=FALSE}
fgr_pred_train_set |>
  write_csv("inst/extdata/fgr_pred_train_set.csv")
```

```{r FGR prediction - Prediction set, eval=FALSE, include=FALSE}
fgr_pred_set <-
  beta_norm_bmiq[
    rownames(fgr_dmp$Yes_to_No)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r FGR prediction - Write prediction set, eval=FALSE, include=FALSE}
fgr_pred_set |>
  write_csv("inst/extdata/fgr_pred_set.csv")

fgr_pred_set |>
  rownames() |>
  saveRDS("data/fgr_pred_set_rownames.rds")
```

```{r FGR prediction - Read prediction results, include=FALSE}
fgr_pred_set_predictions <-
  read_csv("inst/extdata/fgr_pred_prob.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/fgr_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r FGR prediction - Obtain coefficients, include=FALSE}
fgr_pred_coefficients <-
  read_csv(
    "inst/extdata/fgr_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(fgr_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### PE prediction

```{r PE prediction - Identify datasets, eval=FALSE, include=FALSE}
pe_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(pe)) |>
  select(Dataset_ID, outcome = pe)

pe_pred_data <-
  pe_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

pe_pred_data <-
  pe_pred_data |>
  `names<-`(pe_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        pe_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            pe_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r PE prediction - Prediction train set, eval=FALSE, include=FALSE}
pe_pred_train_set <-
  beta_norm_bmiq[
    rownames(pe_dmp$Yes_to_No)
    ,bind_rows(pe_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(pe_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    pe_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r PE prediction - Write prediction train set, eval=FALSE, include=FALSE}
pe_pred_train_set |>
  write_csv("inst/extdata/pe_pred_train_set.csv")
```

```{r PE prediction - Prediction set, eval=FALSE, include=FALSE}
pe_pred_set <-
  beta_norm_bmiq[
    rownames(pe_dmp$Yes_to_No)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r PE prediction - Write prediction set, eval=FALSE, include=FALSE}
pe_pred_set |>
  write_csv("inst/extdata/pe_pred_set.csv")

pe_pred_set |>
  rownames() |>
  saveRDS("data/pe_pred_set_rownames.rds")
```

```{r PE prediction - Read prediction results, include=FALSE}
pe_pred_set_predictions <-
  read_csv("inst/extdata/pe_pred_prob.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/pe_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r PE prediction - Obtain coefficients, include=FALSE}
pe_pred_coefficients <-
  read_csv(
    "inst/extdata/pe_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(pe_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### PE onset prediction

```{r PE onset prediction - Identify datasets, eval=FALSE, include=FALSE}
pe_onset_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!(is.na(pe_onset) | pe_onset == "Not applicable")) |>
  select(Dataset_ID, outcome = pe_onset)

pe_onset_pred_data <-
  pe_onset_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

pe_onset_pred_data <-
  pe_onset_pred_data |>
  `names<-`(pe_onset_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        pe_onset_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            pe_onset_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r PE onset prediction - Prediction train set, eval=FALSE, include=FALSE}
pe_onset_pred_train_set <-
  beta_norm_bmiq[
    rownames(pe_onset_dmp$Early_to_Late)
    ,bind_rows(pe_onset_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(pe_onset_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    pe_onset_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Early")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r PE ons prediction - Write pred train set, eval=FALSE, include=FALSE}
pe_onset_pred_train_set |>
  write_csv("inst/extdata/pe_onset_pred_train_set.csv")
```

```{r PE onset prediction - Prediction set, eval=FALSE, include=FALSE}
pe_onset_pred_set <-
  beta_norm_bmiq[
    rownames(pe_onset_dmp$Early_to_Late)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r PE onset prediction - Write prediction set, eval=FALSE, include=FALSE}
pe_onset_pred_set |>
  write_csv("inst/extdata/pe_onset_pred_set.csv")

pe_onset_pred_set |>
  rownames() |>
  saveRDS("data/pe_onset_pred_set_rownames.rds")
```

```{r PE onset prediction - Read prediction results, include=FALSE}
pe_onset_pred_set_predictions <-
  read_csv(
    "inst/extdata/pe_onset_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/pe_onset_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r PE onset prediction - Obtain coefficients, include=FALSE}
pe_onset_pred_coefficients <-
  read_csv(
    "inst/extdata/pe_onset_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(pe_onset_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### HELLP prediction

```{r HELLP prediction - Identify datasets, eval=FALSE, include=FALSE}
hellp_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(hellp)) |>
  select(Dataset_ID, outcome = hellp)

hellp_pred_data <-
  hellp_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

hellp_pred_data <-
  hellp_pred_data |>
  `names<-`(hellp_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        hellp_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            hellp_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r HELLP prediction - Prediction train set, eval=FALSE, include=FALSE}
hellp_pred_train_set <-
  beta_norm_bmiq[
    rownames(hellp_dmp$No_to_Yes)
    ,bind_rows(hellp_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(hellp_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    hellp_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r HELLP prediction - Write prediction train set, eval=FALSE, include=FALSE}
hellp_pred_train_set |>
  write_csv("inst/extdata/hellp_pred_train_set.csv")
```

```{r HELLP prediction - Prediction set, eval=FALSE, include=FALSE}
hellp_pred_set <-
  beta_norm_bmiq[
    rownames(hellp_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r HELLP prediction - Write prediction set, eval=FALSE, include=FALSE}
hellp_pred_set |>
  write_csv("inst/extdata/hellp_pred_set.csv")

hellp_pred_set |>
  rownames() |>
  saveRDS("data/hellp_pred_set_rownames.rds")
```

```{r HELLP prediction - Read prediction results, include=FALSE}
hellp_pred_set_predictions <-
  read_csv(
    "inst/extdata/hellp_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/hellp_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r HELLP prediction - Obtain coefficients, include=FALSE}
hellp_pred_coefficients <-
  read_csv(
    "inst/extdata/hellp_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(hellp_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Anencephaly prediction

```{r Anencephaly imputation - Identify datasets, eval=FALSE, include=FALSE}
anencephaly_data0 <-
  ano_ori_ga_conds0 |>
  filter(Dataset_ID == "GSE69502") |>
  filter(!is.na(anencephaly)) |>
  select(Dataset_ID, outcome = anencephaly)

anencephaly_data <-
  anencephaly_data0 |>
  pull(Dataset_ID) |>
  unique()

anencephaly_data <-
  anencephaly_data |>
  `names<-`(anencephaly_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        anencephaly_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            anencephaly_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Anencephaly imputation - Identify DMPs, eval=FALSE, include=FALSE}
anencephaly_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(anencephaly_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      anencephaly_data0[
        bind_rows(anencephaly_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
    ,adjPVal = 0.001
    ,adjust.method = "none"
  )

saveRDS(anencephaly_dmp, "data/anencephaly_dmp.rds")
```

```{r Anence prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
anencephaly_dmp <- readRDS("data/anencephaly_dmp.rds")

data.frame(probe_id = rownames(anencephaly_dmp[[1]])) |>
  write_csv("inst/extdata/anencephaly_dmp.csv")
```

```{r Anence prediction - Load pre-written DMPs, include=FALSE}
anencephaly_dmp <- list()

anencephaly_dmp$No_to_Yes <-
  read_csv("inst/extdata/anencephaly_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Anencephaly imputation - Imputation train set, eval=FALSE, include=FALSE}
anencephaly_imp_train_set <-
  beta_norm_bmiq[
    rownames(anencephaly_dmp$No_to_Yes)
    ,bind_rows(anencephaly_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(anencephaly_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    anencephaly_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Anence imputation - Write imp train set, eval=FALSE, include=FALSE}
anencephaly_imp_train_set |>
  write_csv("inst/extdata/anencephaly_imp_train_set.csv")
```

```{r Anencephaly prediction - Identify datasets, eval=FALSE, include=FALSE}
anencephaly_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(anencephaly)) |>
  select(Dataset_ID, outcome = anencephaly)

anencephaly_pred_data <-
  anencephaly_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

anencephaly_pred_data <-
  anencephaly_pred_data |>
  `names<-`(anencephaly_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        anencephaly_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            anencephaly_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Anencephaly prediction - Prediction train set, eval=FALSE, include=FALSE}
anencephaly_pred_train_set <-
  beta_norm_bmiq[
    rownames(anencephaly_dmp$No_to_Yes)
    ,bind_rows(anencephaly_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(anencephaly_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    anencephaly_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Anence prediction - Write pred train set, eval=FALSE, include=FALSE}
anencephaly_pred_train_set |>
  write_csv("inst/extdata/anencephaly_pred_train_set.csv")
```

```{r Anencephaly prediction - Prediction set, eval=FALSE, include=FALSE}
anencephaly_pred_set <-
  beta_norm_bmiq[
    rownames(anencephaly_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Anencephaly prediction - Write prediction set, eval=FALSE, include=FALSE}
anencephaly_pred_set |>
  write_csv("inst/extdata/anencephaly_pred_set.csv")

anencephaly_pred_set |>
  rownames() |>
  saveRDS("data/anencephaly_pred_set_rownames.rds")
```

```{r Anencephaly prediction - Read prediction results, include=FALSE}
anencephaly_pred_set_predictions <-
  read_csv(
    "inst/extdata/anencephaly_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/anencephaly_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Anencephaly prediction - Obtain coefficients, include=FALSE}
anencephaly_pred_coefficients <-
  read_csv(
    "inst/extdata/anencephaly_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(anencephaly_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Spina bifida prediction

```{r Spina bifida imputation - Identify datasets, eval=FALSE, include=FALSE}
spina_bifida_data0 <-
  ano_ori_ga_conds0 |>
  filter(Dataset_ID == "GSE69502") |>
  filter(!is.na(spina_bifida)) |>
  select(Dataset_ID, outcome = spina_bifida)

spina_bifida_data <-
  spina_bifida_data0 |>
  pull(Dataset_ID) |>
  unique()

spina_bifida_data <-
  spina_bifida_data |>
  `names<-`(spina_bifida_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        spina_bifida_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            spina_bifida_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Spina bifida imputation - Identify DMPs, eval=FALSE, include=FALSE}
spina_bifida_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(spina_bifida_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      spina_bifida_data0[
        bind_rows(spina_bifida_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
    ,adjPVal = 0.001
    ,adjust.method = "none"
  )

saveRDS(spina_bifida_dmp, "data/spina_bifida_dmp.rds")
```

```{r Spibif prediction - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
spina_bifida_dmp <- readRDS("data/spina_bifida_dmp.rds")

data.frame(probe_id = rownames(spina_bifida_dmp[[1]])) |>
  write_csv("inst/extdata/spina_bifida_dmp.csv")
```

```{r Spibif prediction - Load pre-written DMPs, include=FALSE}
spina_bifida_dmp <- list()

spina_bifida_dmp$Yes_to_No <-
  read_csv("inst/extdata/spina_bifida_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r Spina bifida imputation - Imputation train set, eval=FALSE, include=FALSE}
spina_bifida_imp_train_set <-
  beta_norm_bmiq[
    rownames(spina_bifida_dmp$Yes_to_No)
    ,bind_rows(spina_bifida_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(spina_bifida_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    spina_bifida_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Spibif imputation - Write imp train set, eval=FALSE, include=FALSE}
spina_bifida_imp_train_set |>
  write_csv("inst/extdata/spina_bifida_imp_train_set.csv")
```

```{r Spina bifida prediction - Identify datasets, eval=FALSE, include=FALSE}
spina_bifida_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(spina_bifida)) |>
  select(Dataset_ID, outcome = spina_bifida)

spina_bifida_pred_data <-
  spina_bifida_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

spina_bifida_pred_data <-
  spina_bifida_pred_data |>
  `names<-`(spina_bifida_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        spina_bifida_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            spina_bifida_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Spina bifida prediction - Prediction train set, eval=FALSE, include=FALSE}
spina_bifida_pred_train_set <-
  beta_norm_bmiq[
    rownames(spina_bifida_dmp$Yes_to_No)
    ,bind_rows(spina_bifida_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(spina_bifida_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    anencephaly_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Spibif prediction - Write pred train set, eval=FALSE, include=FALSE}
spina_bifida_pred_train_set |>
  write_csv("inst/extdata/spina_bifida_pred_train_set.csv")
```

```{r Spina bifida prediction - Prediction set, eval=FALSE, include=FALSE}
spina_bifida_pred_set <-
  beta_norm_bmiq[
    rownames(spina_bifida_dmp$Yes_to_No)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Spina bifida prediction - Write pred set, eval=FALSE, include=FALSE}
spina_bifida_pred_set |>
  write_csv("inst/extdata/spina_bifida_pred_set.csv")

spina_bifida_pred_set |>
  rownames() |>
  saveRDS("data/spina_bifida_pred_set_rownames.rds")
```

```{r Spina bifida prediction - Read prediction results, include=FALSE}
spina_bifida_pred_set_predictions <-
  read_csv(
    "inst/extdata/spina_bifida_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/spina_bifida_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Spina bifida prediction - Obtain coefficients, include=FALSE}
spina_bifida_pred_coefficients <-
  read_csv(
    "inst/extdata/spina_bifida_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(spina_bifida_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Diandric triploid prediction

```{r Diatri prediction - Identify datasets, eval=FALSE, include=FALSE}
diandric_triploid_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(diandric_triploid)) |>
  select(Dataset_ID, outcome = diandric_triploid)

diandric_triploid_pred_data <-
  diandric_triploid_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

diandric_triploid_pred_data <-
  diandric_triploid_pred_data |>
  `names<-`(diandric_triploid_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        diandric_triploid_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            diandric_triploid_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Diatri prediction - Prediction train set, eval=FALSE, include=FALSE}
diandric_triploid_pred_train_set <-
  beta_norm_bmiq[
    rownames(diandric_triploid_dmp$No_to_Yes)
    ,bind_rows(diandric_triploid_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(diandric_triploid_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    diandric_triploid_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Diatri prediction - Write pred train set, eval=FALSE, include=FALSE}
diandric_triploid_pred_train_set |>
  write_csv("inst/extdata/diandric_triploid_pred_train_set.csv")
```

```{r Diandric triploid prediction - Prediction set, eval=FALSE, include=FALSE}
diandric_triploid_pred_set <-
  beta_norm_bmiq[
    rownames(diandric_triploid_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Diatri prediction - Write prediction set, eval=FALSE, include=FALSE}
diandric_triploid_pred_set |>
  write_csv("inst/extdata/diandric_triploid_pred_set.csv")

diandric_triploid_pred_set |>
  rownames() |>
  saveRDS("data/diandric_triploid_pred_set_rownames.rds")
```

```{r Diandric triploid prediction - Read prediction results, include=FALSE}
diandric_triploid_pred_set_predictions <-
  read_csv(
    "inst/extdata/diandric_triploid_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/diandric_triploid_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Diandric triploid prediction - Obtain coefficients, include=FALSE}
diandric_triploid_pred_coefficients <-
  read_csv(
    "inst/extdata/diandric_triploid_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(diandric_triploid_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Miscarriage prediction

```{r Miscarriage prediction - Identify datasets, eval=FALSE, include=FALSE}
miscarriage_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(miscarriage)) |>
  select(Dataset_ID, outcome = miscarriage)

miscarriage_pred_data <-
  miscarriage_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

miscarriage_pred_data <-
  miscarriage_pred_data |>
  `names<-`(miscarriage_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        miscarriage_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            miscarriage_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Miscarriage prediction - Prediction train set, eval=FALSE, include=FALSE}
miscarriage_pred_train_set <-
  beta_norm_bmiq[
    rownames(miscarriage_dmp$No_to_Yes)
    ,bind_rows(miscarriage_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(miscarriage_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    miscarriage_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Miscar prediction - Write pred train set, eval=FALSE, include=FALSE}
miscarriage_pred_train_set |>
  write_csv("inst/extdata/miscarriage_pred_train_set.csv")
```

```{r Miscarriage prediction - Prediction set, eval=FALSE, include=FALSE}
miscarriage_pred_set <-
  beta_norm_bmiq[
    rownames(miscarriage_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Miscarriage prediction - Write prediction set, eval=FALSE, include=FALSE}
miscarriage_pred_set |>
  write_csv("inst/extdata/miscarriage_pred_set.csv")

miscarriage_pred_set |>
  rownames() |>
  saveRDS("data/miscarriage_pred_set_rownames.rds")
```

```{r Miscarriage prediction - Read prediction results, include=FALSE}
miscarriage_pred_set_predictions <-
  read_csv(
    "inst/extdata/miscarriage_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/miscarriage_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Miscarriage prediction - Obtain coefficients, include=FALSE}
miscarriage_pred_coefficients <-
  read_csv(
    "inst/extdata/miscarriage_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(miscarriage_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Preterm prediction

```{r Preterm prediction - Identify datasets, eval=FALSE, include=FALSE}
preterm_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(preterm)) |>
  select(Dataset_ID, outcome = preterm)

preterm_pred_data <-
  preterm_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

preterm_pred_data <-
  preterm_pred_data |>
  `names<-`(preterm_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        preterm_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            preterm_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Preterm prediction - Prediction train set, eval=FALSE, include=FALSE}
preterm_pred_train_set <-
  beta_norm_bmiq[
    rownames(preterm_dmp$Yes_to_No)
    ,bind_rows(preterm_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(preterm_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    preterm_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Pret prediction - Write prediction train set, eval=FALSE, include=FALSE}
preterm_pred_train_set |>
  write_csv("inst/extdata/preterm_pred_train_set.csv")
```

```{r Preterm prediction - Prediction set, eval=FALSE, include=FALSE}
preterm_pred_set <-
  beta_norm_bmiq[
    rownames(preterm_dmp$Yes_to_No)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Preterm prediction - Write prediction set, eval=FALSE, include=FALSE}
preterm_pred_set |>
  write_csv("inst/extdata/preterm_pred_set.csv")

preterm_pred_set |>
  rownames() |>
  saveRDS("data/preterm_pred_set_rownames.rds")
```

```{r Preterm prediction - Read prediction results, include=FALSE}
preterm_pred_set_predictions <-
  read_csv(
    "inst/extdata/preterm_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/preterm_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Preterm prediction - Obtain coefficients, include=FALSE}
preterm_pred_coefficients <-
  read_csv(
    "inst/extdata/preterm_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(preterm_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### GDM prediction

```{r GDM prediction - Identify datasets, eval=FALSE, include=FALSE}
gdm_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(gdm)) |>
  select(Dataset_ID, outcome = gdm)

gdm_pred_data <-
  gdm_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

gdm_pred_data <-
  gdm_pred_data |>
  `names<-`(gdm_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        gdm_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            gdm_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r GDM prediction - Prediction train set, eval=FALSE, include=FALSE}
gdm_pred_train_set <-
  beta_norm_bmiq[
    rownames(gdm_dmp$No_to_Yes)
    ,bind_rows(gdm_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(gdm_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    gdm_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GDM prediction - Write prediction train set, eval=FALSE, include=FALSE}
gdm_pred_train_set |>
  write_csv("inst/extdata/gdm_pred_train_set.csv")
```

```{r GDM prediction - Prediction set, eval=FALSE, include=FALSE}
gdm_pred_set <-
  beta_norm_bmiq[
    rownames(gdm_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r GDM prediction - Write prediction set, eval=FALSE, include=FALSE}
gdm_pred_set |>
  write_csv("inst/extdata/gdm_pred_set.csv")

gdm_pred_set |>
  rownames() |>
  saveRDS("data/gdm_pred_set_rownames.rds")
```

```{r GDM prediction - Read prediction results, include=FALSE}
gdm_pred_set_predictions <-
  read_csv("inst/extdata/gdm_pred_prob.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/gdm_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GDM prediction - Obtain coefficients, include=FALSE}
gdm_pred_coefficients <-
  read_csv(
    "inst/extdata/gdm_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(gdm_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### LGA prediction

```{r LGA prediction - Identify datasets, eval=FALSE, include=FALSE}
lga_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(lga)) |>
  select(Dataset_ID, outcome = lga)

lga_pred_data <-
  lga_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

lga_pred_data <-
  lga_pred_data |>
  `names<-`(lga_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        lga_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            lga_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r LGA prediction - Prediction train set, eval=FALSE, include=FALSE}
lga_pred_train_set <-
  beta_norm_bmiq[
    rownames(lga_dmp$No_to_Yes)
    ,bind_rows(lga_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(lga_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    lga_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r LGA prediction - Write prediction train set, eval=FALSE, include=FALSE}
lga_pred_train_set |>
  write_csv("inst/extdata/lga_pred_train_set.csv")
```

```{r LGA prediction - Prediction set, eval=FALSE, include=FALSE}
lga_pred_set <-
  beta_norm_bmiq[
    rownames(lga_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r LGA prediction - Write prediction set, eval=FALSE, include=FALSE}
lga_pred_set |>
  write_csv("inst/extdata/lga_pred_set.csv")

lga_pred_set |>
  rownames() |>
  saveRDS("data/lga_pred_set_rownames.rds")
```

```{r LGA prediction - Read prediction results, include=FALSE}
lga_pred_set_predictions <-
  read_csv("inst/extdata/lga_pred_prob.csv", show_col_types = FALSE) |>
  mutate(rowname = readRDS("data/lga_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r LGA prediction - Obtain coefficients, include=FALSE}
lga_pred_coefficients <-
  read_csv(
    "inst/extdata/lga_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(lga_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Subfertility prediction

```{r Subfertility prediction - Identify datasets, eval=FALSE, include=FALSE}
subfertility_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(subfertility)) |>
  select(Dataset_ID, outcome = subfertility)

subfertility_pred_data <-
  subfertility_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

subfertility_pred_data <-
  subfertility_pred_data |>
  `names<-`(subfertility_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        subfertility_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            subfertility_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Subfertility prediction - Prediction train set, eval=FALSE, include=FALSE}
subfertility_pred_train_set <-
  beta_norm_bmiq[
    rownames(subfertility_dmp$No_to_Yes)
    ,bind_rows(subfertility_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(subfertility_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    subfertility_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Subfe prediction - Write prediction train set, eval=FALSE, include=FALSE}
subfertility_pred_train_set |>
  write_csv("inst/extdata/subfertility_pred_train_set.csv")
```

```{r Subfertility prediction - Prediction set, eval=FALSE, include=FALSE}
subfertility_pred_set <-
  beta_norm_bmiq[
    rownames(subfertility_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Subfertility prediction - Write pred set, eval=FALSE, include=FALSE}
subfertility_pred_set |>
  write_csv("inst/extdata/subfertility_pred_set.csv")

subfertility_pred_set |>
  rownames() |>
  saveRDS("data/subfertility_pred_set_rownames.rds")
```

```{r Subfertility prediction - Read prediction results, include=FALSE}
subfertility_pred_set_predictions <-
  read_csv(
    "inst/extdata/subfertility_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/subfertility_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Subfertility prediction - Obtain coefficients, include=FALSE}
subfertility_pred_coefficients <-
  read_csv(
    "inst/extdata/subfertility_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(subfertility_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### Chorioamnionitis prediction

```{r Chor prediction - Identify datasets, eval=FALSE, include=FALSE}
chorioamnionitis_pred_data0 <-
  ano_ori_ga_conds |>
  filter(!is.na(chorioamnionitis)) |>
  select(Dataset_ID, outcome = chorioamnionitis)

chorioamnionitis_pred_data <-
  chorioamnionitis_pred_data0 |>
  pull(Dataset_ID) |>
  unique()

chorioamnionitis_pred_data <-
  chorioamnionitis_pred_data |>
  `names<-`(chorioamnionitis_pred_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        chorioamnionitis_pred_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            chorioamnionitis_pred_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r Chor prediction - Prediction train set, eval=FALSE, include=FALSE}
chorioamnionitis_pred_train_set <-
  beta_norm_bmiq[
    rownames(chorioamnionitis_dmp$No_to_Yes)
    ,bind_rows(chorioamnionitis_pred_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(chorioamnionitis_pred_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    chorioamnionitis_pred_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = as.integer(outcome == "Yes")) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Chor prediction - Write prediction train set, eval=FALSE, include=FALSE}
chorioamnionitis_pred_train_set |>
  write_csv("inst/extdata/chorioamnionitis_pred_train_set.csv")
```

```{r Chor prediction - Prediction set, eval=FALSE, include=FALSE}
chorioamnionitis_pred_set <-
  beta_norm_bmiq[
    rownames(chorioamnionitis_dmp$No_to_Yes)
    ,rownames(ano)
    ,drop = FALSE
  ] |>
  `colnames<-`(ano$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r Chor prediction - Write prediction set, eval=FALSE, include=FALSE}
chorioamnionitis_pred_set |>
  write_csv("inst/extdata/chorioamnionitis_pred_set.csv")

chorioamnionitis_pred_set |>
  rownames() |>
  saveRDS("data/chorioamnionitis_pred_set_rownames.rds")
```

```{r Chorioamnionitis prediction - Read prediction results, include=FALSE}
chorioamnionitis_pred_set_predictions <-
  read_csv(
    "inst/extdata/chorioamnionitis_pred_prob.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/chorioamnionitis_pred_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r Chorioamnionitis prediction - Obtain coefficients, include=FALSE}
chorioamnionitis_pred_coefficients <-
  read_csv(
    "inst/extdata/chorioamnionitis_pred_coefficients.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = rownames(chorioamnionitis_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### GA res-true-conds prediction

```{r GA res-true-conds prediction - Pred train set, eval=FALSE, include=FALSE}
ga_rtc_est_train_set <-
  ano_ori_ga_conds |>
  select_at(var_ga) |>
  mutate_all(\(x) ifelse(x %in% c("Yes", "Early"), 1, 0)) |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      left_join(
        ano_ori_ga_conds |>
          select(outcome = GA) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      )
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

#### GA res-true-conds elastic net

```{r GA RTC prediction - Write prediction train set, eval=FALSE, include=FALSE}
ga_rtc_est_train_set |>
  write_csv("inst/extdata/ga_rtc_est_train_set.csv")
```

```{r GA RTC prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_rtc_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_rtc_est_set.csv")

ga_rtc_est_train_set |>
  rownames() |>
  saveRDS("data/ga_rtc_est_set_rownames.rds")
```

```{r GA res-true-conds prediction - Read prediction results, include=FALSE}
ga_rtc_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_rtc_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_rtc_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

#### GA res-true-conds random forest

```{r GA RTC-RF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_rtc_est_train_set |>
  write_csv("inst/extdata/ga_rtc_rf_est_train_set.csv")
```

```{r GA RTC-RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_rtc_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_rtc_rf_est_set.csv")

ga_rtc_est_train_set |>
  rownames() |>
  saveRDS("data/ga_rtc_rf_est_set_rownames.rds")
```

```{r GA RTC-RF prediction - Read prediction results, include=FALSE}
ga_rtc_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_rtc_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_rtc_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA residual prediction

```{r GA residual prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_est_train_set <-
  paste0("inst/extdata/", var_ga, "_imp_prob.csv") |>
  `names<-`(var_ga) |>
  imap(
    ~ .x |>
      read_csv(show_col_types = FALSE) |>
      mutate(
        rowname =
          readRDS(paste0("data/", .y, "_pred_set_rownames.rds"))
      ) |>
      column_to_rownames(var = "rowname") |>
      rename_at("predicted_probability", \(x) .y)
  ) |>
  reduce(cbind) |>
  select_at(var_ga) |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      left_join(
        ano_ori_ga_conds |>
          select(outcome = GA) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      )
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

#### GA residual elastic net

```{r GA res prediction - Write prediction train set, eval=FALSE, include=FALSE}
ga_res_est_train_set |>
  write_csv("inst/extdata/ga_res_est_train_set.csv")
```

```{r GA res prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_est_set.csv")

ga_res_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_est_set_rownames.rds")
```

```{r GA residual prediction - Read prediction results, include=FALSE}
ga_res_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

#### GA residual random forest

```{r GA rRF prediction - Write prediction train set, eval=FALSE, include=FALSE}
ga_res_est_train_set |>
  write_csv("inst/extdata/ga_res_rf_est_train_set.csv")
```

```{r GA res RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_rf_est_set.csv")

ga_res_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_rf_est_set_rownames.rds")
```

```{r GA res RF prediction - Read prediction results, include=FALSE}
ga_res_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA res-comp-risk prediction

```{r GA res-comp-risk prediction - Pred train set, eval=FALSE, include=FALSE}
ga_res_cr_est_train_set <-
  paste0("inst/extdata/", var_ga, "_imp_prob.csv") |>
  `names<-`(var_ga) |>
  imap(
    ~ .x |>
      read_csv(show_col_types = FALSE) |>
      mutate(
        rowname =
          readRDS(paste0("data/", .y, "_pred_set_rownames.rds"))
      ) |>
      column_to_rownames(var = "rowname") |>
      rename_at("predicted_probability", \(x) .y)
  ) |>
  reduce(cbind) |>
  select_at(var_ga) |>
  rownames_to_column(var = "Sample_accession")

ga_res_cr_est_train_set <-
  ga_res_cr_est_train_set |>
  gather(cond, prob, -Sample_accession) |>
  mutate_at(c("Sample_accession", "cond"), \(x) factor(x, unique(x))) |>
  group_by(Sample_accession) |>
  mutate(
    hi_prob = ifelse(prob == max(prob), "Yes", "No")
    ,hi_cond =
      ifelse(hi_prob == "Yes", cond, NA) |>
      setdiff(NA) |>
      sapply(\(x) paste0("^", x, "$")) |>
      paste0(collapse = "|")
  ) |>
  mutate(
    comp_risk =
      ifelse(
        str_detect(hi_cond, "\\^miscarriage\\$")
        ,"^miscarriage$|^subfertility$"
        ,ifelse(
          str_detect(
            hi_cond
            ,c("\\^anencephaly\\$"
               ,"\\^spina_bifida\\$"
               ,"\\^diandric_triploid\\$"
              ) |>
              paste0(collapse = "|")
          )
          ,c("^anencephaly$"
             ,"^spina_bifida$"
             ,"^diandric_triploid$"
             ,"^miscarriage$"
             ,"^subfertility$"
            ) |>
            paste0(collapse = "|")
          ,"^[:graph:]*$"
        )
      )
    ,prob_comp_risk =
      ifelse(
        str_detect(cond, comp_risk)
        ,prob
        ,0
      )
  ) |>
  select(Sample_accession, cond, prob_comp_risk) |>
  spread(cond, prob_comp_risk, fill = 0) |>
  arrange(Sample_accession) |>
  
  mutate(
    pe_onset = pe * pe_onset
    ,hellp = pe * hellp
    ,fgr = fgr + pe * fgr
    ,lga = lga + gdm * lga
    ,preterm = preterm + pe_onset + hellp + chorioamnionitis
  )

ga_res_cr_est_train_set <-
  ga_res_cr_est_train_set |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      left_join(
        ano_ori_ga_conds |>
          select(outcome = GA) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      )
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

#### GA res-comp-risk elastic net

```{r GA res-comp-risk pred - Write pred train set, eval=FALSE, include=FALSE}
ga_res_cr_est_train_set |>
  write_csv("inst/extdata/ga_res_cr_est_train_set.csv")
```

```{r GA res-comp-risk pred - Write prediction set, eval=FALSE, include=FALSE}
ga_res_cr_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_cr_est_set.csv")

ga_res_cr_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_cr_est_set_rownames.rds")
```

```{r GA res-comp-risk prediction - Read prediction results, include=FALSE}
ga_res_cr_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_cr_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_cr_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

#### GA res-comp-risk random forest

```{r GA rcrRF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_cr_est_train_set |>
  write_csv("inst/extdata/ga_res_cr_rf_est_train_set.csv")
```

```{r GA rcrRF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_cr_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_cr_rf_est_set.csv")

ga_res_cr_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_cr_rf_est_set_rownames.rds")
```

```{r GA rcrRF prediction - Read prediction results, include=FALSE}
ga_res_cr_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_cr_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_cr_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA res-full prediction

```{r GA res-full prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_resfull_est_train_set <-
  data.frame(
    path =
      list.files("inst/extdata", pattern = "_prob.csv", full.names = FALSE)
  ) |>
  mutate(
    cond =
      path |>
      str_remove_all("_imp|_pred|_prob|\\.csv") |>
      factor(var_ga)
    ,type = ifelse(str_detect(path, "imp_prob"), "imp_prob", "pred_prob")
  ) |>
  spread(type, path) |>
  mutate(
    path = ifelse(is.na(pred_prob), imp_prob, pred_prob)
    ,path = paste0("inst/extdata/", path)
  ) |>
  pull(path) |>
  `names<-`(var_ga) |>
  imap(
    ~ .x |>
      read_csv(show_col_types = FALSE) |>
      mutate(
        rowname =
          readRDS(paste0("data/", .y, "_pred_set_rownames.rds"))
      ) |>
      column_to_rownames(var = "rowname") |>
      rename_at("predicted_probability", \(x) .y)
  ) |>
  reduce(cbind) |>
  select_at(var_ga) |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      left_join(
        ano_ori_ga_conds |>
          select(outcome = GA) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      )
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

#### GA res-full elastic net

```{r GA res-full prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_resfull_est_train_set |>
  write_csv("inst/extdata/ga_resfull_est_train_set.csv")
```

```{r GA res-full prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_resfull_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_resfull_est_set.csv")

ga_resfull_est_train_set |>
  rownames() |>
  saveRDS("data/ga_resfull_est_set_rownames.rds")
```

```{r GA res-full prediction - Read prediction results, include=FALSE}
ga_resfull_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_resfull_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_resfull_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

#### GA res-full random forest

```{r GA rfullRF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_resfull_est_train_set |>
  write_csv("inst/extdata/ga_resfull_rf_est_train_set.csv")
```

```{r GA res-full RF prediction - Write pred set, eval=FALSE, include=FALSE}
ga_resfull_est_train_set |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_resfull_rf_est_set.csv")

ga_resfull_est_train_set |>
  rownames() |>
  saveRDS("data/ga_resfull_rf_est_set_rownames.rds")
```

```{r GA res-full RF prediction - Read prediction results, include=FALSE}
ga_resfull_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_resfull_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_resfull_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA res-seq-risk prediction

```{r GA res-seq-risk prediction - Pred train set, eval=FALSE, include=FALSE}
ga_res_sr_est_train_set <-
  paste0("inst/extdata/", var_ga, "_imp_prob.csv") |>
  `names<-`(var_ga) |>
  imap(
    ~ .x |>
      read_csv(show_col_types = FALSE) |>
      mutate(
        rowname =
          readRDS(paste0("data/", .y, "_pred_set_rownames.rds"))
      ) |>
      column_to_rownames(var = "rowname") |>
      rename_at("predicted_probability", \(x) .y)
  ) |>
  reduce(cbind) |>
  select_at(var_ga) |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      left_join(
        ano_ori_ga_conds |>
          select(outcome = GA) |>
          rownames_to_column(var = "Sample_accession")
        ,by = join_by(Sample_accession)
      )
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA res-seq-risk prediction - Sequence of conds, eval=FALSE, include=FALSE}
ga_res_sr_seq_conds <-
  list(
    conds1 = c("miscarriage", "subfertility")
    ,conds2 = c("anencephaly", "spina_bifida", "diandric_triploid")
    ,conds3 = c("fgr", "pe", "chorioamnionitis", "gdm")
    ,conds4 = c("pe_onset", "hellp")
    ,conds5 = c("preterm", "lga")
  )
```

#### GA res-seq-risk elastic net

```{r GA rsr1 prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds1)) |>
  write_csv("inst/extdata/ga_res_sr1_est_train_set.csv")
```

```{r GA rsr1 prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_sr_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds1)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr1_est_set.csv")

ga_res_sr_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr1_est_set_rownames.rds")
```

```{r GA rsr1 prediction - Read prediction results, include=FALSE}
ga_res_sr1_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr1_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr1_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr1 prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr1_est_train_set <-
  ga_res_sr1_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr2 prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr1_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds2)) |>
  write_csv("inst/extdata/ga_res_sr2_est_train_set.csv")
```

```{r GA rsr2 prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_sr1_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds2)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr2_est_set.csv")

ga_res_sr1_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr2_est_set_rownames.rds")
```

```{r GA rsr2 prediction - Read prediction results, include=FALSE}
ga_res_sr2_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr2_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr2_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr2 prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr2_est_train_set <-
  ga_res_sr2_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr1_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr3 prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr2_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds3)) |>
  write_csv("inst/extdata/ga_res_sr3_est_train_set.csv")
```

```{r GA rsr3 prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_sr2_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds3)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr3_est_set.csv")

ga_res_sr2_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr3_est_set_rownames.rds")
```

```{r GA rsr3 prediction - Read prediction results, include=FALSE}
ga_res_sr3_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr3_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr3_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr3 prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr3_est_train_set <-
  ga_res_sr3_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr2_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr4 prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr3_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds4)) |>
  write_csv("inst/extdata/ga_res_sr4_est_train_set.csv")
```

```{r GA rsr4 prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_sr3_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds4)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr4_est_set.csv")

ga_res_sr3_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr4_est_set_rownames.rds")
```

```{r GA rsr4 prediction - Read prediction results, include=FALSE}
ga_res_sr4_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr4_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr4_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr4 prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr4_est_train_set <-
  ga_res_sr4_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr3_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr4_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds5)) |>
  write_csv("inst/extdata/ga_res_sr_est_train_set.csv")
```

```{r GA rsr prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_sr4_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds5)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr_est_set.csv")

ga_res_sr4_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr_est_set_rownames.rds")
```

```{r GA rsr prediction - Read prediction results, include=FALSE}
ga_res_sr_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

#### GA res-seq-risk random forest

```{r GA rsr1RF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds1)) |>
  write_csv("inst/extdata/ga_res_sr1_rf_est_train_set.csv")
```

```{r GA rsr1RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_sr_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds1)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr1_rf_est_set.csv")

ga_res_sr_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr1_rf_est_set_rownames.rds")
```

```{r GA rsr1RF prediction - Read prediction results, include=FALSE}
ga_res_sr1_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr1_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr1_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr1RF prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr1_rf_est_train_set <-
  ga_res_sr1_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr2RF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr1_rf_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds2)) |>
  write_csv("inst/extdata/ga_res_sr2_rf_est_train_set.csv")
```

```{r GA rsr2RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_sr1_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds2)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr2_rf_est_set.csv")

ga_res_sr1_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr2_rf_est_set_rownames.rds")
```

```{r GA rsr2RF prediction - Read prediction results, include=FALSE}
ga_res_sr2_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr2_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr2_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr2RF prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr2_rf_est_train_set <-
  ga_res_sr2_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr1_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr3RF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr2_rf_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds3)) |>
  write_csv("inst/extdata/ga_res_sr3_rf_est_train_set.csv")
```

```{r GA rsr3RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_sr2_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds3)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr3_rf_est_set.csv")

ga_res_sr2_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr3_rf_est_set_rownames.rds")
```

```{r GA rsr3RF prediction - Read prediction results, include=FALSE}
ga_res_sr3_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr3_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr3_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr3RF prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr3_rf_est_train_set <-
  ga_res_sr3_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr2_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsr4RF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr3_rf_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds4)) |>
  write_csv("inst/extdata/ga_res_sr4_rf_est_train_set.csv")
```

```{r GA rsr4RF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_sr3_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds4)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr4_rf_est_set.csv")

ga_res_sr3_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr4_rf_est_set_rownames.rds")
```

```{r GA rsr4RF prediction - Read prediction results, include=FALSE}
ga_res_sr4_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr4_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr4_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA rsr4RF prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_sr4_rf_est_train_set <-
  ga_res_sr4_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_sr3_est_train_set |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(outcome = outcome - prediction) |>
  select(-prediction) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA rsrRF prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_sr4_rf_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds5)) |>
  write_csv("inst/extdata/ga_res_sr_rf_est_train_set.csv")
```

```{r GA rsrRF prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_sr4_est_train_set |>
  select_at(c("outcome", ga_res_sr_seq_conds$conds5)) |>
  select(-outcome) |>
  write_csv("inst/extdata/ga_res_sr_rf_est_set.csv")

ga_res_sr4_est_train_set |>
  rownames() |>
  saveRDS("data/ga_res_sr_rf_est_set_rownames.rds")
```

```{r GA rsrRF prediction - Read prediction results, include=FALSE}
ga_res_sr_rf_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_sr_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_sr_rf_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA res-cpg prediction

Res-CPG-GA model consisted of two models for <37 and 37 weeks gestation estimated by normal-GA model. The model numbers and periods were determined according to clinical knowledge and pursuing normal distribution of residual GA. Pregnancy termination <37 weeks gestation is presumably related to a medical indication, while during term pregnancy, both medical and non-medical indications might be encountered. Fitting residuals in term pregnancy using beta values might lead to overfitting. Nonetheless, we used two approaches, i.e., predicting residual GA during: (1) <37 weeks gestation only (Res-CPG-PR-GA); and (2) both <37 and 37 weeks gestation (Res-CPG-GA). The latter model training was restricted to samples with absolute residual GA >0.05. It was chosen based on visual judgement of quantile-to-quantile (QQ) plot to pursue linearity, hence, easily fitted by elastic net regression. We assumed that such approach would avoid overfitting on predicted GA which might be already well-estimated by normal-GA model. We tested this assumption by training Res-CPG model without the residual-GA restriction (Res-CPG-rev-GA).

```{r GA res-cpg prediction - Identify datasets, eval=FALSE, include=FALSE}
ga_res_cpg_data0 <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ano_ori_ga_conds |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(residual = GA - prediction) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Normal QQ plots of normal-GA, eval=FALSE, include=FALSE}
list(
    preterm =
      ga_data0 |>
      filter(ceiling(outcome) <= 36)
   ,term =
      ga_data0 |>
      filter(ceiling(outcome) >= 37)
  ) |>
  imap(
    ~ .x |>
      ggplot(aes(sample=outcome)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r Normal QQ plots of abnormal-GA, eval=FALSE, include=FALSE}
list(
    preterm_t1t2a =
      ano |>
      filter(!Sample_accession %in% rownames(ga_data0)) |>
      filter(ceiling(GA) <= 20) |>
      mutate(GA = (GA - 5) / (20 - 5)) |>
      mutate(GA = log(GA / (1 - GA)))
    ,preterm_t2b =
      ano |>
      filter(!Sample_accession %in% rownames(ga_data0)) |>
      filter(ceiling(GA) >= 21 & ceiling(GA) <= 26)
    ,preterm_t3 =
      ano |>
      filter(!Sample_accession %in% rownames(ga_data0)) |>
      filter(ceiling(GA) >= 27 & ceiling(GA) <= 36) |>
      mutate(GA = (GA - 26) / (36 - 26)) |>
      mutate(GA = log(GA / (1 - GA)))
    ,norm_term =
      ano |>
      filter(!Sample_accession %in% rownames(ga_data0)) |>
      filter(ceiling(GA) >= 37)
  ) |>
  imap(
    ~ .x |>
      ggplot(aes(sample=GA)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r Normal QQ plots of residual-GA, eval=FALSE, include=FALSE}
list(
    preterm_t1t2a =
      ga_res_cpg_data0 |>
      filter(ceiling(prediction) <= 20)
    ,preterm_t2b =
      ga_res_cpg_data0 |>
      filter(ceiling(prediction) >= 21 & ceiling(prediction) <= 26)
    ,preterm_t3 =
      ga_res_cpg_data0 |>
      filter(ceiling(prediction) >= 27 & ceiling(prediction) <= 36)
    ,term_lores =
      ga_res_cpg_data0 |>
      filter(ceiling(prediction) >= 37) |>
      filter(abs(residual) <= 0.05)
    ,term_hires =
      ga_res_cpg_data0 |>
      filter(ceiling(prediction) >= 37) |>
      filter(abs(residual) > 0.05)
  ) |>
  imap(
    ~ .x |>
      ggplot(aes(sample=residual)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r GA res-cpg pred - Identify datasets continued, eval=FALSE, include=FALSE}
ga_res_cpg_pr_data0 <-
  ga_res_cpg_data0 |>
  filter(ceiling(prediction) <= 36) |>
  select(Dataset_ID, outcome = residual)

ga_res_cpg_pr_data <-
  ga_res_cpg_pr_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_cpg_pr_data <-
  ga_res_cpg_pr_data |>
  `names<-`(ga_res_cpg_pr_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_cpg_pr_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_cpg_pr_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )

ga_res_cpg_thr_data0 <-
  ga_res_cpg_data0 |>
  filter(ceiling(prediction) >= 37) |>
  filter(abs(residual) > 0.05) |>
  select(Dataset_ID, outcome = residual)

ga_res_cpg_thr_data <-
  ga_res_cpg_thr_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_cpg_thr_data <-
  ga_res_cpg_thr_data |>
  `names<-`(ga_res_cpg_thr_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_cpg_thr_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_cpg_thr_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )

ga_res_cpg_t_data0 <-
  ga_res_cpg_data0 |>
  filter(ceiling(prediction) >= 37) |>
  select(Dataset_ID, outcome = residual)

ga_res_cpg_t_data <-
  ga_res_cpg_t_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_cpg_t_data <-
  ga_res_cpg_t_data |>
  `names<-`(ga_res_cpg_t_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_cpg_t_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_cpg_t_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r GA res-cpg prediction - Identify DMPs, eval=FALSE, include=FALSE}
ga_res_cpg_pr_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_cpg_pr_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_cpg_pr_data0[
        bind_rows(ga_res_cpg_pr_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_cpg_pr_dmp, "data/ga_res_cpg_pr_dmp.rds")

ga_res_cpg_thr_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_cpg_thr_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_cpg_thr_data0[
        bind_rows(ga_res_cpg_thr_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_cpg_thr_dmp, "data/ga_res_cpg_thr_dmp.rds")

ga_res_cpg_t_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_cpg_t_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_cpg_t_data0[
        bind_rows(ga_res_cpg_t_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_cpg_t_dmp, "data/ga_res_cpg_t_dmp.rds")
```

```{r GA res-cpg pred - Load pre-identified DMPs, eval=FALSE, include=FALSE}
ga_res_cpg_pr_dmp <- readRDS("data/ga_res_cpg_pr_dmp.rds")
ga_res_cpg_thr_dmp <- readRDS("data/ga_res_cpg_thr_dmp.rds")
ga_res_cpg_t_dmp <- readRDS("data/ga_res_cpg_t_dmp.rds")
```

```{r GA res-cpg prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_pr_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_pr_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_pr_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_pr_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")

ga_res_cpg_thr_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_thr_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_thr_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_thr_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_thr_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")

ga_res_cpg_t_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_t_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_t_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_t_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_t_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA res-cpg prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_train_set |>
  write_csv("inst/extdata/ga_res_cpg_pr_est_train_set.csv")

ga_res_cpg_thr_est_train_set |>
  write_csv("inst/extdata/ga_res_cpg_thr_est_train_set.csv")

ga_res_cpg_t_est_train_set |>
  write_csv("inst/extdata/ga_res_cpg_t_est_train_set.csv")
```

```{r GA res-cpg prediction - Prediction set, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_pr_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_pr_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_pr_data)$Sample_accession) |>
  t() |>
  as.data.frame()

ga_res_cpg_thr_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_thr_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_t_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_t_data)$Sample_accession) |>
  t() |>
  as.data.frame()

ga_res_cpg_t_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_cpg_t_dmp$NumericVariable)
    ,bind_rows(ga_res_cpg_t_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_cpg_t_data)$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r GA res-cpg prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_set |>
  write_csv("inst/extdata/ga_res_cpg_pr_est_set.csv")

ga_res_cpg_pr_est_set |>
  rownames() |>
  saveRDS("data/ga_res_cpg_pr_est_set_rownames.rds")

ga_res_cpg_thr_est_set |>
  write_csv("inst/extdata/ga_res_cpg_thr_est_set.csv")

ga_res_cpg_thr_est_set |>
  rownames() |>
  saveRDS("data/ga_res_cpg_thr_est_set_rownames.rds")

ga_res_cpg_t_est_set |>
  write_csv("inst/extdata/ga_res_cpg_t_est_set.csv")

ga_res_cpg_t_est_set |>
  rownames() |>
  saveRDS("data/ga_res_cpg_t_est_set_rownames.rds")
```

```{r GA res-cpg prediction - Read prediction results, include=FALSE}
ga_res_cpg_pr_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_cpg_pr_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_cpg_pr_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")

ga_res_cpg_thr_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_cpg_thr_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_cpg_thr_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")

ga_res_cpg_t_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_cpg_t_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_cpg_t_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

### GA res-conds prediction

Res-Conds-GA model was similar to Resfull-GA model but we used predictors of multiplication values for each predicted probability and residual GA estimated by a model for the corresponding condition. Specifically, we trained a model using beta values of DMPs among samples with a condition. The rationale was that the conditions have different trajectories of when pregnancies are terminated and each pregnant woman has a different set of probabilities of the conditions. For comparison purpose, we used the true and predicted probabilities of the conditions, either from the prediction (Resfull-GA) or imputation models (Res-GA) for the conditions, and without the multiplication procedure (all the probabilities equal to 1). This comparison tested the importance of phenotypic prediction performances in the parent model accuracy and robustness.

```{r GA res-conds prediction - Identify datasets, eval=FALSE, include=FALSE}
ga_res_conds_all_data <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ano_ori_ga_conds |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(residual = GA - prediction) |>
  column_to_rownames(var = "Sample_accession")

ga_res_conds_data0 <-
  var_ga |>
  `names<-`(var_ga) |>
  imap(
    ~ ga_res_conds_all_data |>
      rename_at(.x, \(x) "strata") |>
      filter(strata  %in% c("Yes", "Early")) |>
      select(Dataset_ID, outcome = residual)
  )
```

```{r Normal QQ plots of conds-GA, eval=FALSE, include=FALSE}
ga_res_conds_data0 |>
  imap(
    ~ .x |>
      ggplot(aes(sample=outcome)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r GA res-conds pred - Identify datasets cont, eval=FALSE, include=FALSE}
ga_res_conds_data <-
  ga_res_conds_data0 |>
  lapply(pull, Dataset_ID) |>
  lapply(unique)

ga_res_conds_data <-
  ga_res_conds_data |>
  imap(~ `names<-`(.x, ga_res_conds_data[[.y]])) |>
  imap(
    ~ .x |>
      lapply(
        \(x)
        list(
          Sample_accession =
            ga_res_conds_data0[[.y]] |>
            filter(Dataset_ID == x) |>
            rownames()
          ,Sample_Name =
            ano |>
            filter(
              Sample_accession
              %in% c(
                ga_res_conds_data0[[.y]] |>
                  filter(Dataset_ID == x) |>
                  rownames()
              )
            ) |>
            pull(Sample_Name)
        )
      )
  )
```

```{r GA res-conds prediction - Identify DMPs, eval=FALSE, include=FALSE}
ga_res_conds_dmp0 <-
  ga_res_conds_data0 |>
  imap(~ list(data0 = .x, data = ga_res_conds_data[[.y]]))

var_ga_res_conds_small <- c("diandric_triploid", "miscarriage", "subfertility")

for(i in var_ga[!var_ga %in% var_ga_res_conds_small]){
  ga_res_conds_dmp0[[i]]$dmp <-
    champ.DMP(
      beta =
        beta_norm_bmiq[
          
          ,bind_rows(ga_res_conds_dmp0[[i]]$data)$Sample_Name
          ,drop = FALSE
        ] |>
        as.matrix() |>
        `class<-`("matrix_array")
      ,pheno =
        ga_res_conds_dmp0[[i]]$data0[
          bind_rows(ga_res_conds_dmp0[[i]]$data)$Sample_accession
          ,
          ,drop = FALSE
        ] |>
        pull(outcome)
    )
  
  ga_res_conds_dmp0[[i]]$dmp |>
    saveRDS(paste0("data/ga_res_conds_", i, "_dmp.rds"))
}

for(i in var_ga_res_conds_small){
  ga_res_conds_dmp0[[i]]$dmp <-
    champ.DMP(
      beta =
        beta_norm_bmiq[
          
          ,bind_rows(ga_res_conds_dmp0[[i]]$data)$Sample_Name
          ,drop = FALSE
        ] |>
        as.matrix() |>
        `class<-`("matrix_array")
      ,pheno =
        ga_res_conds_dmp0[[i]]$data0[
          bind_rows(ga_res_conds_dmp0[[i]]$data)$Sample_accession
          ,
          ,drop = FALSE
        ] |>
        pull(outcome)
      ,adjPVal = 0.001
      ,adjust.method = "none"
    )
  
  ga_res_conds_dmp0[[i]]$dmp |>
    saveRDS(paste0("data/ga_res_conds_", i, "_dmp.rds"))
}

ga_res_conds_dmp <-
  ga_res_conds_dmp0 |>
  lapply(\(x) x$dmp)
```

```{r GA res-conds pred - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
ga_res_conds_dmp0 <-
  ga_res_conds_data0 |>
  imap(~ list(data0 = .x, data = ga_res_conds_data[[.y]]))

for(i in var_ga){
  ga_res_conds_dmp0[[i]]$dmp <-
    readRDS(paste0("data/ga_res_conds_", i, "_dmp.rds"))
}

ga_res_conds_dmp <-
  ga_res_conds_dmp0 |>
  lapply(\(x) x$dmp)

for(i in var_ga){
  data.frame(probe_id = rownames(ga_res_conds_dmp[[i]][[1]])) |>
    write_csv(paste0("inst/extdata/ga_res_conds_", i, "_dmp.csv"))
}
```

```{r GA res-conds prediction - Load pre-written DMPs, include=FALSE}
ga_res_conds_dmp <- list()

for(i in var_ga){
  ga_res_conds_dmp[[i]]$NumericVariable <-
    read_csv(
      paste0("inst/extdata/ga_res_conds_", i, "_dmp.csv")
      , show_col_types = FALSE
    ) |>
    column_to_rownames(var = "probe_id")
}
```

```{r GA res-conds pred - Pred train set & write it, eval=FALSE, include=FALSE}
for(i in var_ga){
  ga_res_conds_est_train_set <-
    beta_norm_bmiq[
      rownames(ga_res_conds_dmp[[i]]$NumericVariable)
      ,bind_rows(ga_res_conds_data[[i]])$Sample_Name
      ,drop = FALSE
    ] |>
    `colnames<-`(bind_rows(ga_res_conds_data[[i]])$Sample_accession) |>
    t() |>
    as.data.frame() |>
    rownames_to_column(var = "Sample_accession") |>
    left_join(
      ga_res_conds_data0[[i]] |>
        rownames_to_column(var = "Sample_accession") |>
        select(-Dataset_ID)
      ,by = join_by(Sample_accession)
    ) |>
    select(outcome, everything()) |>
    column_to_rownames(var = "Sample_accession")
  
  ga_res_conds_est_train_set |>
    write_csv(paste0("inst/extdata/ga_res_conds_", i, "_est_train_set.csv"))
}
```

```{r GA res-conds pred - Prediction set & write it, eval=FALSE, include=FALSE}
for(i in var_ga){
  ga_res_conds_est_set <-
    beta_norm_bmiq[
      rownames(ga_res_conds_dmp[[i]]$NumericVariable)
      ,rownames(ano)
      ,drop = FALSE
    ] |>
    `colnames<-`(ano$Sample_accession) |>
    t() |>
    as.data.frame()
  
  ga_res_conds_est_set |>
    write_csv(paste0("inst/extdata/ga_res_conds_", i, "_est_set.csv"))
  
  ga_res_conds_est_set |>
    rownames() |>
    saveRDS(paste0("data/ga_res_conds_", i, "_est_set_rownames.rds"))
  
  rm(ga_res_conds_est_set)
}
```

```{r GA res-conds prediction - Read cond prediction results, include=FALSE}
ga_res_conds_est_set_predictions <-
  var_ga |>
  `names<-`(var_ga) |>
  pblapply(
    \(x)
      read_csv(
          paste0("inst/extdata/ga_res_conds_", x, "_est_predictions.csv")
          ,show_col_types = FALSE
        ) |>
        mutate(
          rowname =
            readRDS(paste0("data/ga_res_conds_", x, "_est_set_rownames.rds"))
        ) |>
        column_to_rownames(var = "rowname")
  )
```

```{r GA res-conds prediction - Obtain coefficients, include=FALSE}
ga_res_conds_est_coefficients <-
  var_ga |>
  `names<-`(var_ga) |>
  pblapply(
    \(x)
      read_csv(
          paste0("inst/extdata/ga_res_conds_", x, "_est_coefficients.csv")
          ,show_col_types = FALSE
        ) |>
        mutate(rowname = rownames(ga_res_conds_dmp[[x]][[1]])) |>
        column_to_rownames(var = "rowname") |>
        filter(weight != 0)
  )
```

```{r GA res-conds prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_conds_all_est_train_set <-
  list(
    ga_res_conds_true = ga_rtc_est_train_set
    ,ga_res_conds_imp = ga_res_est_train_set
    ,ga_res_conds_pred = ga_resfull_est_train_set
    ,ga_res_conds_null = mutate_all(ga_resfull_est_train_set, \(x) 1)
  ) |>
  lapply(
    \(x)
      ga_res_conds_est_set_predictions |>
        imap(
          ~ .x |>
            rownames_to_column(var = "Sample_accession") |>
            left_join(
              x |>
                rename_at(.y, \(x) "prob") |>
                select(prob) |>
                rownames_to_column(var = "Sample_accession")
              ,by = join_by(Sample_accession)
            ) |>
            column_to_rownames(var = "Sample_accession") |>
            mutate(prob_prediction = prob * prediction) |>
            select(prob_prediction) |>
            rename_at("prob_prediction", \(x) .y)
        ) |>
        reduce(cbind) |>
        rownames_to_column(var = "Sample_accession") |>
        left_join(
          ga_res_conds_all_data |>
            select(outcome = residual) |>
            rownames_to_column(var = "Sample_accession")
          ,by = join_by(Sample_accession)
        ) |>
        column_to_rownames(var = "Sample_accession") |>
        select(outcome, everything())
  )
```

```{r GA res-conds prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_conds_all_est_train_set |>
  imap(
    ~ .x |>
      write_csv(paste0("inst/extdata/", .y, "_est_train_set.csv"))
  )

```

```{r GA res-conds prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_conds_all_est_train_set |>
  imap(
    ~ .x |>
      select(-outcome) |>
      write_csv(paste0("inst/extdata/", .y, "_est_set.csv"))
  )

ga_res_conds_all_est_train_set |>
  imap(
    ~ .x |>
      rownames() |>
      saveRDS(paste0("data/", .y, "_est_set_rownames.rds"))
  )
```

```{r GA res-conds prediction - Read prediction results, include=FALSE}
ga_res_conds_all_est_set_predictions <- c("true", "imp", "pred", "null")

ga_res_conds_all_est_set_predictions <-
  ga_res_conds_all_est_set_predictions |>
  `names<-`(ga_res_conds_all_est_set_predictions) |>
  lapply(
    \(x)
      read_csv(
          paste0("inst/extdata/ga_res_conds_", x, "_est_predictions.csv")
          ,show_col_types = FALSE
        ) |>
        mutate(
          rowname =
            readRDS(paste0("data/ga_res_conds_", x, "_est_set_rownames.rds"))
        ) |>
        column_to_rownames(var = "rowname")
  )
```

### GA res-comb prediction

Res-Comb- and Res-CPG-Comb-GA models were considered because other conditions might affect pregnancy termination, not limited to the 12 conditions. The stacking order was considered different. In Res-Comb-GA model, we limited the degree of freedom of residual fitting using known phenotype information (Res-Conds-GA), thus, the second model only fitted the unexplained residual GA (Res-CPG-GA), simply to boost the prediction. Meanwhile, Res-CPG-GA model was assumed to generally fit both explained and unexplained residual GA. Since the resulting residual was presumably smaller, the overall prediction was less affected by the imperfect accuracies of Res-Conds-GA. Both Res-Comb- and Res-CPG-Comb-GA models consisted of three models for <37, 37 and 40, and >40 weeks gestation estimated by normal-GA model. The model numbers and periods were also determined according to clinical knowledge and pursuing normal distribution of residual GA. The estimated delivery date falls on 40 weeks gestation. Before this date, a pregnant woman might seek termination in advance due to a medical condition. Meanwhile, a normal pregnant woman might seek for termination since the delivery date. We used three approaches, i.e., predicting residual GA during: (1) <37 weeks gestation only (Res-Comb-PR-GA); (2) both <37 and 37 and 40 weeks gestation, i.e., term before the estimated delivery date (Res-Comb-PRTB-GA); and (3) all the three periods (Res-Comb-GA).

```{r GA res-comb prediction - Identify datasets, eval=FALSE, include=FALSE}
ga_res_comb_data0 <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction_conds = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    ano_ori_ga_conds |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(residual = GA - prediction_conds) |>
  column_to_rownames(var = "Sample_accession")
```

```{r Normal QQ plots of residual-conds-GA, eval=FALSE, include=FALSE}
list(
    preterm_t1t2 =
      ga_res_comb_data0 |>
      filter(ceiling(prediction_conds) <= 34)
    ,preterm_t3 =
      ga_res_comb_data0 |>
      filter(ceiling(prediction_conds) >= 27 & ceiling(prediction_conds) <= 36)
    ,term_predate =
      ga_res_comb_data0 |>
      filter(ceiling(prediction_conds) >= 37 & ceiling(prediction_conds) <= 40)
    ,term_postdate =
      ga_res_comb_data0 |>
      filter(ceiling(prediction_conds) >= 41)
  ) |>
  imap(
    ~ .x |>
      ggplot(aes(sample=residual)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r GA res-comb pred - Identify datasets continued, eval=FALSE, include=FALSE}
ga_res_comb_pr_data0 <-
  ga_res_comb_data0 |>
  filter(ceiling(prediction_conds) <= 36) |>
  select(Dataset_ID, outcome = residual)

ga_res_comb_pr_data <-
  ga_res_comb_pr_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_comb_pr_data <-
  ga_res_comb_pr_data |>
  `names<-`(ga_res_comb_pr_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_comb_pr_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_comb_pr_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )

ga_res_comb_tb_data0 <-
  ga_res_comb_data0 |>
  filter(ceiling(prediction_conds) >= 37  & ceiling(prediction_conds) <= 40) |>
  select(Dataset_ID, outcome = residual)

ga_res_comb_tb_data <-
  ga_res_comb_tb_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_comb_tb_data <-
  ga_res_comb_tb_data |>
  `names<-`(ga_res_comb_tb_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_comb_tb_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_comb_tb_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )

ga_res_comb_ta_data0 <-
  ga_res_comb_data0 |>
  filter(ceiling(prediction) >= 41) |>
  select(Dataset_ID, outcome = residual)

ga_res_comb_ta_data <-
  ga_res_comb_ta_data0 |>
  pull(Dataset_ID) |>
  unique()

ga_res_comb_ta_data <-
  ga_res_comb_ta_data |>
  `names<-`(ga_res_comb_ta_data) |>
  lapply(
    \(x)
    list(
      Sample_accession =
        ga_res_comb_ta_data0 |>
        filter(Dataset_ID == x) |>
        rownames()
      ,Sample_Name =
        ano |>
        filter(
          Sample_accession
          %in% c(
            ga_res_comb_ta_data0 |>
              filter(Dataset_ID == x) |>
              rownames()
          )
        ) |>
        pull(Sample_Name)
    )
  )
```

```{r GA res-comb prediction - Identify DMPs, eval=FALSE, include=FALSE}
ga_res_comb_pr_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_comb_pr_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_comb_pr_data0[
        bind_rows(ga_res_comb_pr_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_comb_pr_dmp, "data/ga_res_comb_pr_dmp.rds")

ga_res_comb_tb_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_comb_tb_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_comb_tb_data0[
        bind_rows(ga_res_comb_tb_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_comb_tb_dmp, "data/ga_res_comb_tb_dmp.rds")

ga_res_comb_ta_dmp <-
  champ.DMP(
    beta =
      beta_norm_bmiq[
        
        ,bind_rows(ga_res_comb_ta_data)$Sample_Name
        ,drop = FALSE
      ] |>
      as.matrix() |>
      `class<-`("matrix_array")
    ,pheno =
      ga_res_comb_ta_data0[
        bind_rows(ga_res_comb_ta_data)$Sample_accession
        ,
        ,drop = FALSE
      ] |>
      pull(outcome)
    ,arraytype = "450K"
  )

saveRDS(ga_res_comb_ta_dmp, "data/ga_res_comb_ta_dmp.rds")
```

```{r GA res-comb pred - Load and write pre-id DMPs, eval=FALSE, include=FALSE}
ga_res_comb_pr_dmp <- readRDS("data/ga_res_comb_pr_dmp.rds")

data.frame(probe_id = rownames(ga_res_comb_pr_dmp[[1]])) |>
  write_csv("inst/extdata/ga_res_comb_pr_dmp.csv")

ga_res_comb_tb_dmp <- readRDS("data/ga_res_comb_tb_dmp.rds")

data.frame(probe_id = rownames(ga_res_comb_tb_dmp[[1]])) |>
  write_csv("inst/extdata/ga_res_comb_tb_dmp.csv")

ga_res_comb_ta_dmp <- readRDS("data/ga_res_comb_ta_dmp.rds")

data.frame(probe_id = rownames(ga_res_comb_ta_dmp[[1]])) |>
  write_csv("inst/extdata/ga_res_comb_ta_dmp.csv")
```

```{r GA res-comb prediction - Load pre-written DMPs, include=FALSE}
ga_res_comb_pr_dmp <- list()

ga_res_comb_pr_dmp$NumericVariable <-
  read_csv("inst/extdata/ga_res_comb_pr_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")

ga_res_comb_tb_dmp <- list()

ga_res_comb_tb_dmp$NumericVariable <-
  read_csv("inst/extdata/ga_res_comb_tb_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")

ga_res_comb_ta_dmp <- list()

ga_res_comb_ta_dmp$NumericVariable <-
  read_csv("inst/extdata/ga_res_comb_ta_dmp.csv", show_col_types = FALSE) |>
  column_to_rownames(var = "probe_id")
```

```{r GA res-comb prediction - Prediction train set, eval=FALSE, include=FALSE}
ga_res_comb_pr_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_pr_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_pr_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_pr_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_comb_pr_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")

ga_res_comb_tb_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_tb_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_tb_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_tb_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_comb_tb_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")

ga_res_comb_ta_est_train_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_ta_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_ta_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_ta_data)$Sample_accession) |>
  t() |>
  as.data.frame() |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_comb_ta_data0 |>
      rownames_to_column(var = "Sample_accession") |>
      select(-Dataset_ID)
    ,by = join_by(Sample_accession)
  ) |>
  select(outcome, everything()) |>
  column_to_rownames(var = "Sample_accession")
```

```{r GA res-comb prediction - Write pred train set, eval=FALSE, include=FALSE}
ga_res_comb_pr_est_train_set |>
  write_csv("inst/extdata/ga_res_comb_pr_est_train_set.csv")

ga_res_comb_tb_est_train_set |>
  write_csv("inst/extdata/ga_res_comb_tb_est_train_set.csv")

ga_res_comb_ta_est_train_set |>
  write_csv("inst/extdata/ga_res_comb_ta_est_train_set.csv")
```

```{r GA res-comb prediction - Prediction set, eval=FALSE, include=FALSE}
ga_res_comb_pr_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_pr_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_pr_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_pr_data)$Sample_accession) |>
  t() |>
  as.data.frame()

ga_res_comb_tb_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_tb_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_tb_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_tb_data)$Sample_accession) |>
  t() |>
  as.data.frame()

ga_res_comb_ta_est_set <-
  beta_norm_bmiq[
    rownames(ga_res_comb_ta_dmp$NumericVariable)
    ,bind_rows(ga_res_comb_ta_data)$Sample_Name
    ,drop = FALSE
  ] |>
  `colnames<-`(bind_rows(ga_res_comb_ta_data)$Sample_accession) |>
  t() |>
  as.data.frame()
```

```{r GA res-comb prediction - Write prediction set, eval=FALSE, include=FALSE}
ga_res_comb_pr_est_set |>
  write_csv("inst/extdata/ga_res_comb_pr_est_set.csv")

ga_res_comb_pr_est_set |>
  rownames() |>
  saveRDS("data/ga_res_comb_pr_est_set_rownames.rds")

ga_res_comb_tb_est_set |>
  write_csv("inst/extdata/ga_res_comb_tb_est_set.csv")

ga_res_comb_tb_est_set |>
  rownames() |>
  saveRDS("data/ga_res_comb_tb_est_set_rownames.rds")

ga_res_comb_ta_est_set |>
  write_csv("inst/extdata/ga_res_comb_ta_est_set.csv")

ga_res_comb_ta_est_set |>
  rownames() |>
  saveRDS("data/ga_res_comb_ta_est_set_rownames.rds")
```

```{r GA res-comb prediction - Read prediction results, include=FALSE}
ga_res_comb_pr_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_comb_pr_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_comb_pr_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")

ga_res_comb_tb_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_comb_tb_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_comb_tb_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")

ga_res_comb_ta_est_set_predictions <-
  read_csv(
    "inst/extdata/ga_res_comb_ta_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = readRDS("data/ga_res_comb_ta_est_set_rownames.rds")) |>
  column_to_rownames(var = "rowname")
```

```{r GA res-comb prediction - Obtain coefficients, include=FALSE}
ga_res_comb_pr_est_coefficients <-
  read_csv(
    "inst/extdata/ga_res_comb_pr_est_coefficients.csv"
    ,show_col_type = FALSE
  ) |>
  mutate(rowname = rownames(ga_res_comb_pr_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)

ga_res_comb_tb_est_coefficients <-
  read_csv(
    "inst/extdata/ga_res_comb_tb_est_coefficients.csv"
    ,show_col_type = FALSE
  ) |>
  mutate(rowname = rownames(ga_res_comb_tb_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)

ga_res_comb_ta_est_coefficients <-
  read_csv(
    "inst/extdata/ga_res_comb_ta_est_coefficients.csv"
    ,show_col_type = FALSE
  ) |>
  mutate(rowname = rownames(ga_res_comb_ta_dmp[[1]])) |>
  column_to_rownames(var = "rowname") |>
  filter(weight != 0)
```

### GA res-cpg-comb prediction

```{r GA res-cpg-comb prediction - Identify datasets, eval=FALSE, include=FALSE}
ga_res_cpg_comb_all_data <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_cpg_t_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_t = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(is.na(residual_prediction_t), 0, residual_prediction_t)
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_t) |>
  left_join(
    ano_ori_ga_conds |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(residual = GA - prediction) |>
  column_to_rownames(var = "Sample_accession")

ga_res_cpg_comb_data0 <-
  var_ga |>
  `names<-`(var_ga) |>
  imap(
    ~ ga_res_cpg_comb_all_data |>
      rename_at(.x, \(x) "strata") |>
      filter(strata  %in% c("Yes", "Early")) |>
      select(Dataset_ID, outcome = residual)
  )
```

```{r Normal QQ plots of residual-cpg-GA, eval=FALSE, include=FALSE}
ga_res_cpg_comb_data0 |>
  imap(
    ~ .x |>
      ggplot(aes(sample=outcome)) +
      stat_qq(na.rm=T) +
      stat_qq_line(color='red',na.rm=T) +
      coord_flip() +
      xlab('Normal Quantiles') +
      ylab(.y)
  )
```

```{r GA res-cpg-comb pred - Identify datasets cont, eval=FALSE, include=FALSE}
ga_res_cpg_comb_data <-
  ga_res_cpg_comb_data0 |>
  lapply(pull, Dataset_ID) |>
  lapply(unique)

ga_res_cpg_comb_data <-
  ga_res_cpg_comb_data |>
  imap(~ `names<-`(.x, ga_res_cpg_comb_data[[.y]])) |>
  imap(
    ~ .x |>
      lapply(
        \(x)
        list(
          Sample_accession =
            ga_res_cpg_comb_data0[[.y]] |>
            filter(Dataset_ID == x) |>
            rownames()
          ,Sample_Name =
            ano |>
            filter(
              Sample_accession
              %in% c(
                ga_res_cpg_comb_data0[[.y]] |>
                  filter(Dataset_ID == x) |>
                  rownames()
              )
            ) |>
            pull(Sample_Name)
        )
      )
  )
```

```{r GA res-cpg-comb prediction - Identify DMPs, eval=FALSE, include=FALSE}
ga_res_cpg_comb_dmp0 <-
  ga_res_cpg_comb_data0 |>
  imap(~ list(data0 = .x, data = ga_res_cpg_comb_data[[.y]]))

var_ga_res_cpg_comb_small <- c()

for(i in var_ga[!var_ga %in% var_ga_res_cpg_comb_small]){
  ga_res_cpg_comb_dmp0[[i]]$dmp <-
    champ.DMP(
      beta =
        beta_norm_bmiq[
          
          ,bind_rows(ga_res_cpg_comb_dmp0[[i]]$data)$Sample_Name
          ,drop = FALSE
        ] |>
        as.matrix() |>
        `class<-`("matrix_array")
      ,pheno =
        ga_res_cpg_comb_dmp0[[i]]$data0[
          bind_rows(ga_res_cpg_comb_dmp0[[i]]$data)$Sample_accession
          ,
          ,drop = FALSE
        ] |>
        pull(outcome)
    )
  
  ga_res_cpg_comb_dmp0[[i]]$dmp |>
    saveRDS(paste0("data/ga_res_cpg_comb_", i, "_dmp.rds"))
}

for(i in var_ga_res_cpg_comb_small){
  ga_res_cpg_comb_dmp0[[i]]$dmp <-
    champ.DMP(
      beta =
        beta_norm_bmiq[
          
          ,bind_rows(ga_res_cpg_comb_dmp0[[i]]$data)$Sample_Name
          ,drop = FALSE
        ] |>
        as.matrix() |>
        `class<-`("matrix_array")
      ,pheno =
        ga_res_cpg_comb_dmp0[[i]]$data0[
          bind_rows(ga_res_cpg_comb_dmp0[[i]]$data)$Sample_accession
          ,
          ,drop = FALSE
        ] |>
        pull(outcome)
      ,adjPVal = 0.001
      ,adjust.method = "none"
    )
  
  ga_res_cpg_comb_dmp0[[i]]$dmp |>
    saveRDS(paste0("data/ga_res_cpg_comb_", i, "_dmp.rds"))
}

ga_res_cpg_comb_dmp <-
  ga_res_cpg_comb_dmp0 |>
  lapply(\(x) x$dmp)
```

```{r GA res-cpg-comb pred - Load pre-id DMPs, eval=FALSE, include=FALSE}
ga_res_cpg_comb_dmp0 <-
  ga_res_cpg_comb_data0 |>
  imap(~ list(data0 = .x, data = ga_res_cpg_comb_data[[.y]]))

for(i in var_ga){
  ga_res_cpg_comb_dmp0[[i]]$dmp <-
    readRDS(paste0("data/ga_res_cpg_comb_", i, "_dmp.rds"))
}

ga_res_cpg_comb_dmp <-
  ga_res_cpg_comb_dmp0 |>
  lapply(\(x) x$dmp)
```

```{r GA res-cpg-comb - Pred train set & write it, eval=FALSE, include=FALSE}
for(i in var_ga){
  ga_res_cpg_comb_est_train_set <-
    beta_norm_bmiq[
      rownames(ga_res_cpg_comb_dmp[[i]]$NumericVariable)
      ,bind_rows(ga_res_cpg_comb_data[[i]])$Sample_Name
      ,drop = FALSE
    ] |>
    `colnames<-`(bind_rows(ga_res_cpg_comb_data[[i]])$Sample_accession) |>
    t() |>
    as.data.frame() |>
    rownames_to_column(var = "Sample_accession") |>
    left_join(
      ga_res_cpg_comb_data0[[i]] |>
        rownames_to_column(var = "Sample_accession") |>
        select(-Dataset_ID)
      ,by = join_by(Sample_accession)
    ) |>
    select(outcome, everything()) |>
    column_to_rownames(var = "Sample_accession")
  
  ga_res_cpg_comb_est_train_set |>
    write_csv(paste0("inst/extdata/ga_res_cpg_comb_", i, "_est_train_set.csv"))
}
```

```{r GA res-cpg-comb pred - Pred set & write it, eval=FALSE, include=FALSE}
for(i in var_ga){
  ga_res_cpg_comb_est_set <-
    beta_norm_bmiq[
      rownames(ga_res_cpg_comb_dmp[[i]]$NumericVariable)
      ,rownames(ano)
      ,drop = FALSE
    ] |>
    `colnames<-`(ano$Sample_accession) |>
    t() |>
    as.data.frame()
  
  ga_res_cpg_comb_est_set |>
    write_csv(paste0("inst/extdata/ga_res_cpg_comb_", i, "_est_set.csv"))
  
  ga_res_cpg_comb_est_set |>
    rownames() |>
    saveRDS(paste0("data/ga_res_cpg_comb_", i, "_est_set_rownames.rds"))
  
  rm(ga_res_cpg_comb_est_set)
}
```

```{r GA res-cpg-comb prediction - Read cond prediction results, include=FALSE}
ga_res_cpg_comb_est_set_predictions <-
  var_ga |>
  `names<-`(var_ga) |>
  pblapply(
    \(x)
      read_csv(
          paste0("inst/extdata/ga_res_cpg_comb_", x, "_est_predictions.csv")
          ,show_col_types = FALSE
        ) |>
        mutate(
          rowname =
            readRDS(paste0("data/ga_res_cpg_comb_", x, "_est_set_rownames.rds"))
        ) |>
        column_to_rownames(var = "rowname")
  )
```

```{r GA res-cpg-comb prediction - Pred train set, eval=FALSE, include=FALSE}
ga_res_cpg_comb_all_est_train_set <-
  list(
    ga_res_cpg_comb_true = ga_rtc_est_train_set
    ,ga_res_cpg_comb_imp = ga_res_est_train_set
    ,ga_res_cpg_comb_pred = ga_resfull_est_train_set
    ,ga_res_cpg_comb_null = mutate_all(ga_resfull_est_train_set, \(x) 1)
  ) |>
  lapply(
    \(x)
      ga_res_cpg_comb_est_set_predictions |>
        imap(
          ~ .x |>
            rownames_to_column(var = "Sample_accession") |>
            left_join(
              x |>
                rename_at(.y, \(x) "prob") |>
                select(prob) |>
                rownames_to_column(var = "Sample_accession")
              ,by = join_by(Sample_accession)
            ) |>
            column_to_rownames(var = "Sample_accession") |>
            mutate(prob_prediction = prob * prediction) |>
            select(prob_prediction) |>
            rename_at("prob_prediction", \(x) .y)
        ) |>
        reduce(cbind) |>
        rownames_to_column(var = "Sample_accession") |>
        left_join(
          ga_res_cpg_comb_all_data |>
            select(outcome = residual) |>
            rownames_to_column(var = "Sample_accession")
          ,by = join_by(Sample_accession)
        ) |>
        column_to_rownames(var = "Sample_accession") |>
        select(outcome, everything())
  )
```

```{r GA res-cpg-comb pred - Write pred train set, eval=FALSE, include=FALSE}
ga_res_cpg_comb_all_est_train_set |>
  imap(
    ~ .x |>
      write_csv(paste0("inst/extdata/", .y, "_est_train_set.csv"))
  )
```

```{r GA res-cpg-comb prediction - Write pred set, eval=FALSE, include=FALSE}
ga_res_cpg_comb_all_est_train_set |>
  imap(
    ~ .x |>
      select(-outcome) |>
      write_csv(paste0("inst/extdata/", .y, "_est_set.csv"))
  )

ga_res_cpg_comb_all_est_train_set |>
  imap(
    ~ .x |>
      rownames() |>
      saveRDS(paste0("data/", .y, "_est_set_rownames.rds"))
  )
```

```{r GA res-cpg-comb prediction - Read prediction results, include=FALSE}
ga_res_cpg_comb_all_est_set_predictions <- c("true", "imp", "pred", "null")

ga_res_cpg_comb_all_est_set_predictions <-
  ga_res_cpg_comb_all_est_set_predictions |>
  `names<-`(ga_res_cpg_comb_all_est_set_predictions) |>
  lapply(
    \(x)
      read_csv(
          paste0("inst/extdata/ga_res_cpg_comb_", x, "_est_predictions.csv")
          ,show_col_types = FALSE
        ) |>
        mutate(
          rowname =
            readRDS(paste0("data/ga_res_cpg_comb_", x, "_est_set_rownames.rds"))
        ) |>
        column_to_rownames(var = "rowname")
  )
```

# Model evaluation

We considered the first-iteration models were underperformed based on the validation set compared to other participants model using the 450k probes (Figure 2). In training set, we also observed the accuracies of those models inconsistently won against the top performer across different phenotypic subgroups (Figure 3). Hence, the phenotype information might be not as useful as expected. Nevertheless, Resfull-GA model performance was slightly more consistent than Res-, Res-CR-, and Res-Seq-GA models although their performances were better using whole training set. In the first iteration, we learned that the accuracies in predicting the conditions matter, since Resfull-GA employed the larger-training-size, prediction models instead of smaller-training-size, imputation models. The prediction model that used the true probabilities were underperformed. However, we argue that it was because the degree of freedom was lacking due to binary/dichotomic values (0/1). This finding also indicated the predicted probabilities were also lacking for the degree of freedom.

```{r Leaderboard data, include=FALSE}
leaderboard <-
  data.frame(
    metric =c("RMSE", "MAE", "r")
    ,current_best = c(1.076, 0.863, 0.966)
  )

submission <-
  read_csv("inst/extdata/submission_info.csv", show_col_types = FALSE)

performance <-
  read_csv("inst/extdata/leaderboard_performance.csv", show_col_types = FALSE)
```

```{r Normal-GA model evaluation, include=FALSE}
ga_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  add_strata_data()

ga_est_eval_results <- stratified_performance(ga_est_eval, "Normal-GA") 
ga_est_eval_plot <- stratified_plot_element(ga_est_eval_results)
```

```{r Res-true-conds-GA model evaluation, include=FALSE}
ga_rtc_est_eval <-
  ga_rtc_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_rtc_est_eval_results <-
  stratified_performance(ga_rtc_est_eval, "Res-GA (true)")

ga_rtc_est_eval_plot <- stratified_plot_element(ga_rtc_est_eval_results)
```

```{r Res-true-conds-RF-GA model evaluation, include=FALSE}
ga_rtc_rf_est_eval <-
  ga_rtc_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_rtc_rf_est_eval_results <-
  stratified_performance(ga_rtc_rf_est_eval, "Res-GA (true)*")

ga_rtc_rf_est_eval_plot <- stratified_plot_element(ga_rtc_rf_est_eval_results)
```

```{r Residual-GA model evaluation, include=FALSE}
ga_res_est_eval <-
  ga_res_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_est_eval_results <-
  stratified_performance(ga_res_est_eval, "Res-GA")

ga_res_est_eval_plot <- stratified_plot_element(ga_res_est_eval_results)
```

```{r ResRF-GA model evaluation, include=FALSE}
ga_res_rf_est_eval <-
  ga_res_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_rf_est_eval_results <-
  stratified_performance(ga_res_rf_est_eval, "Res-GA*")

ga_res_rf_est_eval_plot <- stratified_plot_element(ga_res_rf_est_eval_results)
```

```{r Res-comp-risk-GA model evaluation, include=FALSE}
ga_res_cr_est_eval <-
  ga_res_cr_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cr_est_eval_results <-
  stratified_performance(ga_res_cr_est_eval, "Res-CR-GA")

ga_res_cr_est_eval_plot <- stratified_plot_element(ga_res_cr_est_eval_results)
```

```{r RcrRF-GA model evaluation, include=FALSE}
ga_res_cr_rf_est_eval <-
  ga_res_cr_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cr_rf_est_eval_results <-
  stratified_performance(ga_res_cr_rf_est_eval, "Res-CR-GA*")

ga_res_cr_rf_est_eval_plot <-
  stratified_plot_element(ga_res_cr_rf_est_eval_results)
```

```{r Resfull-GA model evaluation, include=FALSE}
ga_resfull_est_eval <-
  ga_resfull_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_resfull_est_eval_results <-
  stratified_performance(ga_resfull_est_eval, "Resfull-GA")

ga_resfull_est_eval_plot <- stratified_plot_element(ga_resfull_est_eval_results)
```

```{r ResfullRF-GA model evaluation, include=FALSE}
ga_resfull_rf_est_eval <-
  ga_resfull_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction = prediction) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_resfull_rf_est_eval_results <-
  stratified_performance(ga_resfull_rf_est_eval, "Resfull-GA*")

ga_resfull_rf_est_eval_plot <-
  stratified_plot_element(ga_resfull_rf_est_eval_results)
```

```{r Res-seq-risk-GA model evaluation, include=FALSE}
ga_res_sr_est_eval <-
  ga_res_sr1_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction1 = prediction) |>
  left_join(
    ga_res_sr2_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction2 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr3_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction3 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr4_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction4 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + residual_prediction1
      + residual_prediction2
      + residual_prediction3
      + residual_prediction4
      + residual_prediction
  ) |>
  select(
    -residual_prediction1
    ,-residual_prediction2
    ,-residual_prediction3
    ,-residual_prediction4
    ,-residual_prediction
  ) |>
  add_strata_data()

ga_res_sr_est_eval_results <-
  stratified_performance(ga_res_sr_est_eval, "Res-Seq-GA")

ga_res_sr_est_eval_plot <- stratified_plot_element(ga_res_sr_est_eval_results)
```

```{r RsrRF-GA model evaluation, include=FALSE}
ga_res_sr_rf_est_eval <-
  ga_res_sr1_rf_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  rename(residual_prediction1 = prediction) |>
  left_join(
    ga_res_sr2_rf_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction2 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr3_rf_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction3 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr4_rf_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction4 = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_sr_rf_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_est_set_predictions |>
      rownames_to_column(var = "Sample_accession")
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + residual_prediction1
      + residual_prediction2
      + residual_prediction3
      + residual_prediction4
      + residual_prediction
  ) |>
  select(
    -residual_prediction1
    ,-residual_prediction2
    ,-residual_prediction3
    ,-residual_prediction4
    ,-residual_prediction
  ) |>
  add_strata_data()

ga_res_sr_rf_est_eval_results <-
  stratified_performance(ga_res_sr_rf_est_eval, "Res-Seq-GA*")

ga_res_sr_rf_est_eval_plot <-
  stratified_plot_element(ga_res_sr_rf_est_eval_results)
```

Res-CPG-GA model family mostly outperformed the first-iteration models in terms of the rank in the validation set, which jumped up to top-3 positions based on RMSE, and the subgroup-wise winning rates (Figure 2). Res-CPG- and Res-CPG-rev-GA models for <37 and 37 weeks gestation consistently outperformed Res-CPG-PR-GA model across the subgroups (Figure 3), although it coincidentally outperformed Res-CPG- and Res-CPG-rev-GA models in the validation set (Figure 2). This finding opposed our assumption on the overfitting potential (see Modeling strategies). We were inspired to improve the degree of freedom of phenotypic predicted probabilities, resulting in Res-Conds-GA model.

```{r ResCPG-pr-GA model evaluation, include=FALSE}
ga_res_cpg_pr_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,0
      )
  ) |>
  select(-residual_prediction_pr) |>
  add_strata_data()

ga_res_cpg_pr_est_eval_results <-
  stratified_performance(ga_res_cpg_pr_est_eval, "Res-CPG-PR-GA")

ga_res_cpg_pr_est_eval_plot <-
  stratified_plot_element(ga_res_cpg_pr_est_eval_results)
```

```{r ResCPG-GA model evaluation, include=FALSE}
ga_res_cpg_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_cpg_thr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_thr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(is.na(residual_prediction_thr), 0, residual_prediction_thr)
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_thr) |>
  add_strata_data()

ga_res_cpg_est_eval_results <-
  stratified_performance(ga_res_cpg_est_eval, "Res-CPG-GA")

ga_res_cpg_est_eval_plot <- stratified_plot_element(ga_res_cpg_est_eval_results)
```

```{r ResCPGrev-GA model evaluation, include=FALSE}
ga_res_cpgrev_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_cpg_t_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_t = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(is.na(residual_prediction_t), 0, residual_prediction_t)
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_t) |>
  add_strata_data()

ga_res_cpgrev_est_eval_results <-
  stratified_performance(ga_res_cpgrev_est_eval, "Res-CPG-rev-GA")

ga_res_cpgrev_est_eval_plot <-
  stratified_plot_element(ga_res_cpgrev_est_eval_results)
```

Overall, Res-Conds-GA model family did not outperform Res-CPG and Res-CPG-rev models (Figure 2). The winning rates across the subgroups between both model families were also similar, but Res-Conds-GA models finally won in a few subgroups in which the previous models consistently lose, e.g., diandric triploid and miscarriage (Figure 3). Oppositely, Res-Conds-GA models lose in 3 of the 16 datasets of origin, compared to only 1 dataset (GSE74738) for Res-CPG and Res-CPG-rev models. Res-Conds-GA models using the true and predicted probabilities were more accurate and robust than those using imputation models or without the multiplication (Table 1). In the validation set, the Res-Conds-GA using the predicted probabilities had higher RMSE than those of Res-CPG-rev, Res-CPG, and Res-CPG-PR, consecutively, but not the top performer.

```{r ResCondsTrue-GA model evaluation, include=FALSE}
ga_res_conds_true_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$true |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_conds_true_est_eval_results <-
  stratified_performance(ga_res_conds_true_est_eval, "Res-Conds-GA")

ga_res_conds_true_est_eval_plot <-
  stratified_plot_element(ga_res_conds_true_est_eval_results)
```

```{r ResCondsImp-GA model evaluation, include=FALSE}
ga_res_conds_imp_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$imp |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_conds_imp_est_eval_results <-
  stratified_performance(ga_res_conds_imp_est_eval, "Res-Conds-GA*")

ga_res_conds_imp_est_eval_plot <-
  stratified_plot_element(ga_res_conds_imp_est_eval_results)
```

```{r ResCondsPred-GA model evaluation, include=FALSE}
ga_res_conds_pred_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_conds_pred_est_eval_results <-
  stratified_performance(ga_res_conds_pred_est_eval, "Res-Conds-GA")

ga_res_conds_pred_est_eval_plot <-
  stratified_plot_element(ga_res_conds_pred_est_eval_results)
```

```{r ResCondsNull-GA model evaluation data, include=FALSE}
ga_res_conds_null_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$null |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_conds_null_est_eval_results <-
  stratified_performance(ga_res_conds_null_est_eval, "Res-Conds-GA")

ga_res_conds_null_est_eval_plot <-
  stratified_plot_element(ga_res_conds_null_est_eval_results)
```

Eventually, we were inspired to combine Res-CPG- and Res-Conds-GA modeling approaches. Res-Comb-GA model won in almost all subgroups, except 1 dataset of origin (GSE228149) (Figure 4), and was almost the same RMSE with the top performer in the validation set (1.077 vs. 1.076) (Table 1). Both r values were 0.966, but MAE of Res-Comb-GA was slightly higher than the top performer (0.888 vs. 0.863). Meanwhile, Res-CPG-Comb-GA model had the poorest performance compared to all of the other multistage prediction models (Figure 3). This finding underlines the importance of phenotypic-related information, although it was not perfectly accurate in predicting the conditions. Therefore, considering the robustness across subgroups, Res-Comb-GA model has a higher chance to win in the test set, particularly in sub-challenge 1. Hence, we stopped the iterative process in model development.

```{r ResComb-pr-GA model evaluation, include=FALSE}
ga_res_comb_pr_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    ga_res_comb_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,0
      )
  ) |>
  select(-residual_prediction_pr) |>
  add_strata_data()

ga_res_comb_pr_est_eval_results <-
  stratified_performance(ga_res_comb_pr_est_eval, "Res-Comb-PR-GA")

ga_res_comb_pr_est_eval_plot <-
  stratified_plot_element(ga_res_comb_pr_est_eval_results)
```

```{r ResComb-prtb-GA model evaluation, include=FALSE}
ga_res_comb_prtb_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    ga_res_comb_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_comb_tb_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_tb = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(
          ceiling(prediction) >= 37 & ceiling(prediction) <= 40
          ,ifelse(is.na(residual_prediction_tb), 0, residual_prediction_tb)
          ,0
        )
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_tb) |>
  add_strata_data()

ga_res_comb_prtb_est_eval_results <-
  stratified_performance(ga_res_comb_prtb_est_eval, "Res-Comb-PRTB-GA")

ga_res_comb_prtb_est_eval_plot <-
  stratified_plot_element(ga_res_comb_prtb_est_eval_results)
```

```{r ResComb-GA model evaluation, include=FALSE}
ga_res_comb_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_conds_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    ga_res_comb_pr_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_comb_tb_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_tb = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    ga_res_comb_ta_est_set_predictions |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction_ta = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(
          ceiling(prediction) >= 37 & ceiling(prediction) <= 40
          ,ifelse(is.na(residual_prediction_tb), 0, residual_prediction_tb)
          ,ifelse(is.na(residual_prediction_ta), 0, residual_prediction_ta)
        )
      )
  ) |>
  select(
    -residual_prediction_pr
    ,-residual_prediction_tb
    ,-residual_prediction_ta
  ) |>
  add_strata_data()

ga_res_comb_est_eval_results <-
  stratified_performance(ga_res_comb_est_eval, "Res-Comb-GA")

ga_res_comb_est_eval_plot <-
  stratified_plot_element(ga_res_comb_est_eval_results)
```

```{r ResCPGcombTrue-GA model evaluation, include=FALSE}
ga_res_cpg_comb_true_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_comb_all_est_set_predictions$true |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cpg_comb_true_est_eval_results <-
  stratified_performance(ga_res_cpg_comb_true_est_eval, "Res-CPG-Comb-GA")

ga_res_cpg_comb_true_est_eval_plot <-
  stratified_plot_element(ga_res_cpg_comb_true_est_eval_results)
```

```{r ResCPGcombImp-GA model evaluation, include=FALSE}
ga_res_cpg_comb_imp_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_comb_all_est_set_predictions$imp |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cpg_comb_imp_est_eval_results <-
  stratified_performance(ga_res_cpg_comb_imp_est_eval, "Res-CPG-Comb-GA")

ga_res_cpg_comb_imp_est_eval_plot <-
  stratified_plot_element(ga_res_cpg_comb_imp_est_eval_results)
```

```{r ResCPGcombPred-GA model evaluation, include=FALSE}
ga_res_cpg_comb_pred_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_comb_all_est_set_predictions$pred |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cpg_comb_pred_est_eval_results <-
  stratified_performance(ga_res_cpg_comb_pred_est_eval, "Res-CPG-Comb-GA")

ga_res_cpg_comb_pred_est_eval_plot <-
  stratified_plot_element(ga_res_cpg_comb_pred_est_eval_results)
```

```{r ResCPGcombNull-GA model evaluation data, include=FALSE}
ga_res_cpg_comb_null_est_eval <-
  ga_est_set_predictions |>
  rownames_to_column(var = "Sample_accession") |>
  left_join(
    ga_res_cpg_comb_all_est_set_predictions$null |>
      rownames_to_column(var = "Sample_accession") |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  add_strata_data()

ga_res_cpg_comb_null_est_eval_results <-
  stratified_performance(ga_res_cpg_comb_null_est_eval, "Res-CPG-Comb-GA")

ga_res_cpg_comb_null_est_eval_plot <-
  stratified_plot_element(ga_res_cpg_comb_null_est_eval_results)
```

### Performance comparison

```{r Compile evaluation results, include=FALSE}
eval_results <-
  rbind(
    ga_est_eval_results
    ,ga_rtc_est_eval_results
    ,ga_rtc_rf_est_eval_results
    ,ga_res_est_eval_results
    ,ga_res_rf_est_eval_results
    ,ga_res_cr_est_eval_results
    ,ga_res_cr_rf_est_eval_results
    ,ga_resfull_est_eval_results
    ,ga_resfull_rf_est_eval_results
    ,ga_res_sr_est_eval_results
    ,ga_res_sr_rf_est_eval_results
    ,ga_res_cpg_pr_est_eval_results
    ,ga_res_cpg_est_eval_results
    ,ga_res_cpgrev_est_eval_results
    ,ga_res_conds_true_est_eval_results
    ,ga_res_conds_imp_est_eval_results
    ,ga_res_conds_pred_est_eval_results
    ,ga_res_conds_null_est_eval_results
    ,ga_res_comb_pr_est_eval_results
    ,ga_res_comb_prtb_est_eval_results
    ,ga_res_comb_est_eval_results
    ,ga_res_cpg_comb_true_est_eval_results
    ,ga_res_cpg_comb_imp_est_eval_results
    ,ga_res_cpg_comb_pred_est_eval_results
    ,ga_res_cpg_comb_null_est_eval_results
  ) |>
  filter(strata == "all") |>
  select(-strata, -level) |>
  select(model, everything()) |>
  mutate_at("model", \(x) factor(x, unique(x))) |>
  arrange(model, metric)
```

```{r table-s9, echo=FALSE}
eval_results |>
  left_join(submission, by = join_by(model)) |>
  left_join(performance, by = join_by(sub, metric)) |>
  select(-all, -robust) |>
  mutate_all(\(x) ifelse(is.na(x), "", x)) |>
  kable(caption = "Table S9. Model evaluation") |>
  kable_classic()
```

```{r Generate figure-s7, include=FALSE}
figure_s7 <-
  eval_results |>
  mutate_at("model", \(x) factor(x, rev(levels(x)))) |>
  mutate(
    iteration =
      case_when(
        str_detect(model, "Normal|Res-GA|Res-CR|Resfull|Res-Seq") ~ "Iteration 1"
        ,str_detect(model, "Res-CPG-PR|Res-CPG-GA|Res-CPG-rev") ~ "Iteration 2"
        ,str_detect(model, "Res-Conds") ~ "Iteration 3"
        ,TRUE ~ "Iteration 4"
      )
  ) |>
  ggplot(aes(model, avg, color = win)) +
  geom_errorbar(aes(ymin = lb, ymax = ub), width = 1) +
  geom_point() +
  geom_label(
    data =
      submission |>
      right_join(
        performance
        ,by = join_by(sub)
        ,relationship = "many-to-many"
      ) |>
      left_join(leaderboard, by = join_by(metric)) |>
      mutate(
        metric = factor(metric, c("RMSE", "MAE", "r"))
        ,lb = val
        ,ub = val
        ,win =
          ifelse(
            metric == "r"
            ,ifelse(lb > current_best, "Yes", "No")
            ,ifelse(ub < current_best, "Yes", "No")
          )
      ) |>
      mutate(
        iteration =
          case_when(
            str_detect(model, "Normal|Res-GA|Res-CR|Resfull|Res-Seq") ~ "Iteration 1"
            ,str_detect(model, "Res-CPG-PR|Res-CPG-GA|Res-CPG-rev") ~ "Iteration 2"
            ,str_detect(model, "Res-Conds") ~ "Iteration 3"
            ,TRUE ~ "Iteration 4"
          )
      )
    ,aes(model, val, label = rank)
    ,size = 2.5
    ,label.padding = unit(0.1, "line")
    ,na.rm = TRUE
    ,show.legend = FALSE
  ) +
  facet_grid(iteration ~ metric, scales = "free", space = "free_y") +
  xlab("") +
  ylab("") +
  scale_color_discrete("Win") +
  coord_flip()
```

```{r figure-s7, echo=FALSE, fig.height=5, fig.width=12}
figure_s7
```

```{r Save figure-s7, eval=FALSE, include=FALSE}
ggsave("doc/figure2a.eps", height = 5, width = 12, units = "in")
ggsave("doc/figure2a.svg", height = 5, width = 12, units = "in")
```

```{r Generate figure-s8, include=FALSE}
figure_s8 <-
  stratified_plot(
    ga_est_eval_plot, "Normal-GA"
    ,ga_res_cpgrev_est_eval_plot, "Res-CPG-rev-GA"
    ,ga_res_conds_pred_est_eval_plot, "Res-Conds-GA*"
    ,ga_res_comb_est_eval_plot, "Res-Comb-GA"
  )
```

```{r figure-s8, echo=FALSE, fig.height=9, fig.width=12}
figure_s8
```

```{r Save figure-s8, eval=FALSE, include=FALSE}
ggsave("doc/figure3a.eps", height = 9, width = 12, units = "in")
ggsave("doc/figure3a.svg", height = 9, width = 12, units = "in")
```

```{r figure-s9, echo=FALSE}
list(
    ga_est_eval
    ,ga_res_cpgrev_est_eval
    ,ga_res_conds_pred_est_eval
    ,ga_res_comb_est_eval
  ) |>
  `names<-`(
    c(ga_est_eval_results$model[1]
      ,ga_res_cpgrev_est_eval_results$model[1]
      ,ga_res_conds_pred_est_eval_results$model[1]
      ,ga_res_comb_est_eval_results$model[1]
    )
  ) |>
  imap(
    ~ .x |>
      ggplot(aes(Y, Yhat, color = Dataset_ID)) +
      geom_point(alpha = 0.25, show.legend = FALSE, na.rm = TRUE) +
      geom_abline(intercept = 0, slope = 1, lty = 2) +
      geom_smooth(
        method = "lm"
        ,formula = y ~ x
        ,show.legend = FALSE
        ,na.rm = TRUE
      ) +
      geom_smooth(
        method = "lm"
        ,formula = y ~ x
        ,color = "black"
        ,show.legend = FALSE
        ,na.rm = TRUE
      ) +
      coord_equal() +
      scale_x_continuous("GA", limits = c(5, 44)) +
      scale_y_continuous(paste0("Predicted GA (", .y, ")"), limits = c(5, 44))
  )
```

## Best model predictors

```{r ResComb-GA model predictor, include=FALSE}
ga_res_comb_est_coefficients <-
  list(
    ga_est = ga_est_coefficients
    
    ,fgr_pred = fgr_pred_coefficients
    ,pe_pred = pe_pred_coefficients
    ,pe_onset_pred = pe_onset_pred_coefficients
    ,hellp_pred = hellp_pred_coefficients
    ,anencephaly_pred = anencephaly_pred_coefficients
    ,spina_bifida_pred = spina_bifida_pred_coefficients
    ,diandric_triploid_pred = diandric_triploid_pred_coefficients
    ,miscarriage_pred = miscarriage_pred_coefficients
    ,preterm_pred = preterm_pred_coefficients
    ,gdm_pred = gdm_pred_coefficients
    ,lga_pred = lga_pred_coefficients
    ,subfertility_pred = subfertility_pred_coefficients
    ,chorioamnionitis_pred = chorioamnionitis_pred_coefficients
    
    ,ga_res_conds_fgr_est = ga_res_conds_est_coefficients$fgr
    ,ga_res_conds_pe_est = ga_res_conds_est_coefficients$pe
    ,ga_res_conds_pe_onset_est = ga_res_conds_est_coefficients$pe_onset
    ,ga_res_conds_hellp_est = ga_res_conds_est_coefficients$hellp
    ,ga_res_conds_anencephaly_est = ga_res_conds_est_coefficients$anencephaly
    ,ga_res_conds_spina_bifida_est = ga_res_conds_est_coefficients$spina_bifida
    ,ga_res_conds_diandric_triploid_est =
      ga_res_conds_est_coefficients$diandric_triploid
    ,ga_res_conds_miscarriage_est = ga_res_conds_est_coefficients$miscarriage
    ,ga_res_conds_preterm_est = ga_res_conds_est_coefficients$preterm
    ,ga_res_conds_gdm_est = ga_res_conds_est_coefficients$gdm
    ,ga_res_conds_lga_est = ga_res_conds_est_coefficients$lga
    ,ga_res_conds_subfertility_est = ga_res_conds_est_coefficients$subfertility
    ,ga_res_conds_chorioamnionitis_est =
      ga_res_conds_est_coefficients$chorioamnionitis
    
    ,ga_res_comb_pr_est = ga_res_comb_pr_est_coefficients
    ,ga_res_comb_tb_est = ga_res_comb_tb_est_coefficients
    ,ga_res_comb_ta_est = ga_res_comb_ta_est_coefficients
  ) |>
  imap(~ mutate(.x, model = .y)) |>
  lapply(rownames_to_column, var = "probe_id") |>
  reduce(rbind) |>
  mutate_at(c("probe_id", "model"), \(x) factor(x, unique(x))) |>
  spread(model, weight, fill = 0) |>
  column_to_rownames(var = "probe_id")
```

```{r Write ResComb-GA model predictor, eval=FALSE, include=FALSE}
ga_res_comb_est_coefficients |>
  rownames_to_column(var = "cpg_name") |>
  write_csv("inst/extdata/ga_res_comb_est_coefficients.csv")
```

# Model submission

```{bash Submission login, eval=FALSE, include=FALSE}
docker login docker.synapse.org --username herdiants # Synapse authToken (PAT)
```

```{r Create input and output directories, eval=FALSE, include=FALSE}
if (!dir.exists("input")) {
  dir.create("input", recursive = TRUE)
}

if (!dir.exists("output")) {
  dir.create("output", recursive = TRUE)
}
```

```{r Simulate leaderboard input sample IDs, eval=FALSE, include=FALSE}
Sample_IDs <- readRDS("data/Sample_IDs.rds")
```

```{r Simulate leaderboard input data, eval=FALSE, include=FALSE}
df_train0 |>
  select_at(Sample_IDs) |>
  rownames_to_column(var = "probe_id") |>
  write_csv("input/Leaderboard_beta_subchallenge1.csv")
```

## Submission 1: GA res-comp-risk random forest (testthewaters)

```{r GA rcrRF prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub1/data")) {
  dir.create("pcd1_sub1/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub1/inst/extdata")) {
  dir.create("pcd1_sub1/inst/extdata", recursive = TRUE)
}
```

```{r GA rcrRF prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub1_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    ,"data/var_ga.rds"
    
    ,"data/fgr_dmp.rds"
    ,"data/pe_dmp.rds"
    ,"data/pe_onset_dmp.rds"
    ,"data/preterm_dmp.rds"
    ,"data/anencephaly_dmp.rds"
    ,"data/spina_bifida_dmp.rds"
    ,"data/gdm_dmp.rds"
    ,"data/diandric_triploid_dmp.rds"
    ,"data/miscarriage_dmp.rds"
    ,"data/lga_dmp.rds"
    ,"data/subfertility_dmp.rds"
    ,"data/hellp_dmp.rds"
    ,"data/chorioamnionitis_dmp.rds"
    
    ,"data/ga_dmp.rds"
    
    ,"inst/extdata/fgr_imp_scaler.joblib"
    ,"inst/extdata/pe_imp_scaler.joblib"
    ,"inst/extdata/pe_onset_imp_scaler.joblib"
    ,"inst/extdata/preterm_imp_scaler.joblib"
    ,"inst/extdata/anencephaly_imp_scaler.joblib"
    ,"inst/extdata/spina_bifida_imp_scaler.joblib"
    ,"inst/extdata/gdm_imp_scaler.joblib"
    ,"inst/extdata/diandric_triploid_imp_scaler.joblib"
    ,"inst/extdata/miscarriage_imp_scaler.joblib"
    ,"inst/extdata/lga_imp_scaler.joblib"
    ,"inst/extdata/subfertility_imp_scaler.joblib"
    ,"inst/extdata/hellp_imp_scaler.joblib"
    ,"inst/extdata/chorioamnionitis_imp_scaler.joblib"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_cr_rf_est_scaler.joblib"
    
    ,"inst/extdata/fgr_imp_best_model.joblib"
    ,"inst/extdata/pe_imp_best_model.joblib"
    ,"inst/extdata/pe_onset_imp_best_model.joblib"
    ,"inst/extdata/preterm_imp_best_model.joblib"
    ,"inst/extdata/anencephaly_imp_best_model.joblib"
    ,"inst/extdata/spina_bifida_imp_best_model.joblib"
    ,"inst/extdata/gdm_imp_best_model.joblib"
    ,"inst/extdata/diandric_triploid_imp_best_model.joblib"
    ,"inst/extdata/miscarriage_imp_best_model.joblib"
    ,"inst/extdata/lga_imp_best_model.joblib"
    ,"inst/extdata/subfertility_imp_best_model.joblib"
    ,"inst/extdata/hellp_imp_best_model.joblib"
    ,"inst/extdata/chorioamnionitis_imp_best_model.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_cr_rf_est_best_model.joblib"
  )

pcd1_sub1_files_non_default |>
  file.copy(paste0("pcd1_sub1/", pcd1_sub1_files_non_default))
```

```{bash GA rcrRF prediction - Build docker locally, eval=FALSE, include=FALSE}
cd pcd1_sub1
docker build --load -t pcd1_sub1:latest .
cd ..
```

```{bash GA rcrRF prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub1:latest
```

```{r GA rcrRF prediction - Results during dev, eval=FALSE, include=FALSE}
ga_res_cr_rf_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_res_cr_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(
    Sample_accession =
      readRDS("data/ga_res_cr_rf_est_set_rownames.rds")
  ) |>
  rename(residual_prediction = prediction) |>
  left_join(
    read_csv(
        "inst/extdata/ga_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds"))
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r GA rcrRF prediction - Results during dep, eval=FALSE, include=FALSE}
ga_res_cr_rf_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_cr_rf_est_set_predictions_dep_file) == 1){
  ga_res_cr_rf_est_set_predictions_dep_file |>
    file.rename(
      ga_res_cr_rf_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub1_")
    )
}

ga_res_cr_rf_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub1_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r GA rcrRF prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_res_cr_rf_est_set_predictions_dep |>
  left_join(
    ga_res_cr_rf_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash GA rcrRF prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub1
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-cr-rf-est:v1.6 --push .
cd ..
```

## Submission 2: GA res-full random forest (isitthedarkhorse)

```{r GA rfullRF prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub2/data")) {
  dir.create("pcd1_sub2/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub2/inst/extdata")) {
  dir.create("pcd1_sub2/inst/extdata", recursive = TRUE)
}
```

```{r GA rfullRF prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub2_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    ,"data/var_ga.rds"
    
    ,"data/fgr_dmp.rds"
    ,"data/pe_dmp.rds"
    ,"data/pe_onset_dmp.rds"
    ,"data/preterm_dmp.rds"
    ,"data/anencephaly_dmp.rds"
    ,"data/spina_bifida_dmp.rds"
    ,"data/gdm_dmp.rds"
    ,"data/diandric_triploid_dmp.rds"
    ,"data/miscarriage_dmp.rds"
    ,"data/lga_dmp.rds"
    ,"data/subfertility_dmp.rds"
    ,"data/hellp_dmp.rds"
    ,"data/chorioamnionitis_dmp.rds"
    
    ,"data/ga_dmp.rds"
    
    ,"inst/extdata/fgr_pred_scaler.joblib"
    ,"inst/extdata/pe_pred_scaler.joblib"
    ,"inst/extdata/pe_onset_pred_scaler.joblib"
    ,"inst/extdata/preterm_pred_scaler.joblib"
    ,"inst/extdata/anencephaly_pred_scaler.joblib"
    ,"inst/extdata/spina_bifida_pred_scaler.joblib"
    ,"inst/extdata/gdm_pred_scaler.joblib"
    ,"inst/extdata/diandric_triploid_pred_scaler.joblib"
    ,"inst/extdata/miscarriage_pred_scaler.joblib"
    ,"inst/extdata/lga_pred_scaler.joblib"
    ,"inst/extdata/subfertility_pred_scaler.joblib"
    ,"inst/extdata/hellp_pred_scaler.joblib"
    ,"inst/extdata/chorioamnionitis_pred_scaler.joblib"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_resfull_rf_est_scaler.joblib"
    
    ,"inst/extdata/fgr_pred_best_model.joblib"
    ,"inst/extdata/pe_pred_best_model.joblib"
    ,"inst/extdata/pe_onset_pred_best_model.joblib"
    ,"inst/extdata/preterm_pred_best_model.joblib"
    ,"inst/extdata/anencephaly_pred_best_model.joblib"
    ,"inst/extdata/spina_bifida_pred_best_model.joblib"
    ,"inst/extdata/gdm_pred_best_model.joblib"
    ,"inst/extdata/diandric_triploid_pred_best_model.joblib"
    ,"inst/extdata/miscarriage_pred_best_model.joblib"
    ,"inst/extdata/lga_pred_best_model.joblib"
    ,"inst/extdata/subfertility_pred_best_model.joblib"
    ,"inst/extdata/hellp_pred_best_model.joblib"
    ,"inst/extdata/chorioamnionitis_pred_best_model.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_resfull_rf_est_best_model.joblib"
  )

pcd1_sub2_files_non_default |>
  file.copy(paste0("pcd1_sub2/", pcd1_sub2_files_non_default))
```

```{bash GA rfullRF prediction - Build docker local, eval=FALSE, include=FALSE}
cd pcd1_sub2
docker build --load -t pcd1_sub2:latest .
cd ..
```

```{bash GA rfullRF prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub2:latest
```

```{r GA rfullRF prediction - Results during dev, eval=FALSE, include=FALSE}
ga_resfull_rf_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_resfull_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(
    Sample_accession =
      readRDS("data/ga_resfull_rf_est_set_rownames.rds")
  ) |>
  rename(residual_prediction = prediction) |>
  left_join(
    read_csv(
        "inst/extdata/ga_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds"))
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r GA rfullRF prediction - Results during dep, eval=FALSE, include=FALSE}
ga_resfull_rf_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_resfull_rf_est_set_predictions_dep_file) == 1){
  ga_resfull_rf_est_set_predictions_dep_file |>
    file.rename(
      ga_resfull_rf_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub2_")
    )
}

ga_resfull_rf_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub2_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r GA rfullRF prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_resfull_rf_est_set_predictions_dep |>
  left_join(
    ga_resfull_rf_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash GA rfullRF prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub2
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-resfull-rf-est:v1.0 --push .
cd ..
```

## Submission 3: GA residual random forest (clearcut)

```{r GA res RF prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub3/data")) {
  dir.create("pcd1_sub3/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub3/inst/extdata")) {
  dir.create("pcd1_sub3/inst/extdata", recursive = TRUE)
}
```

```{r GA res RF prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub3_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    ,"data/var_ga.rds"
    
    ,"data/fgr_dmp.rds"
    ,"data/pe_dmp.rds"
    ,"data/pe_onset_dmp.rds"
    ,"data/preterm_dmp.rds"
    ,"data/anencephaly_dmp.rds"
    ,"data/spina_bifida_dmp.rds"
    ,"data/gdm_dmp.rds"
    ,"data/diandric_triploid_dmp.rds"
    ,"data/miscarriage_dmp.rds"
    ,"data/lga_dmp.rds"
    ,"data/subfertility_dmp.rds"
    ,"data/hellp_dmp.rds"
    ,"data/chorioamnionitis_dmp.rds"
    
    ,"data/ga_dmp.rds"
    
    ,"inst/extdata/fgr_imp_scaler.joblib"
    ,"inst/extdata/pe_imp_scaler.joblib"
    ,"inst/extdata/pe_onset_imp_scaler.joblib"
    ,"inst/extdata/preterm_imp_scaler.joblib"
    ,"inst/extdata/anencephaly_imp_scaler.joblib"
    ,"inst/extdata/spina_bifida_imp_scaler.joblib"
    ,"inst/extdata/gdm_imp_scaler.joblib"
    ,"inst/extdata/diandric_triploid_imp_scaler.joblib"
    ,"inst/extdata/miscarriage_imp_scaler.joblib"
    ,"inst/extdata/lga_imp_scaler.joblib"
    ,"inst/extdata/subfertility_imp_scaler.joblib"
    ,"inst/extdata/hellp_imp_scaler.joblib"
    ,"inst/extdata/chorioamnionitis_imp_scaler.joblib"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_rf_est_scaler.joblib"
    
    ,"inst/extdata/fgr_imp_best_model.joblib"
    ,"inst/extdata/pe_imp_best_model.joblib"
    ,"inst/extdata/pe_onset_imp_best_model.joblib"
    ,"inst/extdata/preterm_imp_best_model.joblib"
    ,"inst/extdata/anencephaly_imp_best_model.joblib"
    ,"inst/extdata/spina_bifida_imp_best_model.joblib"
    ,"inst/extdata/gdm_imp_best_model.joblib"
    ,"inst/extdata/diandric_triploid_imp_best_model.joblib"
    ,"inst/extdata/miscarriage_imp_best_model.joblib"
    ,"inst/extdata/lga_imp_best_model.joblib"
    ,"inst/extdata/subfertility_imp_best_model.joblib"
    ,"inst/extdata/hellp_imp_best_model.joblib"
    ,"inst/extdata/chorioamnionitis_imp_best_model.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_rf_est_best_model.joblib"
  )

pcd1_sub3_files_non_default |>
  file.copy(paste0("pcd1_sub3/", pcd1_sub3_files_non_default))
```

```{bash GA res RF prediction - Build docker locally, eval=FALSE, include=FALSE}
cd pcd1_sub3
docker build --load -t pcd1_sub3:latest .
cd ..
```

```{bash GA res RF prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub3:latest
```

```{r GA res RF prediction - Results during dev, eval=FALSE, include=FALSE}
ga_res_rf_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_res_rf_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(
    Sample_accession =
      readRDS("data/ga_res_rf_est_set_rownames.rds")
  ) |>
  rename(residual_prediction = prediction) |>
  left_join(
    read_csv(
        "inst/extdata/ga_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds"))
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r GA res RF prediction - Results during dep, eval=FALSE, include=FALSE}
ga_res_rf_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_rf_est_set_predictions_dep_file) == 1){
  ga_res_rf_est_set_predictions_dep_file |>
    file.rename(
      ga_res_rf_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub3_")
    )
}

ga_res_rf_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub3_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r GA res RF prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_res_rf_est_set_predictions_dep |>
  left_join(
    ga_res_rf_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash GA res RF prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub3
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-rf-est:v1.0 --push .
cd ..
```

## Submission 4: GA res-cpg-pr elastic net (stepback)

```{r RcpgPR-GA prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub4/data")) {
  dir.create("pcd1_sub4/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub4/inst/extdata")) {
  dir.create("pcd1_sub4/inst/extdata", recursive = TRUE)
}
```

```{r RcpgPR-GA prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub4_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    
    ,"data/ga_dmp.rds"
    ,"data/ga_res_cpg_pr_dmp.rds"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_scaler.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_best_model.joblib"
  )

pcd1_sub4_files_non_default |>
  file.copy(paste0("pcd1_sub4/", pcd1_sub4_files_non_default))
```

```{bash RcpgPR-GA prediction - Build docker local, eval=FALSE, include=FALSE}
cd pcd1_sub4
docker build --load -t pcd1_sub4:latest .
cd ..
```

```{bash RcpgPR-GA prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub4:latest
```

```{r RcpgPR-GA prediction - Results during dev, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds")) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_cpg_pr_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_cpg_pr_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,0
      )
  ) |>
  select(-residual_prediction_pr) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r RcpgPR-GA prediction - Results during dep, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_cpg_pr_est_set_predictions_dep_file) == 1){
  ga_res_cpg_pr_est_set_predictions_dep_file |>
    file.rename(
      ga_res_cpg_pr_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub4_")
    )
}

ga_res_cpg_pr_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub4_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r RcpgPR-GA prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_res_cpg_pr_est_set_predictions_dep |>
  left_join(
    ga_res_cpg_pr_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash RcpgPR-GA prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub4
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-cpg-pr-est:v1.0 --push .
cd ..
```

## Submission 5: GA res-cpg elastic net (stepbackfurther)

```{r Rcpg-GA prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub5/data")) {
  dir.create("pcd1_sub5/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub5/inst/extdata")) {
  dir.create("pcd1_sub5/inst/extdata", recursive = TRUE)
}
```

```{r Rcpg-GA prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub5_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    
    ,"data/ga_dmp.rds"
    ,"data/ga_res_cpg_pr_dmp.rds"
    ,"data/ga_res_cpg_thr_dmp.rds"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_scaler.joblib"
    ,"inst/extdata/ga_res_cpg_thr_est_scaler.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_best_model.joblib"
    ,"inst/extdata/ga_res_cpg_thr_est_best_model.joblib"
  )

pcd1_sub5_files_non_default |>
  file.copy(paste0("pcd1_sub5/", pcd1_sub5_files_non_default))
```

```{bash Rcpg-GA prediction - Build docker local, eval=FALSE, include=FALSE}
cd pcd1_sub5
docker build --load -t pcd1_sub5:latest .
cd ..
```

```{bash Rcpg-GA prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub5:latest
```

```{r Rcpg-GA prediction - Results during dev, eval=FALSE, include=FALSE}
ga_res_cpg_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds")) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_cpg_pr_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_cpg_pr_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_cpg_thr_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_cpg_thr_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_thr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(is.na(residual_prediction_thr), 0, residual_prediction_thr)
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_thr) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r Rcpg-GA prediction - Results during dep, eval=FALSE, include=FALSE}
ga_res_cpg_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_cpg_est_set_predictions_dep_file) == 1){
  ga_res_cpg_est_set_predictions_dep_file |>
    file.rename(
      ga_res_cpg_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub5_")
    )
}

ga_res_cpg_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub5_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r Rcpg-GA prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_res_cpg_est_set_predictions_dep |>
  left_join(
    ga_res_cpg_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash Rcpg-GA prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub5
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-cpg-est:v1.0 --push .
cd ..
```

## Submission 6: GA res-cpg-rev elastic net (stepbackabit)

```{r Rcpgrev-GA prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub6/data")) {
  dir.create("pcd1_sub6/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub6/inst/extdata")) {
  dir.create("pcd1_sub6/inst/extdata", recursive = TRUE)
}
```

```{r Rcpgrev-GA prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub6_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    
    ,"data/ga_dmp.rds"
    ,"data/ga_res_cpg_pr_dmp.rds"
    ,"data/ga_res_cpg_t_dmp.rds"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_scaler.joblib"
    ,"inst/extdata/ga_res_cpg_t_est_scaler.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_cpg_pr_est_best_model.joblib"
    ,"inst/extdata/ga_res_cpg_t_est_best_model.joblib"
  )

pcd1_sub6_files_non_default |>
  file.copy(paste0("pcd1_sub6/", pcd1_sub6_files_non_default))
```

```{bash Rcpgrev-GA prediction - Build docker local, eval=FALSE, include=FALSE}
cd pcd1_sub6
docker build --load -t pcd1_sub6:latest .
cd ..
```

```{bash Rcpgrev-GA prediction - Test docker locally, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub6:latest
```

```{r Rcpgrev-GA prediction - Results during dev, eval=FALSE, include=FALSE}
ga_res_cpgrev_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds")) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_cpg_pr_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_cpg_pr_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_cpg_t_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_cpg_t_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_t = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(is.na(residual_prediction_t), 0, residual_prediction_t)
      )
  ) |>
  select(-residual_prediction_pr, -residual_prediction_t) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r Rcpgrev-GA prediction - Results during dep, eval=FALSE, include=FALSE}
ga_res_cpgrev_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_cpgrev_est_set_predictions_dep_file) == 1){
  ga_res_cpgrev_est_set_predictions_dep_file |>
    file.rename(
      ga_res_cpgrev_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub6_")
    )
}

ga_res_cpgrev_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub6_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r Rcpgrev-GA prediction - Compare dev and dep, eval=FALSE, include=FALSE}
ga_res_cpgrev_est_set_predictions_dep |>
  left_join(
    ga_res_cpgrev_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.0000001)
```

```{bash Rcpgrev-GA prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub6
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-cpg-rev-est:v1.0 --push .
cd ..
```

## Submission 7: GA res-conds-pred elastic net (slidingdoors)

```{r ResCondsPred-GA prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub7/data")) {
  dir.create("pcd1_sub7/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub7/inst/extdata")) {
  dir.create("pcd1_sub7/inst/extdata", recursive = TRUE)
}
```

```{r ResCondsPred-GA prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub7_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    ,"data/var_ga.rds"
    
    ,"data/ga_res_conds_fgr_dmp.rds"
    ,"data/ga_res_conds_pe_dmp.rds"
    ,"data/ga_res_conds_pe_onset_dmp.rds"
    ,"data/ga_res_conds_preterm_dmp.rds"
    ,"data/ga_res_conds_anencephaly_dmp.rds"
    ,"data/ga_res_conds_spina_bifida_dmp.rds"
    ,"data/ga_res_conds_gdm_dmp.rds"
    ,"data/ga_res_conds_diandric_triploid_dmp.rds"
    ,"data/ga_res_conds_miscarriage_dmp.rds"
    ,"data/ga_res_conds_lga_dmp.rds"
    ,"data/ga_res_conds_subfertility_dmp.rds"
    ,"data/ga_res_conds_hellp_dmp.rds"
    ,"data/ga_res_conds_chorioamnionitis_dmp.rds"
    
    ,"data/fgr_dmp.rds"
    ,"data/pe_dmp.rds"
    ,"data/pe_onset_dmp.rds"
    ,"data/preterm_dmp.rds"
    ,"data/anencephaly_dmp.rds"
    ,"data/spina_bifida_dmp.rds"
    ,"data/gdm_dmp.rds"
    ,"data/diandric_triploid_dmp.rds"
    ,"data/miscarriage_dmp.rds"
    ,"data/lga_dmp.rds"
    ,"data/subfertility_dmp.rds"
    ,"data/hellp_dmp.rds"
    ,"data/chorioamnionitis_dmp.rds"
    
    ,"data/ga_dmp.rds"
    
    ,"inst/extdata/ga_res_conds_fgr_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pe_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pe_onset_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_preterm_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_anencephaly_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_spina_bifida_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_gdm_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_diandric_triploid_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_miscarriage_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_lga_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_subfertility_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_hellp_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_chorioamnionitis_est_scaler.joblib"
    
    ,"inst/extdata/fgr_pred_scaler.joblib"
    ,"inst/extdata/pe_pred_scaler.joblib"
    ,"inst/extdata/pe_onset_pred_scaler.joblib"
    ,"inst/extdata/preterm_pred_scaler.joblib"
    ,"inst/extdata/anencephaly_pred_scaler.joblib"
    ,"inst/extdata/spina_bifida_pred_scaler.joblib"
    ,"inst/extdata/gdm_pred_scaler.joblib"
    ,"inst/extdata/diandric_triploid_pred_scaler.joblib"
    ,"inst/extdata/miscarriage_pred_scaler.joblib"
    ,"inst/extdata/lga_pred_scaler.joblib"
    ,"inst/extdata/subfertility_pred_scaler.joblib"
    ,"inst/extdata/hellp_pred_scaler.joblib"
    ,"inst/extdata/chorioamnionitis_pred_scaler.joblib"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pred_est_scaler.joblib"
    
    ,"inst/extdata/ga_res_conds_fgr_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pe_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pe_onset_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_preterm_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_anencephaly_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_spina_bifida_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_gdm_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_diandric_triploid_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_miscarriage_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_lga_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_subfertility_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_hellp_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_chorioamnionitis_est_best_model.joblib"
    
    ,"inst/extdata/fgr_pred_best_model.joblib"
    ,"inst/extdata/pe_pred_best_model.joblib"
    ,"inst/extdata/pe_onset_pred_best_model.joblib"
    ,"inst/extdata/preterm_pred_best_model.joblib"
    ,"inst/extdata/anencephaly_pred_best_model.joblib"
    ,"inst/extdata/spina_bifida_pred_best_model.joblib"
    ,"inst/extdata/gdm_pred_best_model.joblib"
    ,"inst/extdata/diandric_triploid_pred_best_model.joblib"
    ,"inst/extdata/miscarriage_pred_best_model.joblib"
    ,"inst/extdata/lga_pred_best_model.joblib"
    ,"inst/extdata/subfertility_pred_best_model.joblib"
    ,"inst/extdata/hellp_pred_best_model.joblib"
    ,"inst/extdata/chorioamnionitis_pred_best_model.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pred_est_best_model.joblib"
  )

pcd1_sub7_files_non_default |>
  file.copy(paste0("pcd1_sub7/", pcd1_sub7_files_non_default))
```

```{bash ResCondsPred-GA prediction - Build loc, eval=FALSE, include=FALSE}
cd pcd1_sub7
docker build --load -t pcd1_sub7:latest .
cd ..
```

```{bash ResCondsPred-GA prediction - Test docker loc, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub7:latest
```

```{r ResCondsPred-GA prediction - Results in dev, eval=FALSE, include=FALSE}
ga_res_conds_pred_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds")) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_conds_pred_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_conds_pred_est_set_rownames.rds")
      ) |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r ResCondsPred-GA prediction - Results in dep, eval=FALSE, include=FALSE}
ga_res_conds_pred_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_conds_pred_est_set_predictions_dep_file) == 1){
  ga_res_conds_pred_est_set_predictions_dep_file |>
    file.rename(
      ga_res_conds_pred_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub7_")
    )
}

ga_res_conds_pred_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub7_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r ResCondsPred-GA prediction - Comp dev and dep, eval=FALSE, include=FALSE}
ga_res_conds_pred_est_set_predictions_dep |>
  left_join(
    ga_res_conds_pred_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.000001)
```

```{bash ResCondsPred-GA prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub7
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-conds-pred-est:v1.0 --push .
cd ..
```

## Submission 8: GA res-comb-pred elastic net (pointofdivergence/imperfectmirror)

```{r ResComb-GA prediction - Create model dir, eval=FALSE, include=FALSE}
if (!dir.exists("pcd1_sub8/data")) {
  dir.create("pcd1_sub8/data", recursive = TRUE)
}

if (!dir.exists("pcd1_sub8/inst/extdata")) {
  dir.create("pcd1_sub8/inst/extdata", recursive = TRUE)
}
```

```{r ResComb-GA prediction - Copy the files, eval=FALSE, include=FALSE}
pcd1_sub8_files_non_default <-
  c("data/probe_info_epic.rds"
    ,"data/probe_info_others.rds"
    ,"data/beta_train_filter_detp_rownames.rds"
    ,"data/var_ga.rds"
    
    ,"data/ga_res_conds_fgr_dmp.rds"
    ,"data/ga_res_conds_pe_dmp.rds"
    ,"data/ga_res_conds_pe_onset_dmp.rds"
    ,"data/ga_res_conds_preterm_dmp.rds"
    ,"data/ga_res_conds_anencephaly_dmp.rds"
    ,"data/ga_res_conds_spina_bifida_dmp.rds"
    ,"data/ga_res_conds_gdm_dmp.rds"
    ,"data/ga_res_conds_diandric_triploid_dmp.rds"
    ,"data/ga_res_conds_miscarriage_dmp.rds"
    ,"data/ga_res_conds_lga_dmp.rds"
    ,"data/ga_res_conds_subfertility_dmp.rds"
    ,"data/ga_res_conds_hellp_dmp.rds"
    ,"data/ga_res_conds_chorioamnionitis_dmp.rds"
    
    ,"data/fgr_dmp.rds"
    ,"data/pe_dmp.rds"
    ,"data/pe_onset_dmp.rds"
    ,"data/preterm_dmp.rds"
    ,"data/anencephaly_dmp.rds"
    ,"data/spina_bifida_dmp.rds"
    ,"data/gdm_dmp.rds"
    ,"data/diandric_triploid_dmp.rds"
    ,"data/miscarriage_dmp.rds"
    ,"data/lga_dmp.rds"
    ,"data/subfertility_dmp.rds"
    ,"data/hellp_dmp.rds"
    ,"data/chorioamnionitis_dmp.rds"
    
    ,"data/ga_dmp.rds"
    ,"data/ga_res_comb_pr_dmp.rds"
    ,"data/ga_res_comb_tb_dmp.rds"
    ,"data/ga_res_comb_ta_dmp.rds"
    
    ,"inst/extdata/ga_res_conds_fgr_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pe_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pe_onset_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_preterm_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_anencephaly_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_spina_bifida_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_gdm_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_diandric_triploid_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_miscarriage_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_lga_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_subfertility_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_hellp_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_chorioamnionitis_est_scaler.joblib"
    
    ,"inst/extdata/fgr_pred_scaler.joblib"
    ,"inst/extdata/pe_pred_scaler.joblib"
    ,"inst/extdata/pe_onset_pred_scaler.joblib"
    ,"inst/extdata/preterm_pred_scaler.joblib"
    ,"inst/extdata/anencephaly_pred_scaler.joblib"
    ,"inst/extdata/spina_bifida_pred_scaler.joblib"
    ,"inst/extdata/gdm_pred_scaler.joblib"
    ,"inst/extdata/diandric_triploid_pred_scaler.joblib"
    ,"inst/extdata/miscarriage_pred_scaler.joblib"
    ,"inst/extdata/lga_pred_scaler.joblib"
    ,"inst/extdata/subfertility_pred_scaler.joblib"
    ,"inst/extdata/hellp_pred_scaler.joblib"
    ,"inst/extdata/chorioamnionitis_pred_scaler.joblib"
    
    ,"inst/extdata/ga_est_scaler.joblib"
    ,"inst/extdata/ga_res_conds_pred_est_scaler.joblib"
    ,"inst/extdata/ga_res_comb_pr_est_scaler.joblib"
    ,"inst/extdata/ga_res_comb_tb_est_scaler.joblib"
    ,"inst/extdata/ga_res_comb_ta_est_scaler.joblib"
    
    ,"inst/extdata/ga_res_conds_fgr_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pe_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pe_onset_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_preterm_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_anencephaly_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_spina_bifida_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_gdm_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_diandric_triploid_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_miscarriage_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_lga_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_subfertility_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_hellp_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_chorioamnionitis_est_best_model.joblib"
    
    ,"inst/extdata/fgr_pred_best_model.joblib"
    ,"inst/extdata/pe_pred_best_model.joblib"
    ,"inst/extdata/pe_onset_pred_best_model.joblib"
    ,"inst/extdata/preterm_pred_best_model.joblib"
    ,"inst/extdata/anencephaly_pred_best_model.joblib"
    ,"inst/extdata/spina_bifida_pred_best_model.joblib"
    ,"inst/extdata/gdm_pred_best_model.joblib"
    ,"inst/extdata/diandric_triploid_pred_best_model.joblib"
    ,"inst/extdata/miscarriage_pred_best_model.joblib"
    ,"inst/extdata/lga_pred_best_model.joblib"
    ,"inst/extdata/subfertility_pred_best_model.joblib"
    ,"inst/extdata/hellp_pred_best_model.joblib"
    ,"inst/extdata/chorioamnionitis_pred_best_model.joblib"
    
    ,"inst/extdata/ga_est_best_model.joblib"
    ,"inst/extdata/ga_res_conds_pred_est_best_model.joblib"
    ,"inst/extdata/ga_res_comb_pr_est_best_model.joblib"
    ,"inst/extdata/ga_res_comb_tb_est_best_model.joblib"
    ,"inst/extdata/ga_res_comb_ta_est_best_model.joblib"
  )

pcd1_sub8_files_non_default |>
  file.copy(paste0("pcd1_sub8/", pcd1_sub8_files_non_default))
```

```{bash ResComb-GA prediction - Build loc, eval=FALSE, include=FALSE}
cd pcd1_sub8
docker build --load -t pcd1_sub8:latest .
cd ..
```

```{bash ResComb-GA prediction - Test docker loc, eval=FALSE, include=FALSE}
docker run -v $(pwd)/input/:/input -v $(pwd)/output/:/output -it pcd1_sub8:latest
```

```{r ResComb-GA prediction - Results in dev, eval=FALSE, include=FALSE}
ga_res_comb_est_set_predictions_dev <-
  read_csv(
    "inst/extdata/ga_est_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(Sample_accession = readRDS("data/ga_est_set_rownames.rds")) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_conds_pred_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_conds_pred_est_set_rownames.rds")
      ) |>
      rename(residual_prediction = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(prediction = prediction + residual_prediction) |>
  select(-residual_prediction) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_comb_pr_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_comb_pr_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_pr = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_comb_tb_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_comb_tb_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_tb = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  left_join(
    read_csv(
        "inst/extdata/ga_res_comb_ta_est_predictions.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(
        Sample_accession =
          readRDS("data/ga_res_comb_ta_est_set_rownames.rds")
      ) |>
      rename(residual_prediction_ta = prediction)
    ,by = join_by(Sample_accession)
  ) |>
  mutate(
    prediction =
      prediction
      + ifelse(
        ceiling(prediction) <= 36
        ,ifelse(is.na(residual_prediction_pr), 0, residual_prediction_pr)
        ,ifelse(
          ceiling(prediction) >= 37 & ceiling(prediction) <= 40
          ,ifelse(is.na(residual_prediction_tb), 0, residual_prediction_tb)
          ,ifelse(is.na(residual_prediction_ta), 0, residual_prediction_ta)
        )
      )
  ) |>
  select(
    -residual_prediction_pr
    ,-residual_prediction_tb
    ,-residual_prediction_ta
  ) |>
  left_join(
    read_csv(
        "inst/extdata/Sample_annotation.csv"
        ,show_col_types = FALSE
      ) |>
      mutate(ID = Sample_ID) |>
      select(ID, Sample_accession)
    ,by = join_by(Sample_accession)
  ) |>
  select(-Sample_accession)
```

```{r ResComb-GA prediction - Results in dep, eval=FALSE, include=FALSE}
ga_res_comb_est_set_predictions_dep_file <-
  list.files(
    "output"
    ,pattern = "^predictions.csv"
    ,full.names = TRUE
  )

if(length(ga_res_comb_est_set_predictions_dep_file) == 1){
  ga_res_comb_est_set_predictions_dep_file |>
    file.rename(
      ga_res_comb_est_set_predictions_dep_file |>
        str_replace_all("output/","output/pcd1_sub8_")
    )
}

ga_res_comb_est_set_predictions_dep <-
  read_csv(
    "output/pcd1_sub8_predictions.csv"
    ,show_col_types = FALSE
  ) |>
  mutate(rowname = Sample_IDs) |>
  column_to_rownames(var = "rowname")
```

```{r ResComb-GA prediction - Comp dev and dep, eval=FALSE, include=FALSE}
ga_res_comb_est_set_predictions_dep |>
  left_join(
    ga_res_comb_est_set_predictions_dev
    ,by = join_by(ID)
  ) |>
  mutate(diff = GA_prediction - prediction) |>
  filter(diff != 0) |>
  filter(abs(diff) > 0.00001)
```

```{bash ResComb-GA prediction - Push command, eval=FALSE, include=FALSE}
cd pcd1_sub8
docker buildx build --platform linux/amd64 -t docker.synapse.org/syn62407322/ga-res-comb-est:v1.0 --push .
cd ..
```










